{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for ingesting data from BPA REST API\n",
    "\n",
    "BPA API is built on CKAN - docs: https://docs.ckan.org/en/latest/api/\n",
    "\n",
    "Base URL for BPA is https://data.bioplatforms.com/api/3\n",
    "\n",
    "Ingest using wildcard search: [`https://data.bioplatforms.com/api/3/action/package_search?q=*:*&rows=1000`](https://data.bioplatforms.com/api/3/action/package_search?q=*:*&rows=1000)\n",
    "\n",
    "Note script will need to paginate 1000 records at a time (48,438 results)\n",
    "\n",
    "### To Do\n",
    "\n",
    "[x] Add ignore list for JSON data to skip\n",
    "\n",
    "[x] Map to DwC and extendions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total: 48812 50\n",
      "api_dataset size 48812\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sw/fck698gd1wv_56c640ksf4r0009m0h/T/ipykernel_43430/4149659864.py:19: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead.\n",
      "  df = json_normalize(api_dataset) # transformation to dataframe via normalize function\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 48812 entries, 0 to 48811\n",
      "Columns: 976 entries, library_ng_ul to license_url\n",
      "dtypes: bool(3), int64(2), object(971)\n",
      "memory usage: 362.5+ MB\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "from pandas.io.json import json_normalize \n",
    "import numpy as nmp\n",
    "\n",
    "url = \"https://data.bioplatforms.com/api/3/action/package_search?q=*:*&rows=\"\n",
    "pagesize = 1000\n",
    "total = requests.get(f\"{url}0\").json()['result']['count']\n",
    "totalPages = int(nmp.ceil(total / pagesize)) + 1 # range is \"exclusive\" so needs an extra 1\n",
    "print(\"Total:\", total, totalPages)\n",
    "api_dataset = requests.get(f\"{url}{pagesize}\").json()['result']['results']\n",
    "\n",
    "for page in range (1, totalPages):\n",
    "    # print(\"Paginating:\", page, \"start=\", (page * pagesize))\n",
    "    response = requests.get(f\"{url}{pagesize}&start={(page * pagesize)}\").json()['result']['results']\n",
    "    api_dataset.extend(response)\n",
    "\n",
    "print(\"api_dataset size\", len(api_dataset))\n",
    "df = json_normalize(api_dataset) # transformation to dataframe via normalize function\n",
    "\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library_ng_ul</th>\n",
       "      <th>dna_treatment</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>sequence_data_type</th>\n",
       "      <th>genomic_material_preparation_process</th>\n",
       "      <th>library_type</th>\n",
       "      <th>nagoya_protocol_permit_number</th>\n",
       "      <th>metadata_modified</th>\n",
       "      <th>title</th>\n",
       "      <th>base_url</th>\n",
       "      <th>...</th>\n",
       "      <th>library_id_description</th>\n",
       "      <th>dataset_context</th>\n",
       "      <th>reference_genome_link</th>\n",
       "      <th>update_frequency</th>\n",
       "      <th>theme</th>\n",
       "      <th>data_homepage</th>\n",
       "      <th>language</th>\n",
       "      <th>data_portal</th>\n",
       "      <th>doi</th>\n",
       "      <th>license_url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>75.2</td>\n",
       "      <td>Shearing on Diagenode megaruptor-3, FFPE repai...</td>\n",
       "      <td>Genomics WA</td>\n",
       "      <td>pacbio-hifi</td>\n",
       "      <td>HMW DNA Prep https://www.protocols.io/view/sca...</td>\n",
       "      <td>Long read Hi Fi library with custom steps</td>\n",
       "      <td>DNP Reference: 2020/21-1</td>\n",
       "      <td>2022-07-20T23:02:57.994196</td>\n",
       "      <td>GAP PacBio HiFi, Reference genomes, 376315,</td>\n",
       "      <td>https://downloads-qcif.bioplatforms.com/bpa/pl...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>48.9</td>\n",
       "      <td>Shearing on Diagenode megaruptor-3, FFPE repai...</td>\n",
       "      <td>Genomics WA</td>\n",
       "      <td>pacbio-hifi</td>\n",
       "      <td>HMW DNA Prep https://www.protocols.io/view/sca...</td>\n",
       "      <td>Long read Hi Fi library with custom steps</td>\n",
       "      <td>SL100305</td>\n",
       "      <td>2022-07-20T23:02:57.488354</td>\n",
       "      <td>GAP PacBio HiFi, Reference genomes, 376318,</td>\n",
       "      <td>https://downloads-qcif.bioplatforms.com/bpa/pl...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>81.3</td>\n",
       "      <td>Shearing on Diagenode megaruptor-3, FFPE repai...</td>\n",
       "      <td>Genomics WA</td>\n",
       "      <td>pacbio-hifi</td>\n",
       "      <td>HMW DNA Prep https://www.protocols.io/view/sca...</td>\n",
       "      <td>Long read Hi Fi library with custom steps</td>\n",
       "      <td>SL100305</td>\n",
       "      <td>2022-07-20T23:02:56.882937</td>\n",
       "      <td>GAP PacBio HiFi, Reference genomes, 376316,</td>\n",
       "      <td>https://downloads-qcif.bioplatforms.com/bpa/pl...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>48.9</td>\n",
       "      <td>Shearing on Diagenode megaruptor-3, FFPE repai...</td>\n",
       "      <td>Genomics WA</td>\n",
       "      <td>pacbio-hifi</td>\n",
       "      <td>HMW DNA Prep https://www.protocols.io/view/sca...</td>\n",
       "      <td>Long read Hi Fi library with custom steps</td>\n",
       "      <td>SL100305</td>\n",
       "      <td>2022-07-20T23:02:56.423997</td>\n",
       "      <td>GAP PacBio HiFi, Reference genomes, 376318,</td>\n",
       "      <td>https://downloads-qcif.bioplatforms.com/bpa/pl...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>75.2</td>\n",
       "      <td>Shearing on Diagenode megaruptor-3, FFPE repai...</td>\n",
       "      <td>Genomics WA</td>\n",
       "      <td>pacbio-hifi</td>\n",
       "      <td>HMW DNA Prep https://www.protocols.io/view/sca...</td>\n",
       "      <td>Long read Hi Fi library with custom steps</td>\n",
       "      <td>DNP Reference: 2020/21-1</td>\n",
       "      <td>2022-07-20T23:02:55.889979</td>\n",
       "      <td>GAP PacBio HiFi, Reference genomes, 376315,</td>\n",
       "      <td>https://downloads-qcif.bioplatforms.com/bpa/pl...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 976 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  library_ng_ul                                      dna_treatment  \\\n",
       "0          75.2  Shearing on Diagenode megaruptor-3, FFPE repai...   \n",
       "1          48.9  Shearing on Diagenode megaruptor-3, FFPE repai...   \n",
       "2          81.3  Shearing on Diagenode megaruptor-3, FFPE repai...   \n",
       "3          48.9  Shearing on Diagenode megaruptor-3, FFPE repai...   \n",
       "4          75.2  Shearing on Diagenode megaruptor-3, FFPE repai...   \n",
       "\n",
       "  sequencing_facility sequence_data_type  \\\n",
       "0         Genomics WA        pacbio-hifi   \n",
       "1         Genomics WA        pacbio-hifi   \n",
       "2         Genomics WA        pacbio-hifi   \n",
       "3         Genomics WA        pacbio-hifi   \n",
       "4         Genomics WA        pacbio-hifi   \n",
       "\n",
       "                genomic_material_preparation_process  \\\n",
       "0  HMW DNA Prep https://www.protocols.io/view/sca...   \n",
       "1  HMW DNA Prep https://www.protocols.io/view/sca...   \n",
       "2  HMW DNA Prep https://www.protocols.io/view/sca...   \n",
       "3  HMW DNA Prep https://www.protocols.io/view/sca...   \n",
       "4  HMW DNA Prep https://www.protocols.io/view/sca...   \n",
       "\n",
       "                                library_type nagoya_protocol_permit_number  \\\n",
       "0  Long read Hi Fi library with custom steps      DNP Reference: 2020/21-1   \n",
       "1  Long read Hi Fi library with custom steps                      SL100305   \n",
       "2  Long read Hi Fi library with custom steps                      SL100305   \n",
       "3  Long read Hi Fi library with custom steps                      SL100305   \n",
       "4  Long read Hi Fi library with custom steps      DNP Reference: 2020/21-1   \n",
       "\n",
       "            metadata_modified                                        title  \\\n",
       "0  2022-07-20T23:02:57.994196  GAP PacBio HiFi, Reference genomes, 376315,   \n",
       "1  2022-07-20T23:02:57.488354  GAP PacBio HiFi, Reference genomes, 376318,   \n",
       "2  2022-07-20T23:02:56.882937  GAP PacBio HiFi, Reference genomes, 376316,   \n",
       "3  2022-07-20T23:02:56.423997  GAP PacBio HiFi, Reference genomes, 376318,   \n",
       "4  2022-07-20T23:02:55.889979  GAP PacBio HiFi, Reference genomes, 376315,   \n",
       "\n",
       "                                            base_url  ...  \\\n",
       "0  https://downloads-qcif.bioplatforms.com/bpa/pl...  ...   \n",
       "1  https://downloads-qcif.bioplatforms.com/bpa/pl...  ...   \n",
       "2  https://downloads-qcif.bioplatforms.com/bpa/pl...  ...   \n",
       "3  https://downloads-qcif.bioplatforms.com/bpa/pl...  ...   \n",
       "4  https://downloads-qcif.bioplatforms.com/bpa/pl...  ...   \n",
       "\n",
       "  library_id_description dataset_context reference_genome_link  \\\n",
       "0                    NaN             NaN                   NaN   \n",
       "1                    NaN             NaN                   NaN   \n",
       "2                    NaN             NaN                   NaN   \n",
       "3                    NaN             NaN                   NaN   \n",
       "4                    NaN             NaN                   NaN   \n",
       "\n",
       "  update_frequency theme data_homepage language data_portal  doi license_url  \n",
       "0              NaN   NaN           NaN      NaN         NaN  NaN         NaN  \n",
       "1              NaN   NaN           NaN      NaN         NaN  NaN         NaN  \n",
       "2              NaN   NaN           NaN      NaN         NaN  NaN         NaN  \n",
       "3              NaN   NaN           NaN      NaN         NaN  NaN         NaN  \n",
       "4              NaN   NaN           NaN      NaN         NaN  NaN         NaN  \n",
       "\n",
       "[5 rows x 976 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a list of the fields with most number of rows containing data\n",
    "# df.notna().sum().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 48812 entries, 0 to 48811\n",
      "Columns: 168 entries, dna_treatment to sample_name\n",
      "dtypes: bool(3), int64(2), object(163)\n",
      "memory usage: 61.6+ MB\n"
     ]
    }
   ],
   "source": [
    "fields_to_keep = [\n",
    "    \"access_control_date\",\n",
    "    \"access_control_mode\",\n",
    "    \"access_control_reason\",\n",
    "    \"access_rights\",\n",
    "    \"ala_specimen_url\",\n",
    "    \"amplicon\",\n",
    "    \"analysis_software_version\",\n",
    "    \"associated_media\",\n",
    "    \"barcode_id\",\n",
    "    \"base_url\",\n",
    "    \"biotic_relationship\",\n",
    "    \"birth_date\",\n",
    "    \"bpa_dataset_id\",\n",
    "    \"bpa_library_id\",\n",
    "    \"bpa_sample_id\",\n",
    "    \"citation\",\n",
    "    \"class\",\n",
    "    \"coastal_id\",\n",
    "    \"collection_date\",\n",
    "    \"collection_method\",\n",
    "    \"collector\",\n",
    "    \"collector_sample_id\",\n",
    "    \"color\",\n",
    "    \"comments\",\n",
    "    \"common_name\",\n",
    "    \"coord_uncertainty_metres\",\n",
    "    \"country\",\n",
    "    \"creator_user_id\",\n",
    "    \"custodian\",\n",
    "    \"dataset_id\",\n",
    "    \"dataset_url\",\n",
    "    \"data_custodian\",\n",
    "    \"data_generated\",\n",
    "    \"data_type\",\n",
    "    \"date_data_published\",\n",
    "    \"date_of_transfer\",\n",
    "    \"date_of_transfer_to_archive\",\n",
    "    \"date_since_change_in_land_use\",\n",
    "    \"ddrad_dataset_ids\",\n",
    "    \"death_date\",\n",
    "    \"description\",\n",
    "    \"dna_extraction_date\",\n",
    "    \"dna_treatment\",\n",
    "    \"download\",\n",
    "    \"facility\",\n",
    "    \"facility_sample_id\",\n",
    "    \"family\",\n",
    "    \"fire\",\n",
    "    \"fire_intensity_if_known\",\n",
    "    \"flooding\",\n",
    "    \"flow_id\",\n",
    "    \"folder_name\",\n",
    "    \"fouling_organisms\",\n",
    "    \"funding_agency\",\n",
    "    \"genome_data\",\n",
    "    \"genome_dataset_ids\",\n",
    "    \"genomic_material_associated_references\",\n",
    "    \"genus\",\n",
    "    \"geo_loc_name\",\n",
    "    \"grazing_number\",\n",
    "    \"groups\",\n",
    "    \"herbarium_code\",\n",
    "    \"host_state\",\n",
    "    \"host_type\",\n",
    "    \"id\",\n",
    "    \"identified_by\",\n",
    "    \"id_vetting_by\",\n",
    "    \"imos_site_code\",\n",
    "    \"information\",\n",
    "    \"institution_name\",\n",
    "    \"isopen\",\n",
    "    \"latitude\",\n",
    "    \"lat_lon\",\n",
    "    \"license_id\",\n",
    "    \"license_title\",\n",
    "    \"life_stage\",\n",
    "    \"living_collections_catalog_number\",\n",
    "    \"living_collections_material_sample_rna\",\n",
    "    \"living_collections_recorded_by\",\n",
    "    \"living_collections_record_number\",\n",
    "    \"location_text\",\n",
    "    \"longitude\",\n",
    "    \"metadata_created\",\n",
    "    \"metadata_modified\",\n",
    "    \"nagoya_protocol_permit_number\",\n",
    "    \"name\",\n",
    "    \"ncbi_bioproject_accession\",\n",
    "    \"ncbi_biosample_accession\",\n",
    "    \"notes\",\n",
    "    \"nrs_sample_code\",\n",
    "    \"nrs_trip_code\",\n",
    "    \"num_resources\",\n",
    "    \"num_tags\",\n",
    "    \"order\",\n",
    "    \"organization.approval_status\",\n",
    "    \"organization.created\",\n",
    "    \"organization.description\",\n",
    "    \"organization.id\",\n",
    "    \"organization.image_url\",\n",
    "    \"organization.is_organization\",\n",
    "    \"organization.name\",\n",
    "    \"organization.revision_id\",\n",
    "    \"organization.state\",\n",
    "    \"organization.title\",\n",
    "    \"organization.type\",\n",
    "    \"owner_org\",\n",
    "    \"phylum\",\n",
    "    \"plant_id\",\n",
    "    \"private\",\n",
    "    \"project_aim\",\n",
    "    \"reads\",\n",
    "    \"read_length\",\n",
    "    \"relationships_as_object\",\n",
    "    \"relationships_as_subject\",\n",
    "    \"resources\",\n",
    "    \"resource_permissions\",\n",
    "    \"revision_id\",\n",
    "    \"sample_attribution\",\n",
    "    \"sample_extraction_id\",\n",
    "    \"sample_id\",\n",
    "    \"sample_name\",\n",
    "    \"sample_submission_date\",\n",
    "    \"sample_submitter\",\n",
    "    \"sample_type\",\n",
    "    \"scientific_name\",\n",
    "    \"scientific_name_authorship\",\n",
    "    \"scientific_name_notes\",\n",
    "    \"sequencer\",\n",
    "    \"sequence_data_type\",\n",
    "    \"sequence_length\",\n",
    "    \"sequencing_facility\",\n",
    "    \"sequencing_platform\",\n",
    "    \"sequencing_run_number\",\n",
    "    \"sex\",\n",
    "    \"spatial\",\n",
    "    \"species\",\n",
    "    \"species_name\",\n",
    "    \"specific_host\",\n",
    "    \"state\",\n",
    "    \"state_or_region\",\n",
    "    \"subspecies_or_variant\",\n",
    "    \"synonyms\",\n",
    "    \"tags\",\n",
    "    \"target\",\n",
    "    \"taxonomic_group\",\n",
    "    \"taxon_id\",\n",
    "    \"texture\",\n",
    "    \"ticket\",\n",
    "    \"tissue_collection\",\n",
    "    \"tissue_preservation\",\n",
    "    \"tissue_type\",\n",
    "    \"title\",\n",
    "    \"touching_organisms\",\n",
    "    \"type\",\n",
    "    \"type_status\",\n",
    "    \"url\",\n",
    "    \"utc_time_sampled\",\n",
    "    \"vegetation_type\",\n",
    "    \"voucher_herbarium_catalog_number\",\n",
    "    \"voucher_herbarium_collector_id\",\n",
    "    \"voucher_herbarium_event_date\",\n",
    "    \"voucher_herbarium_recorded_by\",\n",
    "    \"voucher_herbarium_record_number\",\n",
    "    \"voucher_number\",\n",
    "    \"voucher_or_tissue_number\",\n",
    "    \"voyage_code\",\n",
    "    \"voyage_survey_link\",\n",
    "    \"wild_captive\",\n",
    "]\n",
    "df.drop(columns=df.columns.difference(fields_to_keep), inplace=True)\n",
    "df.info()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Cleanup field names and map to DwC where possible\n",
    "field_mapping = {\n",
    "    # DwC fields\n",
    "    'access_rights': 'accessRights',\n",
    "    'date_of_transfer': 'eventDate',\n",
    "    'base_url': 'geneticAccessionURI', \n",
    "    'citation': 'MaterialCitation',\n",
    "    'collection_method': 'measurementType',\n",
    "    'collector': 'recordedBy',\n",
    "    'collector_sample_id': 'materialSampleID',\n",
    "    'common_name': 'vernacularName',\n",
    "    'coord_uncertainty_metres': 'coordinateUncertaintyInMeters',\n",
    "    'dataset_id': 'datasetID',\n",
    "    'geo_loc_name': 'locality',\n",
    "    'id': 'occurrenceID',\n",
    "    'identified_by': 'identifiedBy',\n",
    "    'latitude': 'decimalLatitude',\n",
    "    'life_stage': 'lifeStage',\n",
    "    'living_collections_catalog_number': 'otherCatalogNumbers',\n",
    "    'living_collections_recorded_by': 'recordedBy',\n",
    "    'longitude': 'decimalLongitude',\n",
    "    'resources': 'associatedSequences',\n",
    "    'sample_id': 'materialSampleID',\n",
    "    # 'scientific_name': 'acceptedScientificName',\n",
    "    'scientific_name_authorship': 'scientificNameAuthorship',\n",
    "    'species': 'specificEpithet',\n",
    "    # 'species_name': 'scientificiName',\n",
    "    'state': 'stateProvince',\n",
    "    'taxon_id': 'taxonID',\n",
    "    'tissue_preservation': 'preparations',\n",
    "    'type_status': 'typeStatus',\n",
    "    'voucher_herbarium_catalog_number': 'otherCatalogNumbers',\n",
    "    'voucher_herbarium_record_number': 'dwciri:recordNumber',\n",
    "    'voucher_herbarium_recorded_by': 'dwciri:recordedBy',\n",
    "    'wild_captive': 'degreeOfEstablishment',\n",
    "    # Unmapped fields\n",
    "    'access_control_date': 'bpa_access_control_date',\n",
    "    'access_control_mode': 'bpa_access_control_mode',\n",
    "    'access_control_reason': 'bpa_access_control_reason',\n",
    "    'ala_specimen_url': 'bpa_ala_specimen_url',\n",
    "    'amplicon': 'bpa_amplicon',\n",
    "    'analysis_software_version': 'bpa_analysis_software_version',\n",
    "    'associated_media': 'bpa_associated_media',\n",
    "    'barcode_id': 'bpa_barcode_id',\n",
    "    'base_url': 'bpa_base_url',\n",
    "    'biotic_relationship': 'bpa_biotic_relationship',\n",
    "    'birth_date': 'bpa_birth_date',\n",
    "    'bpa_dataset_id': 'bpa_bpa_dataset_id',\n",
    "    'bpa_library_id': 'bpa_bpa_library_id',\n",
    "    'bpa_sample_id': 'bpa_bpa_sample_id',\n",
    "    'coastal_id': 'bpa_coastal_id',\n",
    "    'herbarium_code': 'bpa_herbarium_code',\n",
    "    'collection_date': 'bpa_collection_date',\n",
    "    'color': 'bpa_color',\n",
    "    'comments': 'bpa_comments',\n",
    "    'creator_user_id': 'bpa_creator_user_id',\n",
    "    'custodian': 'bpa_custodian',\n",
    "    'data_custodian': 'bpa_data_custodian',\n",
    "    'data_generated': 'bpa_data_generated',\n",
    "    'data_type': 'bpa_data_type',\n",
    "    'dataset_url': 'bpa_dataset_url',\n",
    "    'date_data_published': 'bpa_date_data_published',\n",
    "    'date_of_transfer_to_archive': 'bpa_date_of_transfer_to_archive',\n",
    "    'date_since_change_in_land_use': 'bpa_date_since_change_in_land_use',\n",
    "    'ddrad_dataset_ids': 'bpa_ddrad_dataset_ids',\n",
    "    'death_date': 'bpa_death_date',\n",
    "    'description': 'bpa_description',\n",
    "    'dna_extraction_date': 'bpa_dna_extraction_date',\n",
    "    'dna_treatment': 'bpa_dna_treatment',\n",
    "    'download': 'bpa_download',\n",
    "    'facility': 'bpa_facility',\n",
    "    'facility_sample_id': 'bpa_facility_sample_id',\n",
    "    'fire': 'bpa_fire',\n",
    "    'fire_intensity_if_known': 'bpa_fire_intensity_if_known',\n",
    "    'flooding': 'bpa_flooding',\n",
    "    'flow_id': 'bpa_flow_id',\n",
    "    'folder_name': 'bpa_folder_name',\n",
    "    'fouling_organisms': 'bpa_fouling_organisms',\n",
    "    'funding_agency': 'bpa_funding_agency',\n",
    "    'genome_data': 'bpa_genome_data',\n",
    "    'genome_dataset_ids': 'bpa_genome_dataset_ids',\n",
    "    'genomic_material_associated_references': 'bpa_genomic_material_associated_references',\n",
    "    'grazing_number': 'bpa_grazing_number',\n",
    "    'groups': 'bpa_groups',\n",
    "    'host_state': 'bpa_host_state',\n",
    "    'host_type': 'bpa_host_type',\n",
    "    'id_vetting_by': 'bpa_id_vetting_by',\n",
    "    'imos_site_code': 'bpa_imos_site_code',\n",
    "    'information': 'bpa_information',\n",
    "    'institution_name': 'bpa_institution_name',\n",
    "    'isopen': 'bpa_isopen',\n",
    "    'lat_lon': 'bpa_lat_lon',\n",
    "    'license_id': 'bpa_license_id',\n",
    "    'license_title': 'bpa_license_title',\n",
    "    'living_collections_material_sample_rna': 'bpa_living_collections_material_sample_rna',\n",
    "    'living_collections_record_number': 'bpa_living_collections_record_number',\n",
    "    'location_text': 'bpa_location_text',\n",
    "    'metadata_created': 'bpa_metadata_created',\n",
    "    'metadata_modified': 'bpa_metadata_modified',\n",
    "    'nagoya_protocol_permit_number': 'bpa_nagoya_protocol_permit_number',\n",
    "    'name': 'bpa_name',\n",
    "    'ncbi_bioproject_accession': 'ncbi_bioproject',\n",
    "    'ncbi_biosample_accession': 'ncbi_biosample',\n",
    "    'notes': 'bpa_notes',\n",
    "    'nrs_sample_code': 'bpa_nrs_sample_code',\n",
    "    'nrs_trip_code': 'bpa_nrs_trip_code',\n",
    "    'num_resources': 'bpa_num_resources',\n",
    "    'num_tags': 'bpa_num_tags',\n",
    "    'order': 'bpa_order',\n",
    "    'organization.approval_status': 'bpa_organization_approval_status',\n",
    "    'organization.created': 'bpa_organization_created',\n",
    "    'organization.description': 'bpa_organization_description',\n",
    "    'organization.id': 'bpa_organization_id',\n",
    "    'organization.image_url': 'bpa_organization_image_url',\n",
    "    'organization.is_organization': 'bpa_organization_is_organization',\n",
    "    'organization.name': 'bpa_organization_name',\n",
    "    'organization.revision_id': 'bpa_organization_revision_id',\n",
    "    'organization.state': 'bpa_organization_state',\n",
    "    'organization.title': 'bpa_organization_title',\n",
    "    'organization.type': 'bpa_organization_type',\n",
    "    'owner_org': 'bpa_owner_org',\n",
    "    'plant_id': 'bpa_plant_id',\n",
    "    'private': 'bpa_private',\n",
    "    'project_aim': 'bpa_project_aim',\n",
    "    'read_length': 'bpa_read_length',\n",
    "    'reads': 'bpa_reads',\n",
    "    'relationships_as_object': 'bpa_relationships_as_object',\n",
    "    'relationships_as_subject': 'bpa_relationships_as_subject',\n",
    "    'resource_permissions': 'bpa_resource_permissions',\n",
    "    'revision_id': 'bpa_revision_id',\n",
    "    'sample_attribution': 'bpa_sample_attribution',\n",
    "    'sample_extraction_id': 'bpa_sample_extraction_id',\n",
    "    'sample_name': 'bpa_sample_name',\n",
    "    'sample_submission_date': 'bpa_sample_submission_date',\n",
    "    'sample_submitter': 'bpa_sample_submitter',\n",
    "    'sample_type': 'bpa_sample_type',\n",
    "    'scientific_name_notes': 'bpa_scientific_name_notes',\n",
    "    'sequence_data_type': 'bpa_sequence_data_type',\n",
    "    'sequence_length': 'bpa_sequence_length',\n",
    "    'sequencer': 'bpa_sequencer',\n",
    "    'sequencing_facility': 'bpa_sequencing_facility',\n",
    "    'sequencing_platform': 'bpa_sequencing_platform',\n",
    "    'sequencing_run_number': 'bpa_sequencing_run_number',\n",
    "    'spatial': 'bpa_spatial',\n",
    "    'species': 'bpa_species',\n",
    "    'state_or_region': 'bpa_state_or_region',\n",
    "    'subspecies_or_variant': 'bpa_subspecies_or_variant',\n",
    "    'synonyms': 'bpa_synonyms',\n",
    "    'tags': 'bpa_tags',\n",
    "    'target': 'bpa_target',\n",
    "    'taxonomic_group': 'bpa_taxonomic_group',\n",
    "    'texture': 'bpa_texture',\n",
    "    'ticket': 'bpa_ticket',\n",
    "    'tissue_collection': 'bpa_tissue_collection',\n",
    "    'tissue_type': 'bpa_tissue_type',\n",
    "    'title': 'bpa_title',\n",
    "    'touching_organisms': 'bpa_touching_organisms',\n",
    "    'type': 'bpa_type',\n",
    "    'url': 'bpa_url',\n",
    "    'utc_time_sampled': 'bpa_utc_time_sampled',\n",
    "    'vegetation_type': 'bpa_vegetation_type',\n",
    "    'voucher_herbarium_collector_id': 'bpa_voucher_herbarium_collector_id',\n",
    "    'voucher_herbarium_event_date': 'bpa_voucher_herbarium_event_date',\n",
    "    'voucher_number': 'bpa_voucher_number',\n",
    "    'voucher_or_tissue_number': 'bpa_voucher_or_tissue_number',\n",
    "    'voyage_code': 'bpa_voyage_code',\n",
    "    'voyage_survey_link': 'bpa_voyage_survey_link',\n",
    "}\n",
    "\n",
    "# copy these fields first\n",
    "df['bpa_id'] = df['id'] \n",
    "df['bpa_specific_host'] = df['specific_host'] \n",
    "df['bpa_species'] = df['species'] \n",
    "df['bpa_collection_date'] = df['collection_date'] \n",
    "df['bpa_date_of_transfer'] = df['date_of_transfer'] \n",
    "# df['acceptedScientificName'] = nmp.where(df.scientific_name.notna(), df.scientific_name, df.species_name)\n",
    "\n",
    "# Attempt to find a scientificName value \n",
    "def mergeNames(row):\n",
    "  name = nmp.nan\n",
    "  if (pd.notnull(row['scientific_name'])):\n",
    "    name = row['scientific_name']\n",
    "  elif (pd.notnull(row['species_name'])):\n",
    "    name = row['species_name']\n",
    "  elif (pd.notnull(row['genus']) and pd.notnull(row['species'])):\n",
    "    name = str(row['genus']) + ' ' + str(row['species'])\n",
    "  elif (pd.notnull(row['genus'])):\n",
    "    name = str(row['genus'])\n",
    "  elif (pd.notnull(row['species'])):\n",
    "    name = str(row['species'])\n",
    "  elif (pd.notnull(row['common_name'])):\n",
    "    name = row['common_name']\n",
    "  \n",
    "  return name\n",
    "\n",
    "df['scientificName'] = df.apply(mergeNames, axis=1)\n",
    "\n",
    "# do rename\n",
    "df.rename(\n",
    "  columns = field_mapping, \n",
    "  inplace = True\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "occurrenceID empty count 0\n",
      "scientificName 8338\n",
      "scientific_name 4959\n",
      "species_name 2029\n"
     ]
    }
   ],
   "source": [
    "print(\"occurrenceID empty count\", len(df[df['occurrenceID'].isna()]))\n",
    "print(\"scientificName\", len(df[df['scientificName'].notna()]))\n",
    "print(\"scientific_name\", len(df[df['scientific_name'].notna()]))\n",
    "print(\"species_name\", len(df[df['species_name'].notna()]))\n",
    "# print(\"acceptedScientificName\", len(df[df['acceptedScientificName'].notna()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"/data/arga-data/bpa_export.csv\", index=False)\n",
    "# dr18544 in http://collections-test.ala.org.au"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
