{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iBold/BOLD Systems sequence data ingestion script\n",
    "\n",
    "See https://www.boldsystems.org/index.php/Public_BINSearch?query=Australia for details on data\n",
    "See https://v3.boldsystems.org/index.php/resources/api?type=webservices#combined for API docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1 entries, 0 to 0\n",
      "Data columns (total 3 columns):\n",
      " #   Column                          Non-Null Count  Dtype \n",
      "---  ------                          --------------  ----- \n",
      " 0   @xsi:noNamespaceSchemaLocation  1 non-null      object\n",
      " 1   @xmlns:xsi                      1 non-null      object\n",
      " 2   record                          1 non-null      object\n",
      "dtypes: object(3)\n",
      "memory usage: 152.0+ bytes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>@xsi:noNamespaceSchemaLocation</th>\n",
       "      <th>@xmlns:xsi</th>\n",
       "      <th>record</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>http://v3.boldsystems.org/schemas/BOLDPublic_r...</td>\n",
       "      <td>http://www.w3.org/2001/XMLSchema-instance</td>\n",
       "      <td>[{'record_id': '967799', 'processid': 'BENT040...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      @xsi:noNamespaceSchemaLocation  \\\n",
       "0  http://v3.boldsystems.org/schemas/BOLDPublic_r...   \n",
       "\n",
       "                                  @xmlns:xsi  \\\n",
       "0  http://www.w3.org/2001/XMLSchema-instance   \n",
       "\n",
       "                                              record  \n",
       "0  [{'record_id': '967799', 'processid': 'BENT040...  "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pandas_read_xml as pdx\n",
    "from pandas_read_xml import flatten, fully_flatten, auto_separate_tables\n",
    "\n",
    "df = pdx.read_xml('http://v3.boldsystems.org/index.php/API_Public/combined?format=xml&geo=Australia', ['bold_records'])\n",
    "\n",
    "# df = pd.read_xml(\"http://v3.boldsystems.org/index.php/API_Public/combined?format=xml&geo=Australia&taxon=Reptilia\", dtype={'specimen_identifiers': object}) \n",
    "# df = pd.read_table(\"../data/bold_data.tsv\", dtype='object', parse_dates=['collectiondate']) \n",
    "# columns:\n",
    "# processid\tsampleid\trecordID\tcatalognum\tfieldnum\tinstitution_storing\tbin_uri\tphylum_taxID\tphylum_name\t\n",
    "# class_taxID\tclass_name\torder_taxID\torder_name\tfamily_taxID\tfamily_name\tsubfamily_taxID\tsubfamily_name\t\n",
    "# genus_taxID\tgenus_name\tspecies_taxID\tspecies_name\tsubspecies_taxID\tsubspecies_name\tidentification_provided_by\t\n",
    "# voucher_type\ttissue_type\tcollectors\tcollectiondate\tlifestage\tsex\treproduction\textrainfo\tnotes\tlat\tlon\t\n",
    "# coord_source\tcoord_accuracy\tcountry\tprovince\tregion\texactsite\n",
    "df.info()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 388 entries, 0 to 387\n",
      "Data columns (total 3 columns):\n",
      " #   Column                          Non-Null Count  Dtype \n",
      "---  ------                          --------------  ----- \n",
      " 0   @xsi:noNamespaceSchemaLocation  388 non-null    object\n",
      " 1   @xmlns:xsi                      388 non-null    object\n",
      " 2   record                          388 non-null    object\n",
      "dtypes: object(3)\n",
      "memory usage: 9.2+ KB\n"
     ]
    }
   ],
   "source": [
    "df = df.pipe(flatten)\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 388 entries, 0 to 387\n",
      "Data columns (total 1 columns):\n",
      " #   Column  Non-Null Count  Dtype \n",
      "---  ------  --------------  ----- \n",
      " 0   record  388 non-null    object\n",
      "dtypes: object(1)\n",
      "memory usage: 3.2+ KB\n"
     ]
    }
   ],
   "source": [
    "# df = df.pipe(flatten)\n",
    "df = df.drop('@xsi:noNamespaceSchemaLocation', axis=1)\n",
    "df = df.drop('@xmlns:xsi', axis=1)\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 701 entries, 0 to 700\n",
      "Data columns (total 55 columns):\n",
      " #   Column                                              Non-Null Count  Dtype \n",
      "---  ------                                              --------------  ----- \n",
      " 0   record|bin_uri                                      695 non-null    object\n",
      " 1   record|collection_event|collectiondate              635 non-null    object\n",
      " 2   record|collection_event|collectors                  472 non-null    object\n",
      " 3   record|collection_event|coordinates|coord_accuracy  2 non-null      object\n",
      " 4   record|collection_event|coordinates|coord_source    6 non-null      object\n",
      " 5   record|collection_event|coordinates|lat             642 non-null    object\n",
      " 6   record|collection_event|coordinates|lon             642 non-null    object\n",
      " 7   record|collection_event|country                     701 non-null    object\n",
      " 8   record|collection_event|exactsite                   164 non-null    object\n",
      " 9   record|collection_event|province                    696 non-null    object\n",
      " 10  record|collection_event|region                      34 non-null     object\n",
      " 11  record|notes                                        19 non-null     object\n",
      " 12  record|processid                                    701 non-null    object\n",
      " 13  record|record_id                                    701 non-null    object\n",
      " 14  record|sequences|sequence|genbank_accession         192 non-null    object\n",
      " 15  record|sequences|sequence|markercode                699 non-null    object\n",
      " 16  record|sequences|sequence|nucleotides               699 non-null    object\n",
      " 17  record|sequences|sequence|sequenceID                699 non-null    object\n",
      " 18  record|specimen_details|extrainfo                   57 non-null     object\n",
      " 19  record|specimen_details|lifestage                   139 non-null    object\n",
      " 20  record|specimen_details|reproduction                518 non-null    object\n",
      " 21  record|specimen_details|sex                         146 non-null    object\n",
      " 22  record|specimen_details|voucher_desc                0 non-null      object\n",
      " 23  record|specimen_details|voucher_type                166 non-null    object\n",
      " 24  record|specimen_identifiers|catalognum              652 non-null    object\n",
      " 25  record|specimen_identifiers|fieldnum                323 non-null    object\n",
      " 26  record|specimen_identifiers|institution_storing     701 non-null    object\n",
      " 27  record|specimen_identifiers|sampleid                701 non-null    object\n",
      " 28  record|specimen_imagery|media|caption               0 non-null      object\n",
      " 29  record|specimen_imagery|media|copyright             6 non-null      object\n",
      " 30  record|specimen_imagery|media|image_file            6 non-null      object\n",
      " 31  record|specimen_imagery|media|mediaID               6 non-null      object\n",
      " 32  record|specimen_imagery|media|metatags              0 non-null      object\n",
      " 33  record|taxonomy|class|taxon|name                    701 non-null    object\n",
      " 34  record|taxonomy|class|taxon|taxID                   701 non-null    object\n",
      " 35  record|taxonomy|family|taxon|name                   230 non-null    object\n",
      " 36  record|taxonomy|family|taxon|taxID                  230 non-null    object\n",
      " 37  record|taxonomy|genus|taxon|name                    229 non-null    object\n",
      " 38  record|taxonomy|genus|taxon|taxID                   229 non-null    object\n",
      " 39  record|taxonomy|identification_provided_by          470 non-null    object\n",
      " 40  record|taxonomy|order|taxon|name                    701 non-null    object\n",
      " 41  record|taxonomy|order|taxon|taxID                   701 non-null    object\n",
      " 42  record|taxonomy|phylum|taxon|name                   701 non-null    object\n",
      " 43  record|taxonomy|phylum|taxon|taxID                  701 non-null    object\n",
      " 44  record|taxonomy|species|taxon|name                  229 non-null    object\n",
      " 45  record|taxonomy|species|taxon|taxID                 229 non-null    object\n",
      " 46  record|taxonomy|subfamily|taxon|name                186 non-null    object\n",
      " 47  record|taxonomy|subfamily|taxon|taxID               186 non-null    object\n",
      " 48  record|tracefiles|read|direction                    640 non-null    object\n",
      " 49  record|tracefiles|read|markercode                   640 non-null    object\n",
      " 50  record|tracefiles|read|read_id                      640 non-null    object\n",
      " 51  record|tracefiles|read|run_date                     640 non-null    object\n",
      " 52  record|tracefiles|read|seq_primer                   640 non-null    object\n",
      " 53  record|tracefiles|read|sequencing_center            640 non-null    object\n",
      " 54  record|tracefiles|read|trace_link                   640 non-null    object\n",
      "dtypes: object(55)\n",
      "memory usage: 301.3+ KB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record|bin_uri</th>\n",
       "      <th>record|collection_event|collectiondate</th>\n",
       "      <th>record|collection_event|collectors</th>\n",
       "      <th>record|collection_event|coordinates|coord_accuracy</th>\n",
       "      <th>record|collection_event|coordinates|coord_source</th>\n",
       "      <th>record|collection_event|coordinates|lat</th>\n",
       "      <th>record|collection_event|coordinates|lon</th>\n",
       "      <th>record|collection_event|country</th>\n",
       "      <th>record|collection_event|exactsite</th>\n",
       "      <th>record|collection_event|province</th>\n",
       "      <th>...</th>\n",
       "      <th>record|taxonomy|species|taxon|taxID</th>\n",
       "      <th>record|taxonomy|subfamily|taxon|name</th>\n",
       "      <th>record|taxonomy|subfamily|taxon|taxID</th>\n",
       "      <th>record|tracefiles|read|direction</th>\n",
       "      <th>record|tracefiles|read|markercode</th>\n",
       "      <th>record|tracefiles|read|read_id</th>\n",
       "      <th>record|tracefiles|read|run_date</th>\n",
       "      <th>record|tracefiles|read|seq_primer</th>\n",
       "      <th>record|tracefiles|read|sequencing_center</th>\n",
       "      <th>record|tracefiles|read|trace_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOLD:ACF0415</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.6528</td>\n",
       "      <td>150.146</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>515104</td>\n",
       "      <td>Hydrophiinae</td>\n",
       "      <td>473453</td>\n",
       "      <td>F</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>5001846</td>\n",
       "      <td>2013-04-30T08:55:39</td>\n",
       "      <td>M13F</td>\n",
       "      <td>Biodiversity Institute of Ontario</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BOLD:ACF0415</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.6528</td>\n",
       "      <td>150.146</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>515104</td>\n",
       "      <td>Hydrophiinae</td>\n",
       "      <td>473453</td>\n",
       "      <td>R</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>5001867</td>\n",
       "      <td>2013-04-30T08:55:39</td>\n",
       "      <td>M13R</td>\n",
       "      <td>Biodiversity Institute of Ontario</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BOLD:ACE9617</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.4374</td>\n",
       "      <td>149.799</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>171558</td>\n",
       "      <td>Chelodininae</td>\n",
       "      <td>757233</td>\n",
       "      <td>F</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>5001848</td>\n",
       "      <td>2013-04-30T08:55:39</td>\n",
       "      <td>M13F</td>\n",
       "      <td>Biodiversity Institute of Ontario</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BOLD:ACE9617</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.4374</td>\n",
       "      <td>149.799</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>171558</td>\n",
       "      <td>Chelodininae</td>\n",
       "      <td>757233</td>\n",
       "      <td>R</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>5001869</td>\n",
       "      <td>2013-04-30T08:55:39</td>\n",
       "      <td>M13R</td>\n",
       "      <td>Biodiversity Institute of Ontario</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BOLD:ADI2959</td>\n",
       "      <td>2002-10-07</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>-15.6469</td>\n",
       "      <td>145.22</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Cooktown, ca. 20 km (by road) S of, Black Moun...</td>\n",
       "      <td>Queensland</td>\n",
       "      <td>...</td>\n",
       "      <td>783178</td>\n",
       "      <td>Eugongylinae</td>\n",
       "      <td>757230</td>\n",
       "      <td>f</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>10282446</td>\n",
       "      <td>2014-03-28T10:23:01</td>\n",
       "      <td>REPTBCf</td>\n",
       "      <td>Smithsonian Institution, Laboratories of Analy...</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  record|bin_uri record|collection_event|collectiondate  \\\n",
       "0   BOLD:ACF0415                             2010-10-17   \n",
       "1   BOLD:ACF0415                             2010-10-17   \n",
       "2   BOLD:ACE9617                             2010-10-17   \n",
       "3   BOLD:ACE9617                             2010-10-17   \n",
       "4   BOLD:ADI2959                             2002-10-07   \n",
       "\n",
       "  record|collection_event|collectors  \\\n",
       "0                           P.Hebert   \n",
       "1                           P.Hebert   \n",
       "2                           P.Hebert   \n",
       "3                           P.Hebert   \n",
       "4                               None   \n",
       "\n",
       "  record|collection_event|coordinates|coord_accuracy  \\\n",
       "0                                               None   \n",
       "1                                               None   \n",
       "2                                               None   \n",
       "3                                               None   \n",
       "4                                               None   \n",
       "\n",
       "  record|collection_event|coordinates|coord_source  \\\n",
       "0                                              GPS   \n",
       "1                                              GPS   \n",
       "2                                              GPS   \n",
       "3                                              GPS   \n",
       "4                                             None   \n",
       "\n",
       "  record|collection_event|coordinates|lat  \\\n",
       "0                                -35.6528   \n",
       "1                                -35.6528   \n",
       "2                                -35.4374   \n",
       "3                                -35.4374   \n",
       "4                                -15.6469   \n",
       "\n",
       "  record|collection_event|coordinates|lon record|collection_event|country  \\\n",
       "0                                 150.146                       Australia   \n",
       "1                                 150.146                       Australia   \n",
       "2                                 149.799                       Australia   \n",
       "3                                 149.799                       Australia   \n",
       "4                                  145.22                       Australia   \n",
       "\n",
       "                   record|collection_event|exactsite  \\\n",
       "0                                       Kings Hwy 52   \n",
       "1                                       Kings Hwy 52   \n",
       "2                                       Kings Hwy 52   \n",
       "3                                       Kings Hwy 52   \n",
       "4  Cooktown, ca. 20 km (by road) S of, Black Moun...   \n",
       "\n",
       "  record|collection_event|province  ... record|taxonomy|species|taxon|taxID  \\\n",
       "0                  New South Wales  ...                              515104   \n",
       "1                  New South Wales  ...                              515104   \n",
       "2                  New South Wales  ...                              171558   \n",
       "3                  New South Wales  ...                              171558   \n",
       "4                       Queensland  ...                              783178   \n",
       "\n",
       "  record|taxonomy|subfamily|taxon|name record|taxonomy|subfamily|taxon|taxID  \\\n",
       "0                         Hydrophiinae                                473453   \n",
       "1                         Hydrophiinae                                473453   \n",
       "2                         Chelodininae                                757233   \n",
       "3                         Chelodininae                                757233   \n",
       "4                         Eugongylinae                                757230   \n",
       "\n",
       "  record|tracefiles|read|direction record|tracefiles|read|markercode  \\\n",
       "0                                F                            COI-5P   \n",
       "1                                R                            COI-5P   \n",
       "2                                F                            COI-5P   \n",
       "3                                R                            COI-5P   \n",
       "4                                f                            COI-5P   \n",
       "\n",
       "  record|tracefiles|read|read_id record|tracefiles|read|run_date  \\\n",
       "0                        5001846             2013-04-30T08:55:39   \n",
       "1                        5001867             2013-04-30T08:55:39   \n",
       "2                        5001848             2013-04-30T08:55:39   \n",
       "3                        5001869             2013-04-30T08:55:39   \n",
       "4                       10282446             2014-03-28T10:23:01   \n",
       "\n",
       "  record|tracefiles|read|seq_primer  \\\n",
       "0                              M13F   \n",
       "1                              M13R   \n",
       "2                              M13F   \n",
       "3                              M13R   \n",
       "4                           REPTBCf   \n",
       "\n",
       "            record|tracefiles|read|sequencing_center  \\\n",
       "0                  Biodiversity Institute of Ontario   \n",
       "1                  Biodiversity Institute of Ontario   \n",
       "2                  Biodiversity Institute of Ontario   \n",
       "3                  Biodiversity Institute of Ontario   \n",
       "4  Smithsonian Institution, Laboratories of Analy...   \n",
       "\n",
       "                   record|tracefiles|read|trace_link  \n",
       "0  http://trace.boldsystems.org/traceIO/bold.org/...  \n",
       "1  http://trace.boldsystems.org/traceIO/bold.org/...  \n",
       "2  http://trace.boldsystems.org/traceIO/bold.org/...  \n",
       "3  http://trace.boldsystems.org/traceIO/bold.org/...  \n",
       "4  http://trace.boldsystems.org/traceIO/bold.org/...  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.pipe(fully_flatten)\n",
    "df.info()\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 701 entries, 0 to 700\n",
      "Data columns (total 56 columns):\n",
      " #   Column                                       Non-Null Count  Dtype \n",
      "---  ------                                       --------------  ----- \n",
      " 0   bold_bin_uri                                 695 non-null    object\n",
      " 1   eventDate                                    635 non-null    object\n",
      " 2   recordedBy                                   472 non-null    object\n",
      " 3   coordinatePrecision                          2 non-null      object\n",
      " 4   georeferenceSources                          6 non-null      object\n",
      " 5   decimalLatitude                              642 non-null    object\n",
      " 6   decimalLongitude                             642 non-null    object\n",
      " 7   country                                      701 non-null    object\n",
      " 8   locationRemarks                              164 non-null    object\n",
      " 9   stateProvince                                696 non-null    object\n",
      " 10  locality                                     34 non-null     object\n",
      " 11  fieldNotes                                   19 non-null     object\n",
      " 12  recordNumber                                 701 non-null    object\n",
      " 13  occurrenceID                                 701 non-null    object\n",
      " 14  record|sequences|sequence|genbank_accession  192 non-null    object\n",
      " 15  record|sequences|sequence|markercode         699 non-null    object\n",
      " 16  record|sequences|sequence|nucleotides        699 non-null    object\n",
      " 17  record|sequences|sequence|sequenceID         699 non-null    object\n",
      " 18  identificationRemarks                        57 non-null     object\n",
      " 19  lifeStage                                    139 non-null    object\n",
      " 20  bold_reproduction                            518 non-null    object\n",
      " 21  sex                                          146 non-null    object\n",
      " 22  dispositionProperty                          0 non-null      object\n",
      " 23  bold_voucher_type                            166 non-null    object\n",
      " 24  otherCatalogNumbers                          652 non-null    object\n",
      " 25  fieldNumber                                  323 non-null    object\n",
      " 26  ownerInstitutionCode                         701 non-null    object\n",
      " 27  materialSampleID                             701 non-null    object\n",
      " 28  record|specimen_imagery|media|caption        0 non-null      object\n",
      " 29  record|specimen_imagery|media|copyright      6 non-null      object\n",
      " 30  record|specimen_imagery|media|image_file     6 non-null      object\n",
      " 31  record|specimen_imagery|media|mediaID        6 non-null      object\n",
      " 32  record|specimen_imagery|media|metatags       0 non-null      object\n",
      " 33  class                                        701 non-null    object\n",
      " 34  classID                                      701 non-null    object\n",
      " 35  family                                       230 non-null    object\n",
      " 36  familyID                                     230 non-null    object\n",
      " 37  genus                                        229 non-null    object\n",
      " 38  genusID                                      229 non-null    object\n",
      " 39  identifiedBy                                 470 non-null    object\n",
      " 40  order                                        701 non-null    object\n",
      " 41  orderID                                      701 non-null    object\n",
      " 42  phylum                                       701 non-null    object\n",
      " 43  phylumID                                     701 non-null    object\n",
      " 44  species                                      229 non-null    object\n",
      " 45  speciesID                                    229 non-null    object\n",
      " 46  record|taxonomy|subfamily|taxon|name         186 non-null    object\n",
      " 47  record|taxonomy|subfamily|taxon|taxID        186 non-null    object\n",
      " 48  record|tracefiles|read|direction             640 non-null    object\n",
      " 49  record|tracefiles|read|markercode            640 non-null    object\n",
      " 50  record|tracefiles|read|read_id               640 non-null    object\n",
      " 51  record|tracefiles|read|run_date              640 non-null    object\n",
      " 52  record|tracefiles|read|seq_primer            640 non-null    object\n",
      " 53  record|tracefiles|read|sequencing_center     640 non-null    object\n",
      " 54  record|tracefiles|read|trace_link            640 non-null    object\n",
      " 55  scientificName                               229 non-null    object\n",
      "dtypes: object(56)\n",
      "memory usage: 306.8+ KB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bold_bin_uri</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>coordinatePrecision</th>\n",
       "      <th>georeferenceSources</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>country</th>\n",
       "      <th>locationRemarks</th>\n",
       "      <th>stateProvince</th>\n",
       "      <th>...</th>\n",
       "      <th>record|taxonomy|subfamily|taxon|name</th>\n",
       "      <th>record|taxonomy|subfamily|taxon|taxID</th>\n",
       "      <th>record|tracefiles|read|direction</th>\n",
       "      <th>record|tracefiles|read|markercode</th>\n",
       "      <th>record|tracefiles|read|read_id</th>\n",
       "      <th>record|tracefiles|read|run_date</th>\n",
       "      <th>record|tracefiles|read|seq_primer</th>\n",
       "      <th>record|tracefiles|read|sequencing_center</th>\n",
       "      <th>record|tracefiles|read|trace_link</th>\n",
       "      <th>scientificName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOLD:ACF0415</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.6528</td>\n",
       "      <td>150.146</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>Hydrophiinae</td>\n",
       "      <td>473453</td>\n",
       "      <td>F</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>5001846</td>\n",
       "      <td>2013-04-30T08:55:39</td>\n",
       "      <td>M13F</td>\n",
       "      <td>Biodiversity Institute of Ontario</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "      <td>Pseudechis porphyriacus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BOLD:ACF0415</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.6528</td>\n",
       "      <td>150.146</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>Hydrophiinae</td>\n",
       "      <td>473453</td>\n",
       "      <td>R</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>5001867</td>\n",
       "      <td>2013-04-30T08:55:39</td>\n",
       "      <td>M13R</td>\n",
       "      <td>Biodiversity Institute of Ontario</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "      <td>Pseudechis porphyriacus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BOLD:ACE9617</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.4374</td>\n",
       "      <td>149.799</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>Chelodininae</td>\n",
       "      <td>757233</td>\n",
       "      <td>F</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>5001848</td>\n",
       "      <td>2013-04-30T08:55:39</td>\n",
       "      <td>M13F</td>\n",
       "      <td>Biodiversity Institute of Ontario</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "      <td>Chelodina longicollis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BOLD:ACE9617</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.4374</td>\n",
       "      <td>149.799</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>Chelodininae</td>\n",
       "      <td>757233</td>\n",
       "      <td>R</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>5001869</td>\n",
       "      <td>2013-04-30T08:55:39</td>\n",
       "      <td>M13R</td>\n",
       "      <td>Biodiversity Institute of Ontario</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "      <td>Chelodina longicollis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BOLD:ADI2959</td>\n",
       "      <td>2002-10-07</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>-15.6469</td>\n",
       "      <td>145.22</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Cooktown, ca. 20 km (by road) S of, Black Moun...</td>\n",
       "      <td>Queensland</td>\n",
       "      <td>...</td>\n",
       "      <td>Eugongylinae</td>\n",
       "      <td>757230</td>\n",
       "      <td>f</td>\n",
       "      <td>COI-5P</td>\n",
       "      <td>10282446</td>\n",
       "      <td>2014-03-28T10:23:01</td>\n",
       "      <td>REPTBCf</td>\n",
       "      <td>Smithsonian Institution, Laboratories of Analy...</td>\n",
       "      <td>http://trace.boldsystems.org/traceIO/bold.org/...</td>\n",
       "      <td>Liburnascincus scirtetis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   bold_bin_uri   eventDate recordedBy coordinatePrecision  \\\n",
       "0  BOLD:ACF0415  2010-10-17   P.Hebert                None   \n",
       "1  BOLD:ACF0415  2010-10-17   P.Hebert                None   \n",
       "2  BOLD:ACE9617  2010-10-17   P.Hebert                None   \n",
       "3  BOLD:ACE9617  2010-10-17   P.Hebert                None   \n",
       "4  BOLD:ADI2959  2002-10-07       None                None   \n",
       "\n",
       "  georeferenceSources decimalLatitude decimalLongitude    country  \\\n",
       "0                 GPS        -35.6528          150.146  Australia   \n",
       "1                 GPS        -35.6528          150.146  Australia   \n",
       "2                 GPS        -35.4374          149.799  Australia   \n",
       "3                 GPS        -35.4374          149.799  Australia   \n",
       "4                None        -15.6469           145.22  Australia   \n",
       "\n",
       "                                     locationRemarks    stateProvince  ...  \\\n",
       "0                                       Kings Hwy 52  New South Wales  ...   \n",
       "1                                       Kings Hwy 52  New South Wales  ...   \n",
       "2                                       Kings Hwy 52  New South Wales  ...   \n",
       "3                                       Kings Hwy 52  New South Wales  ...   \n",
       "4  Cooktown, ca. 20 km (by road) S of, Black Moun...       Queensland  ...   \n",
       "\n",
       "  record|taxonomy|subfamily|taxon|name record|taxonomy|subfamily|taxon|taxID  \\\n",
       "0                         Hydrophiinae                                473453   \n",
       "1                         Hydrophiinae                                473453   \n",
       "2                         Chelodininae                                757233   \n",
       "3                         Chelodininae                                757233   \n",
       "4                         Eugongylinae                                757230   \n",
       "\n",
       "  record|tracefiles|read|direction record|tracefiles|read|markercode  \\\n",
       "0                                F                            COI-5P   \n",
       "1                                R                            COI-5P   \n",
       "2                                F                            COI-5P   \n",
       "3                                R                            COI-5P   \n",
       "4                                f                            COI-5P   \n",
       "\n",
       "  record|tracefiles|read|read_id record|tracefiles|read|run_date  \\\n",
       "0                        5001846             2013-04-30T08:55:39   \n",
       "1                        5001867             2013-04-30T08:55:39   \n",
       "2                        5001848             2013-04-30T08:55:39   \n",
       "3                        5001869             2013-04-30T08:55:39   \n",
       "4                       10282446             2014-03-28T10:23:01   \n",
       "\n",
       "  record|tracefiles|read|seq_primer  \\\n",
       "0                              M13F   \n",
       "1                              M13R   \n",
       "2                              M13F   \n",
       "3                              M13R   \n",
       "4                           REPTBCf   \n",
       "\n",
       "            record|tracefiles|read|sequencing_center  \\\n",
       "0                  Biodiversity Institute of Ontario   \n",
       "1                  Biodiversity Institute of Ontario   \n",
       "2                  Biodiversity Institute of Ontario   \n",
       "3                  Biodiversity Institute of Ontario   \n",
       "4  Smithsonian Institution, Laboratories of Analy...   \n",
       "\n",
       "                   record|tracefiles|read|trace_link            scientificName  \n",
       "0  http://trace.boldsystems.org/traceIO/bold.org/...   Pseudechis porphyriacus  \n",
       "1  http://trace.boldsystems.org/traceIO/bold.org/...   Pseudechis porphyriacus  \n",
       "2  http://trace.boldsystems.org/traceIO/bold.org/...     Chelodina longicollis  \n",
       "3  http://trace.boldsystems.org/traceIO/bold.org/...     Chelodina longicollis  \n",
       "4  http://trace.boldsystems.org/traceIO/bold.org/...  Liburnascincus scirtetis  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cleanup field names and map to DwC where possible\n",
    "field_mapping = {\n",
    "    # DwC fields\n",
    "    'record|collection_event|collectiondate': 'eventDate',\n",
    "    'record|collection_event|collectors': 'recordedBy',\n",
    "    'record|collection_event|coordinates|coord_accuracy':'coordinatePrecision',\n",
    "    'record|collection_event|coordinates|coord_source':'georeferenceSources',\n",
    "    'record|collection_event|coordinates|lat':'decimalLatitude',\n",
    "    'record|collection_event|coordinates|lon':'decimalLongitude',\n",
    "    'record|collection_event|country':'country',\n",
    "    'record|collection_event|exactsite':'locationRemarks',\n",
    "    'record|collection_event|province':'stateProvince',\n",
    "    'record|collection_event|region':'locality',\n",
    "    'record|notes':'fieldNotes', # eventRemarks\n",
    "    'record|processid':'recordNumber',\n",
    "    'record|record_id':'occurrenceID',\n",
    "    'record|bin_uri':'bold_bin_uri',\n",
    "    # use df.apply to merge following fields into associatedSequences\n",
    "    # 'record|sequences|sequence|genbank_accession':'ncbi_accession',\n",
    "    # 'record|sequences|sequence|sequenceID':'',\n",
    "    # 'record|sequences|sequence|nucleotides':'',\n",
    "    'record|specimen_details|extrainfo':'identificationRemarks',\n",
    "    'record|specimen_details|lifestage':'lifeStage',\n",
    "    'record|specimen_details|reproduction':'bold_reproduction',\n",
    "    'record|specimen_details|sex':'sex',\n",
    "    'record|specimen_details|voucher_desc':'dispositionProperty',\n",
    "    'record|specimen_details|voucher_type':'bold_voucher_type',\n",
    "    'record|specimen_identifiers|catalognum':'otherCatalogNumbers',\n",
    "    'record|specimen_identifiers|fieldnum':'fieldNumber',\n",
    "    'record|specimen_identifiers|institution_storing':'ownerInstitutionCode',\n",
    "    'record|specimen_identifiers|sampleid':'materialSampleID',\n",
    "    # use df.apply to merge following fields into associatedMedia \n",
    "    # specimen_imagery|media|caption \n",
    "    # specimen_imagery|media|copyright  \n",
    "    # specimen_imagery|media|image_file         \n",
    "    # specimen_imagery|media|mediaID\n",
    "    # specimen_imagery|media|metatags\n",
    "    'record|taxonomy|identification_provided_by':'identifiedBy',\n",
    "    'record|taxonomy|class|taxon|name':'class',\n",
    "    'record|taxonomy|class|taxon|taxID':'classID',\n",
    "    'record|taxonomy|family|taxon|name':'family',\n",
    "    'record|taxonomy|family|taxon|taxID':'familyID',\n",
    "    'record|taxonomy|genus|taxon|name':'genus',\n",
    "    'record|taxonomy|genus|taxon|taxID':'genusID',\n",
    "    'record|taxonomy|order|taxon|name':'order',\n",
    "    'record|taxonomy|order|taxon|taxID':'orderID',\n",
    "    'record|taxonomy|phylum|taxon|name':'phylum',\n",
    "    'record|taxonomy|phylum|taxon|taxID':'phylumID',\n",
    "    'record|taxonomy|species|taxon|name':'species',\n",
    "    'record|taxonomy|species|taxon|taxID':'speciesID',\n",
    "    # use df.apply to merge following fields into associatedMedia \n",
    "    # record|tracefiles|read|*\n",
    "}\n",
    "\n",
    "# keep copies of DwC matched columns (just in case)\n",
    "df['scientificName'] = df['record|taxonomy|species|taxon|name'] \n",
    "\n",
    "# do rename\n",
    "df.rename(\n",
    "  columns = field_mapping, \n",
    "  inplace = True\n",
    ")\n",
    "\n",
    "df.info()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 701 entries, 0 to 700\n",
      "Data columns (total 41 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   bold_bin_uri           695 non-null    object\n",
      " 1   eventDate              635 non-null    object\n",
      " 2   recordedBy             472 non-null    object\n",
      " 3   coordinatePrecision    2 non-null      object\n",
      " 4   georeferenceSources    6 non-null      object\n",
      " 5   decimalLatitude        642 non-null    object\n",
      " 6   decimalLongitude       642 non-null    object\n",
      " 7   country                701 non-null    object\n",
      " 8   locationRemarks        164 non-null    object\n",
      " 9   stateProvince          696 non-null    object\n",
      " 10  locality               34 non-null     object\n",
      " 11  fieldNotes             19 non-null     object\n",
      " 12  recordNumber           701 non-null    object\n",
      " 13  occurrenceID           701 non-null    object\n",
      " 14  identificationRemarks  57 non-null     object\n",
      " 15  lifeStage              139 non-null    object\n",
      " 16  bold_reproduction      518 non-null    object\n",
      " 17  sex                    146 non-null    object\n",
      " 18  dispositionProperty    0 non-null      object\n",
      " 19  bold_voucher_type      166 non-null    object\n",
      " 20  otherCatalogNumbers    652 non-null    object\n",
      " 21  fieldNumber            323 non-null    object\n",
      " 22  ownerInstitutionCode   701 non-null    object\n",
      " 23  materialSampleID       701 non-null    object\n",
      " 24  class                  701 non-null    object\n",
      " 25  classID                701 non-null    object\n",
      " 26  family                 230 non-null    object\n",
      " 27  familyID               230 non-null    object\n",
      " 28  genus                  229 non-null    object\n",
      " 29  genusID                229 non-null    object\n",
      " 30  identifiedBy           470 non-null    object\n",
      " 31  order                  701 non-null    object\n",
      " 32  orderID                701 non-null    object\n",
      " 33  phylum                 701 non-null    object\n",
      " 34  phylumID               701 non-null    object\n",
      " 35  species                229 non-null    object\n",
      " 36  speciesID              229 non-null    object\n",
      " 37  scientificName         229 non-null    object\n",
      " 38  associatedSequences    701 non-null    object\n",
      " 39  associatedMedia        701 non-null    object\n",
      " 40  bold_tracefiles_read   701 non-null    object\n",
      "dtypes: object(41)\n",
      "memory usage: 224.7+ KB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bold_bin_uri</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>coordinatePrecision</th>\n",
       "      <th>georeferenceSources</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>country</th>\n",
       "      <th>locationRemarks</th>\n",
       "      <th>stateProvince</th>\n",
       "      <th>...</th>\n",
       "      <th>order</th>\n",
       "      <th>orderID</th>\n",
       "      <th>phylum</th>\n",
       "      <th>phylumID</th>\n",
       "      <th>species</th>\n",
       "      <th>speciesID</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>associatedSequences</th>\n",
       "      <th>associatedMedia</th>\n",
       "      <th>bold_tracefiles_read</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOLD:ACF0415</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.6528</td>\n",
       "      <td>150.146</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>Squamata</td>\n",
       "      <td>367</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>18</td>\n",
       "      <td>Pseudechis porphyriacus</td>\n",
       "      <td>515104</td>\n",
       "      <td>Pseudechis porphyriacus</td>\n",
       "      <td>[{'sequenceID': '5161496', 'genbank_accession'...</td>\n",
       "      <td>[{'mediaID': '1902893', 'image_file': 'http://...</td>\n",
       "      <td>[{'direction': 'F', 'read_id': '5001846', 'mar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BOLD:ACF0415</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.6528</td>\n",
       "      <td>150.146</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>Squamata</td>\n",
       "      <td>367</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>18</td>\n",
       "      <td>Pseudechis porphyriacus</td>\n",
       "      <td>515104</td>\n",
       "      <td>Pseudechis porphyriacus</td>\n",
       "      <td>[{'sequenceID': '5161496', 'genbank_accession'...</td>\n",
       "      <td>[{'mediaID': '1902893', 'image_file': 'http://...</td>\n",
       "      <td>[{'direction': 'R', 'read_id': '5001867', 'mar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BOLD:ACE9617</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.4374</td>\n",
       "      <td>149.799</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>Testudines</td>\n",
       "      <td>384</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>18</td>\n",
       "      <td>Chelodina longicollis</td>\n",
       "      <td>171558</td>\n",
       "      <td>Chelodina longicollis</td>\n",
       "      <td>[{'sequenceID': '5161498', 'genbank_accession'...</td>\n",
       "      <td>[{'mediaID': '1902895', 'image_file': 'http://...</td>\n",
       "      <td>[{'direction': 'F', 'read_id': '5001848', 'mar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BOLD:ACE9617</td>\n",
       "      <td>2010-10-17</td>\n",
       "      <td>P.Hebert</td>\n",
       "      <td>None</td>\n",
       "      <td>GPS</td>\n",
       "      <td>-35.4374</td>\n",
       "      <td>149.799</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Kings Hwy 52</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>...</td>\n",
       "      <td>Testudines</td>\n",
       "      <td>384</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>18</td>\n",
       "      <td>Chelodina longicollis</td>\n",
       "      <td>171558</td>\n",
       "      <td>Chelodina longicollis</td>\n",
       "      <td>[{'sequenceID': '5161498', 'genbank_accession'...</td>\n",
       "      <td>[{'mediaID': '1902895', 'image_file': 'http://...</td>\n",
       "      <td>[{'direction': 'R', 'read_id': '5001869', 'mar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BOLD:ADI2959</td>\n",
       "      <td>2002-10-07</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>-15.6469</td>\n",
       "      <td>145.22</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Cooktown, ca. 20 km (by road) S of, Black Moun...</td>\n",
       "      <td>Queensland</td>\n",
       "      <td>...</td>\n",
       "      <td>Squamata</td>\n",
       "      <td>367</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>18</td>\n",
       "      <td>Liburnascincus scirtetis</td>\n",
       "      <td>783178</td>\n",
       "      <td>Liburnascincus scirtetis</td>\n",
       "      <td>[{'sequenceID': '9474096', 'genbank_accession'...</td>\n",
       "      <td>[{'mediaID': nan, 'image_file': nan, 'metatags...</td>\n",
       "      <td>[{'direction': 'f', 'read_id': '10282446', 'ma...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   bold_bin_uri   eventDate recordedBy coordinatePrecision  \\\n",
       "0  BOLD:ACF0415  2010-10-17   P.Hebert                None   \n",
       "1  BOLD:ACF0415  2010-10-17   P.Hebert                None   \n",
       "2  BOLD:ACE9617  2010-10-17   P.Hebert                None   \n",
       "3  BOLD:ACE9617  2010-10-17   P.Hebert                None   \n",
       "4  BOLD:ADI2959  2002-10-07       None                None   \n",
       "\n",
       "  georeferenceSources decimalLatitude decimalLongitude    country  \\\n",
       "0                 GPS        -35.6528          150.146  Australia   \n",
       "1                 GPS        -35.6528          150.146  Australia   \n",
       "2                 GPS        -35.4374          149.799  Australia   \n",
       "3                 GPS        -35.4374          149.799  Australia   \n",
       "4                None        -15.6469           145.22  Australia   \n",
       "\n",
       "                                     locationRemarks    stateProvince  ...  \\\n",
       "0                                       Kings Hwy 52  New South Wales  ...   \n",
       "1                                       Kings Hwy 52  New South Wales  ...   \n",
       "2                                       Kings Hwy 52  New South Wales  ...   \n",
       "3                                       Kings Hwy 52  New South Wales  ...   \n",
       "4  Cooktown, ca. 20 km (by road) S of, Black Moun...       Queensland  ...   \n",
       "\n",
       "        order orderID    phylum phylumID                   species speciesID  \\\n",
       "0    Squamata     367  Chordata       18   Pseudechis porphyriacus    515104   \n",
       "1    Squamata     367  Chordata       18   Pseudechis porphyriacus    515104   \n",
       "2  Testudines     384  Chordata       18     Chelodina longicollis    171558   \n",
       "3  Testudines     384  Chordata       18     Chelodina longicollis    171558   \n",
       "4    Squamata     367  Chordata       18  Liburnascincus scirtetis    783178   \n",
       "\n",
       "             scientificName  \\\n",
       "0   Pseudechis porphyriacus   \n",
       "1   Pseudechis porphyriacus   \n",
       "2     Chelodina longicollis   \n",
       "3     Chelodina longicollis   \n",
       "4  Liburnascincus scirtetis   \n",
       "\n",
       "                                 associatedSequences  \\\n",
       "0  [{'sequenceID': '5161496', 'genbank_accession'...   \n",
       "1  [{'sequenceID': '5161496', 'genbank_accession'...   \n",
       "2  [{'sequenceID': '5161498', 'genbank_accession'...   \n",
       "3  [{'sequenceID': '5161498', 'genbank_accession'...   \n",
       "4  [{'sequenceID': '9474096', 'genbank_accession'...   \n",
       "\n",
       "                                     associatedMedia  \\\n",
       "0  [{'mediaID': '1902893', 'image_file': 'http://...   \n",
       "1  [{'mediaID': '1902893', 'image_file': 'http://...   \n",
       "2  [{'mediaID': '1902895', 'image_file': 'http://...   \n",
       "3  [{'mediaID': '1902895', 'image_file': 'http://...   \n",
       "4  [{'mediaID': nan, 'image_file': nan, 'metatags...   \n",
       "\n",
       "                                bold_tracefiles_read  \n",
       "0  [{'direction': 'F', 'read_id': '5001846', 'mar...  \n",
       "1  [{'direction': 'R', 'read_id': '5001867', 'mar...  \n",
       "2  [{'direction': 'F', 'read_id': '5001848', 'mar...  \n",
       "3  [{'direction': 'R', 'read_id': '5001869', 'mar...  \n",
       "4  [{'direction': 'f', 'read_id': '10282446', 'ma...  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# process sequences data\n",
    "sequenceFields = {\n",
    "  'sequenceID': 'record|sequences|sequence|sequenceID',\n",
    "  'genbank_accession': 'record|sequences|sequence|genbank_accession',\n",
    "  'markercode': 'record|sequences|sequence|markercode',\n",
    "  'nucleotides': 'record|sequences|sequence|nucleotides',\n",
    "}\n",
    "\n",
    "def addAssociatedeSequenceData(row):\n",
    "  sequenceData = {}\n",
    "  for sd in sequenceFields:\n",
    "    sequenceData[sd] = row[sequenceFields[sd]]\n",
    "  return [sequenceData]\n",
    "\n",
    "df['associatedSequences'] = df.apply(addAssociatedeSequenceData, axis=1)\n",
    "df.drop(sequenceFields.values(), axis=1, inplace=True)\n",
    "\n",
    "# process images data\n",
    "imagesFields = {\n",
    "  'mediaID': 'record|specimen_imagery|media|mediaID',\n",
    "  'image_file': 'record|specimen_imagery|media|image_file',\n",
    "  'metatags': 'record|specimen_imagery|media|metatags',\n",
    "  'copyright': 'record|specimen_imagery|media|copyright',\n",
    "  'caption': 'record|specimen_imagery|media|caption',\n",
    "}\n",
    "\n",
    "def associatedImageData(row):\n",
    "  imageData = {}\n",
    "  for id in imagesFields:\n",
    "    imageData[id] = row[imagesFields[id]]\n",
    "  return [imageData]\n",
    "\n",
    "df['associatedMedia'] = df.apply(associatedImageData, axis=1)\n",
    "df.drop(imagesFields.values(), axis=1, inplace=True)\n",
    "\n",
    "# process traces data\n",
    "tracesFields = {\n",
    "  'direction': 'record|tracefiles|read|direction',\n",
    "  'read_id': 'record|tracefiles|read|read_id',\n",
    "  'markercode': 'record|tracefiles|read|markercode',\n",
    "  'run_date': 'record|tracefiles|read|run_date',\n",
    "  'seq_primer': 'record|tracefiles|read|seq_primer',\n",
    "  'sequencing_center': 'record|tracefiles|read|sequencing_center',\n",
    "  'trace_link': 'record|tracefiles|read|trace_link',\n",
    "}\n",
    "\n",
    "def addTracesData(row):\n",
    "  tdata = {}\n",
    "  for tf in tracesFields:\n",
    "    tdata[tf] = row[tracesFields[tf]]\n",
    "  return [tdata]\n",
    "\n",
    "df['bold_tracefiles_read'] = df.apply(addTracesData, axis=1)\n",
    "df.drop(tracesFields.values(), axis=1, inplace=True)\n",
    "  \n",
    "df.drop(['record|taxonomy|subfamily|taxon|name','record|taxonomy|subfamily|taxon|taxID'], axis=1, inplace=True)\n",
    "df.info()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to CSV file\n",
    "df.to_csv(f\"/data/arga-data/bold_australia_data.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
