{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9df0567f",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# NCBI Genome RefSeq ingestion script\n",
    "\n",
    "This script reads the listing for the:\n",
    "\n",
    "- NCBI genome/refseq via the URL https://ftp.ncbi.nih.gov/genomes/refseq/assembly_summary_refseq.txt \n",
    "- NCBI genome/genbank via the URL https://ftp.ncbi.nih.gov/genomes/genbank/assembly_summary_genbank.txt\n",
    "\n",
    "and attempts to map the columns to Darwin Core (DwC), using any of the DwC extenesions that might apply. More info in [this issue](https://github.com/AtlasOfLivingAustralia/arga-issues/issues/8).\n",
    "\n",
    "### TODO\n",
    "- create URL to `refseq` entry page on NCBI site\n",
    "- fix DwC fields not matched by preingestion code due to being deprecated: `EarliestDateCollected` & ~`MaterialSample` (is a class not a field)~\n",
    "- extract `geo_loc_name` into `locality` from biosample `attrs`\n",
    "- add link to refseq page - e.g. https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_013358625.1/ from accession `ncbi_assembly_accession\":\"GCF_013358625.1\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "e4998905",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "genome_directory = \"refseq\" # \"genbank\"\n",
    "# df = pd.read_table(\"https://ftp.ncbi.nih.gov/genomes/genbank/assembly_summary_genbank.txt\", header=1, dtype='object', parse_dates=['seq_rel_date'])\n",
    "df = pd.read_table(f\"https://ftp.ncbi.nih.gov/genomes/{genome_directory}/assembly_summary_{genome_directory}.txt\", header=1, dtype='object', parse_dates=['seq_rel_date'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a95f5cd9",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Look at the data types detected by pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "a29aa8a0",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 263364 entries, 0 to 263363\n",
      "Data columns (total 23 columns):\n",
      " #   Column                     Non-Null Count   Dtype         \n",
      "---  ------                     --------------   -----         \n",
      " 0   # assembly_accession       263364 non-null  object        \n",
      " 1   bioproject                 263226 non-null  object        \n",
      " 2   biosample                  251649 non-null  object        \n",
      " 3   wgs_master                 219908 non-null  object        \n",
      " 4   refseq_category            263364 non-null  object        \n",
      " 5   taxid                      263364 non-null  object        \n",
      " 6   species_taxid              263364 non-null  object        \n",
      " 7   organism_name              263364 non-null  object        \n",
      " 8   infraspecific_name         244326 non-null  object        \n",
      " 9   isolate                    17785 non-null   object        \n",
      " 10  version_status             263364 non-null  object        \n",
      " 11  assembly_level             263364 non-null  object        \n",
      " 12  release_type               263364 non-null  object        \n",
      " 13  genome_rep                 263364 non-null  object        \n",
      " 14  seq_rel_date               263364 non-null  datetime64[ns]\n",
      " 15  asm_name                   263168 non-null  object        \n",
      " 16  submitter                  251468 non-null  object        \n",
      " 17  gbrs_paired_asm            263364 non-null  object        \n",
      " 18  paired_asm_comp            263364 non-null  object        \n",
      " 19  ftp_path                   263364 non-null  object        \n",
      " 20  excluded_from_refseq       507 non-null     object        \n",
      " 21  relation_to_type_material  27236 non-null   object        \n",
      " 22  asm_not_live_date          263364 non-null  object        \n",
      "dtypes: datetime64[ns](1), object(22)\n",
      "memory usage: 46.2+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "53073a18",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>paired_asm_comp</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>excluded_from_refseq</th>\n",
       "      <th>relation_to_type_material</th>\n",
       "      <th>asm_not_live_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>263354</th>\n",
       "      <td>GCF_940677155.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14288414</td>\n",
       "      <td>CALNWH000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>192955</td>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ke...</td>\n",
       "      <td>strain=11_0799</td>\n",
       "      <td>11_0799</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-07</td>\n",
       "      <td>11_0799 assembly</td>\n",
       "      <td>inrae nouzilly</td>\n",
       "      <td>GCA_940677155.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263355</th>\n",
       "      <td>GCF_940677165.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14288404</td>\n",
       "      <td>CALNWI000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>90371</td>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>strain=08_1537</td>\n",
       "      <td>08_1537</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-07</td>\n",
       "      <td>08_1537 assembly</td>\n",
       "      <td>inrae nouzilly</td>\n",
       "      <td>GCA_940677165.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263356</th>\n",
       "      <td>GCF_940677185.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14288403</td>\n",
       "      <td>CALNWK000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>90371</td>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>strain=04_9275</td>\n",
       "      <td>04_9275</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-07</td>\n",
       "      <td>04_9275 assembly</td>\n",
       "      <td>inrae nouzilly</td>\n",
       "      <td>GCA_940677185.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263357</th>\n",
       "      <td>GCF_940677195.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14288402</td>\n",
       "      <td>CALNWG000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>108619</td>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ne...</td>\n",
       "      <td>strain=01_2174</td>\n",
       "      <td>01_2174</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-07</td>\n",
       "      <td>01_2174 assembly</td>\n",
       "      <td>inrae nouzilly</td>\n",
       "      <td>GCA_940677195.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263358</th>\n",
       "      <td>GCF_940677205.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14305874</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>36835</td>\n",
       "      <td>36835</td>\n",
       "      <td>[Clostridium] colinum</td>\n",
       "      <td>strain=DSM 6011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-11</td>\n",
       "      <td>DSM6011</td>\n",
       "      <td>Center for Biotechnology, Bielefeld University</td>\n",
       "      <td>GCA_940677205.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>assembly from type material</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263359</th>\n",
       "      <td>GCF_940792985.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14305861</td>\n",
       "      <td>CALNWU000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2758090</td>\n",
       "      <td>2758090</td>\n",
       "      <td>Dermabacter sp. Marseille-Q3180</td>\n",
       "      <td>strain=Marseille-Q3180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-07</td>\n",
       "      <td>PRJEB52557</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_940792985.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263360</th>\n",
       "      <td>GCF_940796315.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14305863</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>2759692</td>\n",
       "      <td>2759692</td>\n",
       "      <td>Streptococcus sp. Marseille-Q3533</td>\n",
       "      <td>strain=Marseille-Q3533</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-11</td>\n",
       "      <td>PRJEB52562</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_940796315.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263361</th>\n",
       "      <td>GCF_940796375.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14305859</td>\n",
       "      <td>CALNWW000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2759694</td>\n",
       "      <td>2759694</td>\n",
       "      <td>Gardnerella sp. Marseille-Q2328</td>\n",
       "      <td>strain=Marseille-Q2328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-07</td>\n",
       "      <td>PRJEB52556</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_940796375.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263362</th>\n",
       "      <td>GCF_940807095.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14305862</td>\n",
       "      <td>CALNWX000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2758092</td>\n",
       "      <td>2758092</td>\n",
       "      <td>Chryseobacterium sp. Marseille-Q3244</td>\n",
       "      <td>strain=Marseille-Q3244</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-07</td>\n",
       "      <td>PRJEB52561</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_940807095.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263363</th>\n",
       "      <td>GCF_942548115.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA9965907</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>1736306</td>\n",
       "      <td>1736306</td>\n",
       "      <td>Aeromicrobium sp. Leaf245</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-05-22</td>\n",
       "      <td>Leaf245_closed</td>\n",
       "      <td>ETH ZURICH</td>\n",
       "      <td>GCA_942548115.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       # assembly_accession   bioproject      biosample         wgs_master  \\\n",
       "263354      GCF_940677155.1  PRJNA224116  SAMEA14288414  CALNWH000000000.1   \n",
       "263355      GCF_940677165.1  PRJNA224116  SAMEA14288404  CALNWI000000000.1   \n",
       "263356      GCF_940677185.1  PRJNA224116  SAMEA14288403  CALNWK000000000.1   \n",
       "263357      GCF_940677195.1  PRJNA224116  SAMEA14288402  CALNWG000000000.1   \n",
       "263358      GCF_940677205.1  PRJNA224116  SAMEA14305874                NaN   \n",
       "263359      GCF_940792985.1  PRJNA224116  SAMEA14305861  CALNWU000000000.1   \n",
       "263360      GCF_940796315.1  PRJNA224116  SAMEA14305863                NaN   \n",
       "263361      GCF_940796375.1  PRJNA224116  SAMEA14305859  CALNWW000000000.1   \n",
       "263362      GCF_940807095.1  PRJNA224116  SAMEA14305862  CALNWX000000000.1   \n",
       "263363      GCF_942548115.1  PRJNA224116   SAMEA9965907                NaN   \n",
       "\n",
       "       refseq_category    taxid species_taxid  \\\n",
       "263354              na   192955         28901   \n",
       "263355              na    90371         28901   \n",
       "263356              na    90371         28901   \n",
       "263357              na   108619         28901   \n",
       "263358              na    36835         36835   \n",
       "263359              na  2758090       2758090   \n",
       "263360              na  2759692       2759692   \n",
       "263361              na  2759694       2759694   \n",
       "263362              na  2758092       2758092   \n",
       "263363              na  1736306       1736306   \n",
       "\n",
       "                                            organism_name  \\\n",
       "263354  Salmonella enterica subsp. enterica serovar Ke...   \n",
       "263355  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "263356  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "263357  Salmonella enterica subsp. enterica serovar Ne...   \n",
       "263358                              [Clostridium] colinum   \n",
       "263359                    Dermabacter sp. Marseille-Q3180   \n",
       "263360                  Streptococcus sp. Marseille-Q3533   \n",
       "263361                    Gardnerella sp. Marseille-Q2328   \n",
       "263362               Chryseobacterium sp. Marseille-Q3244   \n",
       "263363                          Aeromicrobium sp. Leaf245   \n",
       "\n",
       "            infraspecific_name  isolate  ... genome_rep seq_rel_date  \\\n",
       "263354          strain=11_0799  11_0799  ...       Full   2022-05-07   \n",
       "263355          strain=08_1537  08_1537  ...       Full   2022-05-07   \n",
       "263356          strain=04_9275  04_9275  ...       Full   2022-05-07   \n",
       "263357          strain=01_2174  01_2174  ...       Full   2022-05-07   \n",
       "263358         strain=DSM 6011      NaN  ...       Full   2022-05-11   \n",
       "263359  strain=Marseille-Q3180      NaN  ...       Full   2022-05-07   \n",
       "263360  strain=Marseille-Q3533      NaN  ...       Full   2022-05-11   \n",
       "263361  strain=Marseille-Q2328      NaN  ...       Full   2022-05-07   \n",
       "263362  strain=Marseille-Q3244      NaN  ...       Full   2022-05-07   \n",
       "263363                     NaN       wt  ...       Full   2022-05-22   \n",
       "\n",
       "                asm_name                                       submitter  \\\n",
       "263354  11_0799 assembly                                  inrae nouzilly   \n",
       "263355  08_1537 assembly                                  inrae nouzilly   \n",
       "263356  04_9275 assembly                                  inrae nouzilly   \n",
       "263357  01_2174 assembly                                  inrae nouzilly   \n",
       "263358           DSM6011  Center for Biotechnology, Bielefeld University   \n",
       "263359        PRJEB52557                    IHU - Mediterranee Infection   \n",
       "263360        PRJEB52562                    IHU - Mediterranee Infection   \n",
       "263361        PRJEB52556                    IHU - Mediterranee Infection   \n",
       "263362        PRJEB52561                    IHU - Mediterranee Infection   \n",
       "263363    Leaf245_closed                                      ETH ZURICH   \n",
       "\n",
       "        gbrs_paired_asm paired_asm_comp  \\\n",
       "263354  GCA_940677155.1       identical   \n",
       "263355  GCA_940677165.1       identical   \n",
       "263356  GCA_940677185.1       identical   \n",
       "263357  GCA_940677195.1       identical   \n",
       "263358  GCA_940677205.1       identical   \n",
       "263359  GCA_940792985.1       identical   \n",
       "263360  GCA_940796315.1       identical   \n",
       "263361  GCA_940796375.1       identical   \n",
       "263362  GCA_940807095.1       identical   \n",
       "263363  GCA_942548115.1       identical   \n",
       "\n",
       "                                                 ftp_path  \\\n",
       "263354  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263355  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263356  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263357  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263358  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263359  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263360  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263361  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263362  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "263363  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "\n",
       "       excluded_from_refseq    relation_to_type_material asm_not_live_date  \n",
       "263354                  NaN                          NaN                na  \n",
       "263355                  NaN                          NaN                na  \n",
       "263356                  NaN                          NaN                na  \n",
       "263357                  NaN                          NaN                na  \n",
       "263358                  NaN  assembly from type material                na  \n",
       "263359                  NaN                          NaN                na  \n",
       "263360                  NaN                          NaN                na  \n",
       "263361                  NaN                          NaN                na  \n",
       "263362                  NaN                          NaN                na  \n",
       "263363                  NaN                          NaN                na  \n",
       "\n",
       "[10 rows x 23 columns]"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b8c9c37",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Field mappings\n",
    "\n",
    "NCBI fields described at https://ftp.ncbi.nih.gov/genomes/README_assembly_summary.txt\n",
    "\n",
    "| Column | field name | field description |\n",
    "|--------| ---- | ---- |\n",
    "| 1      | \"assembly_accession\" | Assembly accession: the assembly accession.version reported in this field is a unique identifier for the set of sequences in this particular version of the genome assembly. |\n",
    "| 2      | \"bioproject\" | BioProject: accession for the BioProject which produced the sequences in the genome assembly. A BioProject is a collection of biological data related to a single initiative, originating from a single organization or from a consortium. A BioProject record provides users a single place to find links to the diverse data types generated for that project. The record can be retrieved from the NCBI BioProject resource: https://www.ncbi.nlm.nih.gov/bioproject/ |\n",
    "| 3      | \"biosample\" | BioSample: accession for the BioSample from which the sequences in the genome assembly were obtained. A BioSample record contains a description of the biological source material used in experimental assays. The record can be retrieved from the NCBI BioSample resource: https://www.ncbi.nlm.nih.gov/biosample/ |\n",
    "| 4      | \"wgs_master\" | WGS-master: the GenBank Nucleotide accession and version for the master record of the Whole Genome Shotgun (WGS) project for the genome assembly. The master record can be retrieved from the NCBI Nucleotide resource: https://www.ncbi.nlm.nih.gov/nuccore Genome assemblies that are complete genomes, and those that are clone-based, do not have WGS-master records in which case this field will be empty. |\n",
    "| 5      | \"refseq_category\" | RefSeq Category: whether the assembly is a reference or representative genome in the NCBI Reference Sequence (RefSeq) project classification. Values: **reference genome** - a manually selected high quality genome assembly that NCBI and the community have identified as being important as a standard against which other data are compared; **representative genome** - a genome computationally or manually selected as a representative from among the best genomes available for a species or clade that does not have a designated reference genome; **na** - no RefSeq category assigned to this assembly Prokaryotes may have more than one reference or representative genome per species. For more information see: https://www.ncbi.nlm.nih.gov/refseq/about/prokaryotes/#referencegenome Eukaryotes have no more than one reference or representative genome per species. If there are no assemblies in RefSeq for a particular eukaryotic species, then the GenBank assembly that RefSeq would select as the best available for that species will be designated as the representative genome. Viruses may have one or more reference genomes per species. The representative genome designation is not applied to viruses. |\n",
    "| 6      | \"taxid\" | Taxonomy ID: the NCBI taxonomy identifier for the organism from which the genome assembly was derived. The NCBI Taxonomy Database is a curated classification and nomenclature for all of the organisms in the public sequence databases. The taxonomy record can be retrieved from the NCBI Taxonomy resource: https://www.ncbi.nlm.nih.gov/taxonomy/ |\n",
    "| 7      | \"species_taxid\" | Species taxonomy ID: the NCBI taxonomy identifier for the species from which the genome assembly was derived. The species taxid will differ from the organism taxid (column 6) only when the organism was reported at a sub- species or strain level. |\n",
    "| 8      | \"organism_name\" | Organism name: the scientific name of the organism from which the sequences in the genome assembly were derived. This name is taken from the NCBI Taxonomy record for the taxid specified in column 6. Some older taxids were assigned at the strain level and for these the organism name will include the strain. Current practice is only to assign taxids at the species level; for these the organism name will be just the species, however, the strain name will be reported in the infraspecific_name field (column 9). |\n",
    "| 9      | \"infraspecific_name\" | Infraspecific name: the strain, breed, cultivar or ecotype of the organism from which the sequences in the genome assembly were derived. Data are reported in the form tag=value, e.g. strain=AF16. Strain, breed, cultivar and ecotype are not expected to be used together, however, if they are then they will be reported in a list separated by \", /\". Empty if no strain, breed, cultivar or ecotype is specified on the genomic sequence records. |\n",
    "| 10     | \"isolate\" | Isolate: the individual isolate from which the sequences in the genome assembly were derived. Empty if no isolate is specified on the genomic sequence records. |\n",
    "| 11     | \"version_status\" | Version status: the release status for the genome assembly version. Values: **latest** - the most recent of all the versions for this assembly chain; **replaced** - this version has been replaced by a newer version of the assembly in the same chain; **suppressed** - this version of the assembly has been suppressed An assembly chain is the collection of all versions for the same assembly accession. |\n",
    "| 12     | \"assembly_level\" | Assembly level: the highest level of assembly for any object in the genome assembly. Values: **Complete genome** - all chromosomes are gapless and have no runs of 10 or more ambiguous bases (Ns), there are no unplaced or unlocalized scaffolds, and all the expected chromosomes are present (i.e. the assembly is not noted as having partial genome representation). Plasmids and organelles may or may not be included in the assembly but if present then the sequences are gapless. **Chromosome** - there is sequence for one or more chromosomes. This could be a completely sequenced chromosome without gaps or a chromosome containing scaffolds or contigs with gaps between them. There may also be unplaced or unlocalized scaffolds. Scaffold - some sequence contigs have been connected across gaps to create scaffolds, but the scaffolds are all unplaced or unlocalized. **Contig** - nothing is assembled beyond the level of sequence contigs |\n",
    "| 13     | \"release_type\" | Release type: whether this version of the genome assembly is a major, minor or patch release. Values: **Major** - changes from the previous assembly version result in a significant change to the coordinate system. The first version of an assembly is always a major release. Most subsequent genome assembly updates are also major releases. **Minor** - changes from the previous assembly version are limited to the following changes, none of which result in a significant change to the coordinate system of the primary assembly-unit: - adding, removing or changing a non-nuclear assembly-unit - dropping unplaced or unlocalized scaffolds - adding up to 50 unplaced or unlocalized scaffolds which are shorter than the current scaffold-N50 value - replacing a component with a gap of the same length Patch - the only change is the addition or modification of a patch assembly-unit. See the NCBI Assembly model web page (https://www.ncbi.nlm.nih.gov/assembly/ model/#asmb_def) for definitions of assembly-units and genome patches. |\n",
    "| 14     | \"genome_rep\" | Genome representation: whether the goal for the assembly was to represent the whole genome or only part of it. Values: **Full** - the data used to generate the assembly was obtained from the whole genome, as in Whole Genome Shotgun (WGS) assemblies for example. There may still be gaps in the assembly. **Partial** - the data used to generate the assembly came from only part of the genome. Most assemblies have full genome representation with a minority being partial genome representation. See the Assembly help web page (https://www.ncbi.nlm.nih.gov/assembly/help/) for reasons that the genome representation would be set to partial. |\n",
    "| 15     | \"seq_rel_date\" | Sequence release date: the date the sequences in the genome assembly were released in the International Nucleotide Sequence Database Collaboration (INSDC) databases, i.e. DDBJ, ENA or GenBank. |\n",
    "| 16     | \"asm_name\" | Assembly name: the submitter's name for the genome assembly, when one was provided, otherwise a default name, in the form ASM#####v#, is provided by NCBI. Assembly names are not unique. |\n",
    "| 17     | \"submitter\" | Submitter: the submitting consortium or first position if a list of organizations. The full submitter information is available in the NCBI BioProject resource: www.ncbi.nlm.nih.gov/bioproject/ |\n",
    "| 18     | \"gbrs_paired_asm\" | GenBank/RefSeq paired assembly: the accession.version of the GenBank assembly that is paired to the given RefSeq assembly, or vice-versa. \"na\" is reported if the assembly is unpaired. |\n",
    "| 19     | \"paired_asm_comp\" | Paired assembly comparison: whether the paired GenBank & RefSeq assemblies are identical or different. Values: **identical** - GenBank and RefSeq assemblies are identical; **different** - GenBank and RefSeq assemblies are not identical; **na** - not applicable since the assembly is unpaired |\n",
    "| 20     | \"ftp_path\" | FTP path: the path to the directory on the NCBI genomes FTP site from which data for this genome assembly can be downloaded. |\n",
    "| 21     | \"excluded_from_refseq\" | Excluded from RefSeq: reasons the assembly was excluded from the NCBI Reference Sequence (RefSeq) project, including any assembly anomalies. See: https://www.ncbi.nlm.nih.gov/assembly/help/anomnotrefseq/ |\n",
    "| 22     | \"relation_to_type_material\" | Relation to type material: contains a value if the sequences in the genome assembly were derived from type material. Values: **assembly from type material** - the sequences in the genome assembly were derived from type material; **assembly from synonym type material** - the sequences in the genome assembly were derived from synonym type material; **assembly from pathotype material** - the sequences in the genome assembly were derived from pathovar material;** **assembly designated as neotype** - the sequences in the genome assembly were derived from neotype material; **assembly designated as reftype** - the sequences in the genome assembly were derived from reference material where type material never was available and is not likely to ever be available ICTV species exemplar - the International Committee on Taxonomy of Viruses (ICTV) designated the genome assembly as the exemplar for the virus species ICTV additional isolate - the International Committee on Taxonomy of Viruses (ICTV) designated the genome assembly an additional isolate for the virus species |\n",
    "| 23     | \"asm_not_live_date\" | Assembly no longer live date: the date the assembly transitioned from version_status latest to either replaced or suppressed. When the assembly is in status latest, \"na\" is reported. |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9f9f151",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Attempt to map NCBI fields to known [DwC terms](https://dwc.tdwg.org/list/), including extensions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "348ef5b0",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# 1:1 mapping to best-guess DwC and extension terms\n",
    "new_names = {\n",
    "  '# assembly_accession': 'occurrenceID',\n",
    "  'bioproject': 'ncbi_bioproject',\n",
    "  'biosample': 'ncbi_biosample',\n",
    "  'wgs_master': 'ncbi_nuccore',\n",
    "  'refseq_category': 'ncbi_refseq_category',\n",
    "  'taxid': 'taxonID',\n",
    "  'species_taxid': 'specific_host',\n",
    "  'organism_name': 'scientificName',\n",
    "  'infraspecific_name': 'cloneStrain',\n",
    "  'isolate': 'source_mat_id',\n",
    "  'version_status': 'ncbi_version_status',\n",
    "  'assembly_level': 'ncbi_assembly_level',\n",
    "  'release_type': 'ncbi_release_type',\n",
    "  'genome_rep': 'ncbi_genome_rep',\n",
    "  'seq_rel_date': 'eventDate',\n",
    "  'asm_name': 'otherCatalogNumbers',\n",
    "  'submitter': 'recordedBy',\n",
    "  'gbrs_paired_asm': 'ncbi_gbrs_paired_asm',\n",
    "  'paired_asm_comp': 'ncbi_paired_asm_comp',\n",
    "  'ftp_path': 'geneticAccessionURI',\n",
    "  'excluded_from_refseq': 'ncbi_excluded_from_refseq',\n",
    "  'relation_to_type_material': 'ncbi_relation_to_type_material',\n",
    "  'asm_not_live_date': 'ncbia_sm_not_live_date'\n",
    "}\n",
    "\n",
    "# keep copies of DwC matched columns (just in case)\n",
    "df['ncbi_assembly_accession'] = df['# assembly_accession'] \n",
    "df['ncbi_taxid'] = df['taxid'] \n",
    "df['ncbi_species_taxid'] = df['species_taxid'] \n",
    "df['ncbi_organism_name'] = df['organism_name']\n",
    "df['ncbi_infraspecific_name'] = df['infraspecific_name']\n",
    "df['ncbi_isolate'] = df['isolate']\n",
    "df['ncbi_seq_rel_date'] = df['seq_rel_date'] \n",
    "df['ncbi_asm_name'] = df['asm_name']\n",
    "df['ncbi_submitter'] = df['submitter']\n",
    "df['ncbi_ftp_path'] = df['ftp_path']\n",
    "\n",
    "# create an `occurrenceID` column\n",
    "\n",
    "# do rename\n",
    "df.rename(\n",
    "  columns = new_names, \n",
    "  inplace = True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff256251",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Merge data from NCBI BioSample database\n",
    "\n",
    "NCBI provides an XML version of its [BioSample database](https://www.ncbi.nlm.nih.gov/books/NBK169436/) at `https://ftp.ncbi.nih.gov/biosample/biosample_set.xml.gz`. By mapping the `biosample` accession number (column 3) to the data extracted from BioSample XML file, we can extract a few more columns for the DwC file. The `parse_biosample_*.py` scripts have done this and saved the data into a SQLite3 database file (in data directory)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "9abff88e",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 263364 entries, 0 to 263363\n",
      "Data columns (total 33 columns):\n",
      " #   Column                          Non-Null Count   Dtype         \n",
      "---  ------                          --------------   -----         \n",
      " 0   occurrenceID                    263364 non-null  object        \n",
      " 1   ncbi_bioproject                 263226 non-null  object        \n",
      " 2   ncbi_biosample                  251649 non-null  object        \n",
      " 3   ncbi_nuccore                    219908 non-null  object        \n",
      " 4   ncbi_refseq_category            263364 non-null  object        \n",
      " 5   taxonID                         263364 non-null  object        \n",
      " 6   specific_host                   263364 non-null  object        \n",
      " 7   scientificName                  263364 non-null  object        \n",
      " 8   cloneStrain                     244326 non-null  object        \n",
      " 9   source_mat_id                   17785 non-null   object        \n",
      " 10  ncbi_version_status             263364 non-null  object        \n",
      " 11  ncbi_assembly_level             263364 non-null  object        \n",
      " 12  ncbi_release_type               263364 non-null  object        \n",
      " 13  ncbi_genome_rep                 263364 non-null  object        \n",
      " 14  eventDate                       263364 non-null  datetime64[ns]\n",
      " 15  otherCatalogNumbers             263168 non-null  object        \n",
      " 16  recordedBy                      251468 non-null  object        \n",
      " 17  ncbi_gbrs_paired_asm            263364 non-null  object        \n",
      " 18  ncbi_paired_asm_comp            263364 non-null  object        \n",
      " 19  geneticAccessionURI             263364 non-null  object        \n",
      " 20  ncbi_excluded_from_refseq       507 non-null     object        \n",
      " 21  ncbi_relation_to_type_material  27236 non-null   object        \n",
      " 22  ncbia_sm_not_live_date          263364 non-null  object        \n",
      " 23  ncbi_assembly_accession         263364 non-null  object        \n",
      " 24  ncbi_taxid                      263364 non-null  object        \n",
      " 25  ncbi_species_taxid              263364 non-null  object        \n",
      " 26  ncbi_organism_name              263364 non-null  object        \n",
      " 27  ncbi_infraspecific_name         244326 non-null  object        \n",
      " 28  ncbi_isolate                    17785 non-null   object        \n",
      " 29  ncbi_seq_rel_date               263364 non-null  datetime64[ns]\n",
      " 30  ncbi_asm_name                   263168 non-null  object        \n",
      " 31  ncbi_submitter                  251468 non-null  object        \n",
      " 32  ncbi_ftp_path                   263364 non-null  object        \n",
      "dtypes: datetime64[ns](2), object(31)\n",
      "memory usage: 66.3+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "bf2db2c6",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>ncbi_bioproject</th>\n",
       "      <th>ncbi_biosample</th>\n",
       "      <th>ncbi_nuccore</th>\n",
       "      <th>ncbi_refseq_category</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>specific_host</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>cloneStrain</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>...</th>\n",
       "      <th>ncbi_assembly_accession</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>ncbi_species_taxid</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "      <th>ncbi_infraspecific_name</th>\n",
       "      <th>ncbi_isolate</th>\n",
       "      <th>ncbi_seq_rel_date</th>\n",
       "      <th>ncbi_asm_name</th>\n",
       "      <th>ncbi_submitter</th>\n",
       "      <th>ncbi_ftp_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>PRJNA164</td>\n",
       "      <td>SAMN02803731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-08-01</td>\n",
       "      <td>Release 6 plus ISO1 MT</td>\n",
       "      <td>The FlyBase Consortium/Berkeley Drosophila Gen...</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>PRJNA168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-03</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>PRJNA169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-06-24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>PRJNA116</td>\n",
       "      <td>SAMN03081427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-03-15</td>\n",
       "      <td>TAIR10.1</td>\n",
       "      <td>The Arabidopsis Information Resource (TAIR)</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>PRJNA70973</td>\n",
       "      <td>SAMN02953622</td>\n",
       "      <td>AAGU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>2009-07-15</td>\n",
       "      <td>Loxafr3.0</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       occurrenceID ncbi_bioproject ncbi_biosample    ncbi_nuccore  \\\n",
       "0   GCF_000001215.4        PRJNA164   SAMN02803731             NaN   \n",
       "1  GCF_000001405.40        PRJNA168            NaN             NaN   \n",
       "2  GCF_000001635.27        PRJNA169            NaN             NaN   \n",
       "3   GCF_000001735.4        PRJNA116   SAMN03081427             NaN   \n",
       "4   GCF_000001905.1      PRJNA70973   SAMN02953622  AAGU00000000.3   \n",
       "\n",
       "    ncbi_refseq_category taxonID specific_host           scientificName  \\\n",
       "0       reference genome    7227          7227  Drosophila melanogaster   \n",
       "1       reference genome    9606          9606             Homo sapiens   \n",
       "2       reference genome   10090         10090             Mus musculus   \n",
       "3       reference genome    3702          3702     Arabidopsis thaliana   \n",
       "4  representative genome    9785          9785       Loxodonta africana   \n",
       "\n",
       "        cloneStrain source_mat_id  ... ncbi_assembly_accession ncbi_taxid  \\\n",
       "0               NaN           NaN  ...         GCF_000001215.4       7227   \n",
       "1               NaN           NaN  ...        GCF_000001405.40       9606   \n",
       "2               NaN           NaN  ...        GCF_000001635.27      10090   \n",
       "3  ecotype=Columbia           NaN  ...         GCF_000001735.4       3702   \n",
       "4               NaN    ISIS603380  ...         GCF_000001905.1       9785   \n",
       "\n",
       "  ncbi_species_taxid       ncbi_organism_name ncbi_infraspecific_name  \\\n",
       "0               7227  Drosophila melanogaster                     NaN   \n",
       "1               9606             Homo sapiens                     NaN   \n",
       "2              10090             Mus musculus                     NaN   \n",
       "3               3702     Arabidopsis thaliana        ecotype=Columbia   \n",
       "4               9785       Loxodonta africana                     NaN   \n",
       "\n",
       "  ncbi_isolate ncbi_seq_rel_date           ncbi_asm_name  \\\n",
       "0          NaN        2014-08-01  Release 6 plus ISO1 MT   \n",
       "1          NaN        2022-02-03              GRCh38.p14   \n",
       "2          NaN        2020-06-24                  GRCm39   \n",
       "3          NaN        2018-03-15                TAIR10.1   \n",
       "4   ISIS603380        2009-07-15               Loxafr3.0   \n",
       "\n",
       "                                      ncbi_submitter  \\\n",
       "0  The FlyBase Consortium/Berkeley Drosophila Gen...   \n",
       "1                        Genome Reference Consortium   \n",
       "2                        Genome Reference Consortium   \n",
       "3        The Arabidopsis Information Resource (TAIR)   \n",
       "4                                    Broad Institute   \n",
       "\n",
       "                                       ncbi_ftp_path  \n",
       "0  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "2  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "3  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "4  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df.dropna(how='all')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "f7da5e29",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import sqlite3 as sl\n",
    "import numpy as np\n",
    "\n",
    "con = sl.connect('../data/biosamples_DBL.db')\n",
    "\n",
    "# Function to add extra columns from biosample XML file\n",
    "def addBioSampleData(row):\n",
    "    biosample_accession = row['ncbi_biosample'] if pd.notnull(row['ncbi_biosample']) else ''\n",
    "    #print(\"biosample_accession =\", biosample_accession)\n",
    "    data = ()\n",
    "    if biosample_accession:\n",
    "        try:\n",
    "            cursor = con.cursor()\n",
    "            query = \"\"\"\n",
    "            SELECT dwc_EarliestDateCollected,\n",
    "                     dwc_MaterialSample,\n",
    "                     dwc_lifeStage,\n",
    "                     dwc_materialSampleID,\n",
    "                     dwc_otherCatalogNumbers,\n",
    "                     dwc_sex,\n",
    "                     email,\n",
    "                     lab,\n",
    "                     scientificName,\n",
    "                     attrs\n",
    "            FROM entry\n",
    "            WHERE id = ?\n",
    "            \"\"\"\n",
    "            cursor.execute(query, (biosample_accession,))\n",
    "            data = cursor.fetchone() # tuple is returned\n",
    "        except sl.Error as error:\n",
    "            print(\"Failed to read data from sqlite table\", error)\n",
    "\n",
    "    if data is not None and len(data) == 10:\n",
    "        # convert empty strings to NaN (so consistent with existing data = compresses better too)\n",
    "        conv = lambda i : i or np.nan\n",
    "        list2 = [conv(i) for i in data]\n",
    "        return tuple(list2)\n",
    "    else:\n",
    "        return  tuple([np.nan] * 10) # return just NaN values as tuple\n",
    "\n",
    "# df2 = df.head(9999).copy(deep=True) # subset data for debugging\n",
    "\n",
    "df[['EarliestDateCollected','MaterialSample','lifeStage','materialSampleID','catalogNumber','sex',\n",
    "    'ncbi_biosample_email','ncbi_biosample_lab','ncbi_biosample_scientificName',\n",
    "    'ncbi_biosample_attributes_json']] = df.apply(addBioSampleData, axis=1, result_type='expand')\n",
    "\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "b56ad59b",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 263364 entries, 0 to 263363\n",
      "Data columns (total 43 columns):\n",
      " #   Column                          Non-Null Count   Dtype         \n",
      "---  ------                          --------------   -----         \n",
      " 0   occurrenceID                    263364 non-null  object        \n",
      " 1   ncbi_bioproject                 263226 non-null  object        \n",
      " 2   ncbi_biosample                  251649 non-null  object        \n",
      " 3   ncbi_nuccore                    219908 non-null  object        \n",
      " 4   ncbi_refseq_category            263364 non-null  object        \n",
      " 5   taxonID                         263364 non-null  object        \n",
      " 6   specific_host                   263364 non-null  object        \n",
      " 7   scientificName                  263364 non-null  object        \n",
      " 8   cloneStrain                     244326 non-null  object        \n",
      " 9   source_mat_id                   17785 non-null   object        \n",
      " 10  ncbi_version_status             263364 non-null  object        \n",
      " 11  ncbi_assembly_level             263364 non-null  object        \n",
      " 12  ncbi_release_type               263364 non-null  object        \n",
      " 13  ncbi_genome_rep                 263364 non-null  object        \n",
      " 14  eventDate                       263364 non-null  datetime64[ns]\n",
      " 15  otherCatalogNumbers             263168 non-null  object        \n",
      " 16  recordedBy                      251468 non-null  object        \n",
      " 17  ncbi_gbrs_paired_asm            263364 non-null  object        \n",
      " 18  ncbi_paired_asm_comp            263364 non-null  object        \n",
      " 19  geneticAccessionURI             263364 non-null  object        \n",
      " 20  ncbi_excluded_from_refseq       507 non-null     object        \n",
      " 21  ncbi_relation_to_type_material  27236 non-null   object        \n",
      " 22  ncbia_sm_not_live_date          263364 non-null  object        \n",
      " 23  ncbi_assembly_accession         263364 non-null  object        \n",
      " 24  ncbi_taxid                      263364 non-null  object        \n",
      " 25  ncbi_species_taxid              263364 non-null  object        \n",
      " 26  ncbi_organism_name              263364 non-null  object        \n",
      " 27  ncbi_infraspecific_name         244326 non-null  object        \n",
      " 28  ncbi_isolate                    17785 non-null   object        \n",
      " 29  ncbi_seq_rel_date               263364 non-null  datetime64[ns]\n",
      " 30  ncbi_asm_name                   263168 non-null  object        \n",
      " 31  ncbi_submitter                  251468 non-null  object        \n",
      " 32  ncbi_ftp_path                   263364 non-null  object        \n",
      " 33  EarliestDateCollected           185561 non-null  object        \n",
      " 34  MaterialSample                  318 non-null     object        \n",
      " 35  lifeStage                       0 non-null       float64       \n",
      " 36  materialSampleID                17877 non-null   object        \n",
      " 37  catalogNumber                   2854 non-null    object        \n",
      " 38  sex                             583 non-null     object        \n",
      " 39  ncbi_biosample_email            169891 non-null  object        \n",
      " 40  ncbi_biosample_lab              2229 non-null    object        \n",
      " 41  ncbi_biosample_scientificName   251353 non-null  object        \n",
      " 42  ncbi_biosample_attributes_json  251188 non-null  object        \n",
      "dtypes: datetime64[ns](2), float64(1), object(40)\n",
      "memory usage: 86.4+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "3ca73f5d",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>ncbi_bioproject</th>\n",
       "      <th>ncbi_biosample</th>\n",
       "      <th>ncbi_nuccore</th>\n",
       "      <th>ncbi_refseq_category</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>specific_host</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>cloneStrain</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>...</th>\n",
       "      <th>EarliestDateCollected</th>\n",
       "      <th>MaterialSample</th>\n",
       "      <th>lifeStage</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>catalogNumber</th>\n",
       "      <th>sex</th>\n",
       "      <th>ncbi_biosample_email</th>\n",
       "      <th>ncbi_biosample_lab</th>\n",
       "      <th>ncbi_biosample_scientificName</th>\n",
       "      <th>ncbi_biosample_attributes_json</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>PRJNA164</td>\n",
       "      <td>SAMN02803731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mixed</td>\n",
       "      <td>celniker@fruitfly.org</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>{\"strain\": \"iso-1\", \"age\": \"missing\", \"biomate...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>PRJNA168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>PRJNA169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>PRJNA116</td>\n",
       "      <td>SAMN03081427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>{\"ecotype\": \"Columbia\"}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>PRJNA70973</td>\n",
       "      <td>SAMN02953622</td>\n",
       "      <td>AAGU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>{\"sex\": \"female\", \"geo_loc_name\": \"USA: San Di...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCF_000001985.1</td>\n",
       "      <td>PRJNA32665</td>\n",
       "      <td>SAMN02953685</td>\n",
       "      <td>ABAR00000000.1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>441960</td>\n",
       "      <td>37727</td>\n",
       "      <td>Talaromyces marneffei ATCC 18224</td>\n",
       "      <td>strain=ATCC 18224</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Talaromyces marneffei ATCC 18224</td>\n",
       "      <td>{\"strain\": \"ATCC 18224\", \"culture_collection\":...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GCF_000002035.6</td>\n",
       "      <td>PRJNA13922</td>\n",
       "      <td>SAMN06930106</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7955</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>{\"strain\": \"Tuebingen\", \"note\": \"combined samp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GCF_000002075.1</td>\n",
       "      <td>PRJNA209509</td>\n",
       "      <td>SAMN02953658</td>\n",
       "      <td>AASC00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>6500</td>\n",
       "      <td>6500</td>\n",
       "      <td>Aplysia californica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F4 #8</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F4 #8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hermaphrodite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aplysia californica</td>\n",
       "      <td>{\"sex\": \"hermaphrodite\", \"isolate\": \"F4 #8\"}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GCF_000002235.5</td>\n",
       "      <td>PRJNA13728</td>\n",
       "      <td>SAMN00829422</td>\n",
       "      <td>AAGJ00000000.6</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>7668</td>\n",
       "      <td>7668</td>\n",
       "      <td>Strongylocentrotus purpuratus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spur 01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spur 01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Strongylocentrotus purpuratus</td>\n",
       "      <td>{\"isolate\": \"Spur 01\", \"sex\": \"male\", \"cell_ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GCF_000002285.5</td>\n",
       "      <td>PRJNA12384</td>\n",
       "      <td>SAMN02953603</td>\n",
       "      <td>AAEX00000000.4</td>\n",
       "      <td>na</td>\n",
       "      <td>9615</td>\n",
       "      <td>9612</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>breed=boxer</td>\n",
       "      <td>Tasha</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tasha</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>{\"sex\": \"female\", \"sub_species\": \"familiaris\",...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GCF_000002295.2</td>\n",
       "      <td>PRJNA15683</td>\n",
       "      <td>SAMN02953611</td>\n",
       "      <td>AAFR00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>13616</td>\n",
       "      <td>13616</td>\n",
       "      <td>Monodelphis domestica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Monodelphis domestica</td>\n",
       "      <td>{\"pop-variant\": \"pop1\"}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GCF_000002335.3</td>\n",
       "      <td>PRJNA15718</td>\n",
       "      <td>SAMN02953634</td>\n",
       "      <td>AAJJ00000000.2</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>7070</td>\n",
       "      <td>7070</td>\n",
       "      <td>Tribolium castaneum</td>\n",
       "      <td>strain=Georgia GA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tribolium castaneum</td>\n",
       "      <td>{\"strain\": \"Georgia GA2\", \"collection_date\": \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GCF_000002415.2</td>\n",
       "      <td>PRJNA20431</td>\n",
       "      <td>SAMN02953638</td>\n",
       "      <td>AAKM00000000.1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>5855</td>\n",
       "      <td>5855</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Salvador I</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Salvador I</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>{\"isolate\": \"Salvador I\"}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GCF_000002425.4</td>\n",
       "      <td>PRJNA28131</td>\n",
       "      <td>SAMN00120126</td>\n",
       "      <td>ABEU00000000.2</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>3218</td>\n",
       "      <td>3218</td>\n",
       "      <td>Physcomitrium patens</td>\n",
       "      <td>ecotype=Gransden 2004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Physcomitrium patens</td>\n",
       "      <td>{\"Sample Number\": \"4088616\", \"ecotype\": \"Grans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GCF_000002435.2</td>\n",
       "      <td>PRJNA15590</td>\n",
       "      <td>SAMN02952905</td>\n",
       "      <td>AACB00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>5741</td>\n",
       "      <td>5741</td>\n",
       "      <td>Giardia intestinalis</td>\n",
       "      <td>strain=WB C6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Giardia intestinalis</td>\n",
       "      <td>{\"strain\": \"WB C6\", \"culture_collection\": \"ATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GCF_000002445.2</td>\n",
       "      <td>PRJNA15565</td>\n",
       "      <td>SAMN02953625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>185431</td>\n",
       "      <td>5691</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>strain=927/4 GUTat10.1</td>\n",
       "      <td>927/4 GUTat10.1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>{\"strain\": \"927/4 GUTat10.1\", \"sub_species\": \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GCF_000002455.1</td>\n",
       "      <td>PRJNA27939</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>227086</td>\n",
       "      <td>227086</td>\n",
       "      <td>Bigelowiella natans</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GCF_000002495.2</td>\n",
       "      <td>PRJNA1433</td>\n",
       "      <td>SAMN02953596</td>\n",
       "      <td>AACU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>242507</td>\n",
       "      <td>318829</td>\n",
       "      <td>Pyricularia oryzae 70-15</td>\n",
       "      <td>strain=70-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pyricularia oryzae 70-15</td>\n",
       "      <td>{\"strain\": \"70-15\"}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GCF_000002515.2</td>\n",
       "      <td>PRJNA12377</td>\n",
       "      <td>SAMEA3138170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>28985</td>\n",
       "      <td>28985</td>\n",
       "      <td>Kluyveromyces lactis</td>\n",
       "      <td>strain=NRRL Y-1140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kluyveromyces lactis</td>\n",
       "      <td>{\"Alias\": \"GCA_000002515\", \"Broker name\": \"Eur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GCF_000002525.2</td>\n",
       "      <td>PRJNA12414</td>\n",
       "      <td>SAMEA3138171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>284591</td>\n",
       "      <td>4952</td>\n",
       "      <td>Yarrowia lipolytica CLIB122</td>\n",
       "      <td>strain=CLIB122</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yarrowia lipolytica CLIB122</td>\n",
       "      <td>{\"Alias\": \"GCA_000002525\", \"Broker name\": \"Eur...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        occurrenceID ncbi_bioproject ncbi_biosample    ncbi_nuccore  \\\n",
       "0    GCF_000001215.4        PRJNA164   SAMN02803731             NaN   \n",
       "1   GCF_000001405.40        PRJNA168            NaN             NaN   \n",
       "2   GCF_000001635.27        PRJNA169            NaN             NaN   \n",
       "3    GCF_000001735.4        PRJNA116   SAMN03081427             NaN   \n",
       "4    GCF_000001905.1      PRJNA70973   SAMN02953622  AAGU00000000.3   \n",
       "5    GCF_000001985.1      PRJNA32665   SAMN02953685  ABAR00000000.1   \n",
       "6    GCF_000002035.6      PRJNA13922   SAMN06930106             NaN   \n",
       "7    GCF_000002075.1     PRJNA209509   SAMN02953658  AASC00000000.3   \n",
       "8    GCF_000002235.5      PRJNA13728   SAMN00829422  AAGJ00000000.6   \n",
       "9    GCF_000002285.5      PRJNA12384   SAMN02953603  AAEX00000000.4   \n",
       "10   GCF_000002295.2      PRJNA15683   SAMN02953611  AAFR00000000.3   \n",
       "11   GCF_000002335.3      PRJNA15718   SAMN02953634  AAJJ00000000.2   \n",
       "12   GCF_000002415.2      PRJNA20431   SAMN02953638  AAKM00000000.1   \n",
       "13   GCF_000002425.4      PRJNA28131   SAMN00120126  ABEU00000000.2   \n",
       "14   GCF_000002435.2      PRJNA15590   SAMN02952905  AACB00000000.3   \n",
       "15   GCF_000002445.2      PRJNA15565   SAMN02953625             NaN   \n",
       "16   GCF_000002455.1      PRJNA27939            NaN             NaN   \n",
       "17   GCF_000002495.2       PRJNA1433   SAMN02953596  AACU00000000.3   \n",
       "18   GCF_000002515.2      PRJNA12377   SAMEA3138170             NaN   \n",
       "19   GCF_000002525.2      PRJNA12414   SAMEA3138171             NaN   \n",
       "\n",
       "     ncbi_refseq_category taxonID specific_host  \\\n",
       "0        reference genome    7227          7227   \n",
       "1        reference genome    9606          9606   \n",
       "2        reference genome   10090         10090   \n",
       "3        reference genome    3702          3702   \n",
       "4   representative genome    9785          9785   \n",
       "5   representative genome  441960         37727   \n",
       "6        reference genome    7955          7955   \n",
       "7   representative genome    6500          6500   \n",
       "8   representative genome    7668          7668   \n",
       "9                      na    9615          9612   \n",
       "10  representative genome   13616         13616   \n",
       "11  representative genome    7070          7070   \n",
       "12  representative genome    5855          5855   \n",
       "13  representative genome    3218          3218   \n",
       "14  representative genome    5741          5741   \n",
       "15  representative genome  185431          5691   \n",
       "16                     na  227086        227086   \n",
       "17  representative genome  242507        318829   \n",
       "18  representative genome   28985         28985   \n",
       "19  representative genome  284591          4952   \n",
       "\n",
       "                       scientificName             cloneStrain  \\\n",
       "0             Drosophila melanogaster                     NaN   \n",
       "1                        Homo sapiens                     NaN   \n",
       "2                        Mus musculus                     NaN   \n",
       "3                Arabidopsis thaliana        ecotype=Columbia   \n",
       "4                  Loxodonta africana                     NaN   \n",
       "5    Talaromyces marneffei ATCC 18224       strain=ATCC 18224   \n",
       "6                         Danio rerio                     NaN   \n",
       "7                 Aplysia californica                     NaN   \n",
       "8       Strongylocentrotus purpuratus                     NaN   \n",
       "9              Canis lupus familiaris             breed=boxer   \n",
       "10              Monodelphis domestica                     NaN   \n",
       "11                Tribolium castaneum      strain=Georgia GA2   \n",
       "12                   Plasmodium vivax                     NaN   \n",
       "13               Physcomitrium patens   ecotype=Gransden 2004   \n",
       "14               Giardia intestinalis            strain=WB C6   \n",
       "15  Trypanosoma brucei brucei TREU927  strain=927/4 GUTat10.1   \n",
       "16                Bigelowiella natans                     NaN   \n",
       "17           Pyricularia oryzae 70-15            strain=70-15   \n",
       "18               Kluyveromyces lactis      strain=NRRL Y-1140   \n",
       "19        Yarrowia lipolytica CLIB122          strain=CLIB122   \n",
       "\n",
       "      source_mat_id  ... EarliestDateCollected MaterialSample lifeStage  \\\n",
       "0               NaN  ...                   NaN            NaN       NaN   \n",
       "1               NaN  ...                   NaN            NaN       NaN   \n",
       "2               NaN  ...                   NaN            NaN       NaN   \n",
       "3               NaN  ...                   NaN            NaN       NaN   \n",
       "4        ISIS603380  ...                   NaN            NaN       NaN   \n",
       "5               NaN  ...                   NaN            NaN       NaN   \n",
       "6               NaN  ...                   NaN            NaN       NaN   \n",
       "7             F4 #8  ...                   NaN            NaN       NaN   \n",
       "8           Spur 01  ...                   NaN            NaN       NaN   \n",
       "9             Tasha  ...                   NaN            NaN       NaN   \n",
       "10              NaN  ...                   NaN            NaN       NaN   \n",
       "11              NaN  ...                   NaN            NaN       NaN   \n",
       "12       Salvador I  ...                   NaN            NaN       NaN   \n",
       "13              NaN  ...                   NaN            NaN       NaN   \n",
       "14              NaN  ...                   NaN            NaN       NaN   \n",
       "15  927/4 GUTat10.1  ...                   NaN            NaN       NaN   \n",
       "16              NaN  ...                   NaN            NaN       NaN   \n",
       "17              NaN  ...                   NaN            NaN       NaN   \n",
       "18              NaN  ...                   NaN            NaN       NaN   \n",
       "19              NaN  ...                   NaN            NaN       NaN   \n",
       "\n",
       "   materialSampleID catalogNumber            sex   ncbi_biosample_email  \\\n",
       "0               NaN           NaN          mixed  celniker@fruitfly.org   \n",
       "1               NaN           NaN            NaN                    NaN   \n",
       "2               NaN           NaN            NaN                    NaN   \n",
       "3               NaN           NaN            NaN                    NaN   \n",
       "4        ISIS603380           NaN         female                    NaN   \n",
       "5               NaN           NaN            NaN                    NaN   \n",
       "6               NaN           NaN            NaN                    NaN   \n",
       "7             F4 #8           NaN  hermaphrodite                    NaN   \n",
       "8           Spur 01           NaN           male                    NaN   \n",
       "9             Tasha           NaN         female                    NaN   \n",
       "10              NaN           NaN            NaN                    NaN   \n",
       "11              NaN           NaN            NaN                    NaN   \n",
       "12       Salvador I           NaN            NaN                    NaN   \n",
       "13              NaN           NaN            NaN                    NaN   \n",
       "14              NaN           NaN            NaN                    NaN   \n",
       "15              NaN           NaN            NaN                    NaN   \n",
       "16              NaN           NaN            NaN                    NaN   \n",
       "17              NaN           NaN            NaN                    NaN   \n",
       "18              NaN           NaN            NaN                    NaN   \n",
       "19              NaN           NaN            NaN                    NaN   \n",
       "\n",
       "   ncbi_biosample_lab      ncbi_biosample_scientificName  \\\n",
       "0                 NaN            Drosophila melanogaster   \n",
       "1                 NaN                                NaN   \n",
       "2                 NaN                                NaN   \n",
       "3                 NaN               Arabidopsis thaliana   \n",
       "4                 NaN                 Loxodonta africana   \n",
       "5                 NaN   Talaromyces marneffei ATCC 18224   \n",
       "6                 NaN                        Danio rerio   \n",
       "7                 NaN                Aplysia californica   \n",
       "8                 NaN      Strongylocentrotus purpuratus   \n",
       "9                 NaN             Canis lupus familiaris   \n",
       "10                NaN              Monodelphis domestica   \n",
       "11                NaN                Tribolium castaneum   \n",
       "12                NaN                   Plasmodium vivax   \n",
       "13                NaN               Physcomitrium patens   \n",
       "14                NaN               Giardia intestinalis   \n",
       "15                NaN  Trypanosoma brucei brucei TREU927   \n",
       "16                NaN                                NaN   \n",
       "17                NaN           Pyricularia oryzae 70-15   \n",
       "18                NaN               Kluyveromyces lactis   \n",
       "19                NaN        Yarrowia lipolytica CLIB122   \n",
       "\n",
       "                       ncbi_biosample_attributes_json  \n",
       "0   {\"strain\": \"iso-1\", \"age\": \"missing\", \"biomate...  \n",
       "1                                                 NaN  \n",
       "2                                                 NaN  \n",
       "3                             {\"ecotype\": \"Columbia\"}  \n",
       "4   {\"sex\": \"female\", \"geo_loc_name\": \"USA: San Di...  \n",
       "5   {\"strain\": \"ATCC 18224\", \"culture_collection\":...  \n",
       "6   {\"strain\": \"Tuebingen\", \"note\": \"combined samp...  \n",
       "7        {\"sex\": \"hermaphrodite\", \"isolate\": \"F4 #8\"}  \n",
       "8   {\"isolate\": \"Spur 01\", \"sex\": \"male\", \"cell_ty...  \n",
       "9   {\"sex\": \"female\", \"sub_species\": \"familiaris\",...  \n",
       "10                            {\"pop-variant\": \"pop1\"}  \n",
       "11  {\"strain\": \"Georgia GA2\", \"collection_date\": \"...  \n",
       "12                          {\"isolate\": \"Salvador I\"}  \n",
       "13  {\"Sample Number\": \"4088616\", \"ecotype\": \"Grans...  \n",
       "14  {\"strain\": \"WB C6\", \"culture_collection\": \"ATC...  \n",
       "15  {\"strain\": \"927/4 GUTat10.1\", \"sub_species\": \"...  \n",
       "16                                                NaN  \n",
       "17                                {\"strain\": \"70-15\"}  \n",
       "18  {\"Alias\": \"GCA_000002515\", \"Broker name\": \"Eur...  \n",
       "19  {\"Alias\": \"GCA_000002525\", \"Broker name\": \"Eur...  \n",
       "\n",
       "[20 rows x 43 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "7ad5498d",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "df.to_csv(f\"/data/arga-data/assembly_summary_{genome_directory}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "42b29997",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "#df = pd.read_csv(f\"/data/arga-data/assembly_summary_{genome_directory}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "bed4d65e",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "#df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0c3259c",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.9.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
