{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9df0567f",
   "metadata": {},
   "source": [
    "# NCBI RefSeq ingestion script\n",
    "\n",
    "This script reads the listing for the NCBI refseq via the URL https://ftp.ncbi.nih.gov/genomes/refseq/assembly_summary_refseq.txt and attempts to map the columns to Darwin Core (DwC), using any of the DwC extenesions that might apply."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4998905",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_table(\"https://ftp.ncbi.nih.gov/genomes/refseq/assembly_summary_refseq.txt\", header=1, dtype = 'object', parse_dates=['seq_rel_date'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a29aa8a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 261398 entries, 0 to 261397\n",
      "Data columns (total 23 columns):\n",
      " #   Column                     Non-Null Count   Dtype         \n",
      "---  ------                     --------------   -----         \n",
      " 0   # assembly_accession       261398 non-null  object        \n",
      " 1   bioproject                 261260 non-null  object        \n",
      " 2   biosample                  249679 non-null  object        \n",
      " 3   wgs_master                 218468 non-null  object        \n",
      " 4   refseq_category            261398 non-null  object        \n",
      " 5   taxid                      261398 non-null  object        \n",
      " 6   species_taxid              261398 non-null  object        \n",
      " 7   organism_name              261398 non-null  object        \n",
      " 8   infraspecific_name         242405 non-null  object        \n",
      " 9   isolate                    17724 non-null   object        \n",
      " 10  version_status             261398 non-null  object        \n",
      " 11  assembly_level             261398 non-null  object        \n",
      " 12  release_type               261398 non-null  object        \n",
      " 13  genome_rep                 261398 non-null  object        \n",
      " 14  seq_rel_date               261398 non-null  datetime64[ns]\n",
      " 15  asm_name                   261202 non-null  object        \n",
      " 16  submitter                  249498 non-null  object        \n",
      " 17  gbrs_paired_asm            261398 non-null  object        \n",
      " 18  paired_asm_comp            261398 non-null  object        \n",
      " 19  ftp_path                   261398 non-null  object        \n",
      " 20  excluded_from_refseq       498 non-null     object        \n",
      " 21  relation_to_type_material  27081 non-null   object        \n",
      " 22  asm_not_live_date          261398 non-null  object        \n",
      "dtypes: datetime64[ns](1), object(22)\n",
      "memory usage: 45.9+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b8c9c37",
   "metadata": {},
   "source": [
    "## Field mappings\n",
    "\n",
    "Attempt to map NCBI fields to known [DwC terms](https://dwc.tdwg.org/list/), including extensions\n",
    "\n",
    "NCBI fields described at https://ftp.ncbi.nih.gov/genomes/README_assembly_summary.txt\n",
    "\n",
    "| # | field name | field description |\n",
    "| ---- | ---- | ---- |\n",
    "| 1 | \"assembly_accession\" | Assembly accession: the assembly accession.version reported in this field is a unique identifier for the set of sequences in this particular version of the genome assembly. |\n",
    "| 2 | \"bioproject\" | BioProject: accession for the BioProject which produced the sequences in the genome assembly. A BioProject is a collection of biological data related to a single initiative, originating from a single organization or from a consortium. A BioProject record provides users a single place to find links to the diverse data types generated for that project. The record can be retrieved from the NCBI BioProject resource: https://www.ncbi.nlm.nih.gov/bioproject/ |\n",
    "| 3 | \"biosample\" | BioSample: accession for the BioSample from which the sequences in the genome assembly were obtained. A BioSample record contains a description of the biological source material used in experimental assays. The record can be retrieved from the NCBI BioSample resource: https://www.ncbi.nlm.nih.gov/biosample/ |\n",
    "| 4 | \"wgs_master\" | WGS-master: the GenBank Nucleotide accession and version for the master record of the Whole Genome Shotgun (WGS) project for the genome assembly. The master record can be retrieved from the NCBI Nucleotide resource: https://www.ncbi.nlm.nih.gov/nuccore Genome assemblies that are complete genomes, and those that are clone-based, do not have WGS-master records in which case this field will be empty. |\n",
    "| 5 | \"refseq_category\" | RefSeq Category: whether the assembly is a reference or representative genome in the NCBI Reference Sequence (RefSeq) project classification. Values: **reference genome** - a manually selected high quality genome assembly that NCBI and the community have identified as being important as a standard against which other data are compared; **representative genome** - a genome computationally or manually selected as a representative from among the best genomes available for a species or clade that does not have a designated reference genome; **na** - no RefSeq category assigned to this assembly Prokaryotes may have more than one reference or representative genome per species. For more information see: https://www.ncbi.nlm.nih.gov/refseq/about/prokaryotes/#referencegenome Eukaryotes have no more than one reference or representative genome per species. If there are no assemblies in RefSeq for a particular eukaryotic species, then the GenBank assembly that RefSeq would select as the best available for that species will be designated as the representative genome. Viruses may have one or more reference genomes per species. The representative genome designation is not applied to viruses. |\n",
    "| 6 | \"taxid\" | Taxonomy ID: the NCBI taxonomy identifier for the organism from which the genome assembly was derived. The NCBI Taxonomy Database is a curated classification and nomenclature for all of the organisms in the public sequence databases. The taxonomy record can be retrieved from the NCBI Taxonomy resource: https://www.ncbi.nlm.nih.gov/taxonomy/ |\n",
    "| 7 | \"species_taxid\" | Species taxonomy ID: the NCBI taxonomy identifier for the species from which the genome assembly was derived. The species taxid will differ from the organism taxid (column 6) only when the organism was reported at a sub- species or strain level. |\n",
    "| 8 | \"organism_name\" | Organism name: the scientific name of the organism from which the sequences in the genome assembly were derived. This name is taken from the NCBI Taxonomy record for the taxid specified in column 6. Some older taxids were assigned at the strain level and for these the organism name will include the strain. Current practice is only to assign taxids at the species level; for these the organism name will be just the species, however, the strain name will be reported in the infraspecific_name field (column 9). |\n",
    "| 9 | \"infraspecific_name\" | Infraspecific name: the strain, breed, cultivar or ecotype of the organism from which the sequences in the genome assembly were derived. Data are reported in the form tag=value, e.g. strain=AF16. Strain, breed, cultivar and ecotype are not expected to be used together, however, if they are then they will be reported in a list separated by \", /\". Empty if no strain, breed, cultivar or ecotype is specified on the genomic sequence records. |\n",
    "| 10 | \"isolate\" | Isolate: the individual isolate from which the sequences in the genome assembly were derived. Empty if no isolate is specified on the genomic sequence records. |\n",
    "| 11 | \"version_status\" | Version status: the release status for the genome assembly version. Values: **latest** - the most recent of all the versions for this assembly chain; **replaced** - this version has been replaced by a newer version of the assembly in the same chain; **suppressed** - this version of the assembly has been suppressed An assembly chain is the collection of all versions for the same assembly accession. |\n",
    "| 12 | \"assembly_level\" | Assembly level: the highest level of assembly for any object in the genome assembly. Values: **Complete genome** - all chromosomes are gapless and have no runs of 10 or more ambiguous bases (Ns), there are no unplaced or unlocalized scaffolds, and all the expected chromosomes are present (i.e. the assembly is not noted as having partial genome representation). Plasmids and organelles may or may not be included in the assembly but if present then the sequences are gapless. **Chromosome** - there is sequence for one or more chromosomes. This could be a completely sequenced chromosome without gaps or a chromosome containing scaffolds or contigs with gaps between them. There may also be unplaced or unlocalized scaffolds. Scaffold - some sequence contigs have been connected across gaps to create scaffolds, but the scaffolds are all unplaced or unlocalized. **Contig** - nothing is assembled beyond the level of sequence contigs |\n",
    "| 13 | \"release_type\" | Release type: whether this version of the genome assembly is a major, minor or patch release. Values: **Major** - changes from the previous assembly version result in a significant change to the coordinate system. The first version of an assembly is always a major release. Most subsequent genome assembly updates are also major releases. **Minor** - changes from the previous assembly version are limited to the following changes, none of which result in a significant change to the coordinate system of the primary assembly-unit: - adding, removing or changing a non-nuclear assembly-unit - dropping unplaced or unlocalized scaffolds - adding up to 50 unplaced or unlocalized scaffolds which are shorter than the current scaffold-N50 value - replacing a component with a gap of the same length Patch - the only change is the addition or modification of a patch assembly-unit. See the NCBI Assembly model web page (https://www.ncbi.nlm.nih.gov/assembly/ model/#asmb_def) for definitions of assembly-units and genome patches. |\n",
    "| 14 | \"genome_rep\" | Genome representation: whether the goal for the assembly was to represent the whole genome or only part of it. Values: **Full** - the data used to generate the assembly was obtained from the whole genome, as in Whole Genome Shotgun (WGS) assemblies for example. There may still be gaps in the assembly. **Partial** - the data used to generate the assembly came from only part of the genome. Most assemblies have full genome representation with a minority being partial genome representation. See the Assembly help web page (https://www.ncbi.nlm.nih.gov/assembly/help/) for reasons that the genome representation would be set to partial. |\n",
    "| 15 | \"seq_rel_date\" | Sequence release date: the date the sequences in the genome assembly were released in the International Nucleotide Sequence Database Collaboration (INSDC) databases, i.e. DDBJ, ENA or GenBank. |\n",
    "| 16 | \"asm_name\" | Assembly name: the submitter's name for the genome assembly, when one was provided, otherwise a default name, in the form ASM#####v#, is provided by NCBI. Assembly names are not unique. |\n",
    "| 17 | \"submitter\" | Submitter: the submitting consortium or first position if a list of organizations. The full submitter information is available in the NCBI BioProject resource: www.ncbi.nlm.nih.gov/bioproject/ |\n",
    "| 18 | \"gbrs_paired_asm\" | GenBank/RefSeq paired assembly: the accession.version of the GenBank assembly that is paired to the given RefSeq assembly, or vice-versa. \"na\" is reported if the assembly is unpaired. |\n",
    "| 19 | \"paired_asm_comp\" | Paired assembly comparison: whether the paired GenBank & RefSeq assemblies are identical or different. Values: **identical** - GenBank and RefSeq assemblies are identical; **different** - GenBank and RefSeq assemblies are not identical; **na** - not applicable since the assembly is unpaired |\n",
    "| 20 | \"ftp_path\" | FTP path: the path to the directory on the NCBI genomes FTP site from which data for this genome assembly can be downloaded. |\n",
    "| 21 | \"excluded_from_refseq\" | Excluded from RefSeq: reasons the assembly was excluded from the NCBI Reference Sequence (RefSeq) project, including any assembly anomalies. See: https://www.ncbi.nlm.nih.gov/assembly/help/anomnotrefseq/ |\n",
    "| 22 | \"relation_to_type_material\" | Relation to type material: contains a value if the sequences in the genome assembly were derived from type material. Values: **assembly from type material** - the sequences in the genome assembly were derived from type material; **assembly from synonym type material** - the sequences in the genome assembly were derived from synonym type material; **assembly from pathotype material** - the sequences in the genome assembly were derived from pathovar material;** **assembly designated as neotype** - the sequences in the genome assembly were derived from neotype material; **assembly designated as reftype** - the sequences in the genome assembly were derived from reference material where type material never was available and is not likely to ever be available ICTV species exemplar - the International Committee on Taxonomy of Viruses (ICTV) designated the genome assembly as the exemplar for the virus species ICTV additional isolate - the International Committee on Taxonomy of Viruses (ICTV) designated the genome assembly an additional isolate for the virus species |\n",
    "| 23 | \"asm_not_live_date\" | Assembly no longer live date: the date the assembly transitioned from version_status latest to either replaced or suppressed. When the assembly is in status latest, \"na\" is reported. |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "348ef5b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1:1 mapping to GGBN terms\n",
    "# NCBI fields described at https://ftp.ncbi.nih.gov/genomes/README_assembly_summary.txt\n",
    "new_names = {\n",
    "  '# assembly_accession': 'geneticAccessionNumber',\n",
    "  'bioproject': 'ncbi_bioproject',\n",
    "  'biosample': 'ncbi_biosample',\n",
    "  'wgs_master': 'ncbi_nuccore',\n",
    "  'refseq_category': 'ncbi_refseq_category',\n",
    "  'taxid': 'taxonID',\n",
    "  'species_taxid': 'specific_host',\n",
    "  'organism_name': 'scientificName',\n",
    "  'infraspecific_name': 'cloneStrain',\n",
    "  'isolate': 'source_mat_id',\n",
    "  'version_status': 'ncbi_version_status',\n",
    "  'assembly_level': 'ncbi_assembly_level',\n",
    "  'release_type': 'ncbi_release_type',\n",
    "  'genome_rep': 'ncbi_genome_rep',\n",
    "  'seq_rel_date': 'eventDate',\n",
    "  'asm_name': 'otherCatalogNumbers',\n",
    "  'submitter': 'redcordedBy',\n",
    "  'gbrs_paired_asm': 'ncbi_gbrs_paired_asm',\n",
    "  'paired_asm_comp': 'ncbi_paired_asm_comp',\n",
    "  'ftp_path': 'geneticAccessionURI',\n",
    "  'excluded_from_refseq': 'ncbi_excluded_from_refseq',\n",
    "  'relation_to_type_material': 'ncbi_relation_to_type_material',\n",
    "  'asm_not_live_date': 'ncbiasm_not_live_date'\n",
    "}\n",
    "\n",
    "#do rename\n",
    "df.rename(\n",
    "  columns = new_names, \n",
    "  inplace = True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3ca73f5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneticAccessionNumber</th>\n",
       "      <th>ncbi_bioproject</th>\n",
       "      <th>ncbi_biosample</th>\n",
       "      <th>ncbi_nuccore</th>\n",
       "      <th>ncbi_refseq_category</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>specific_host</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>infraspecificEpithet</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>...</th>\n",
       "      <th>ncbi_genome_rep</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>otherCatalogNumbers</th>\n",
       "      <th>redcordedBy</th>\n",
       "      <th>ncbi_gbrs_paired_asm</th>\n",
       "      <th>ncbi_paired_asm_comp</th>\n",
       "      <th>geneticAccessionURI</th>\n",
       "      <th>ncbi_excluded_from_refseq</th>\n",
       "      <th>ncbi_relation_to_type_material</th>\n",
       "      <th>ncbiasm_not_live_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>PRJNA164</td>\n",
       "      <td>SAMN02803731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2014-08-01</td>\n",
       "      <td>Release 6 plus ISO1 MT</td>\n",
       "      <td>The FlyBase Consortium/Berkeley Drosophila Gen...</td>\n",
       "      <td>GCA_000001215.4</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>PRJNA168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-02-03</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>GCA_000001405.29</td>\n",
       "      <td>different</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>PRJNA169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2020-06-24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>GCA_000001635.9</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>PRJNA116</td>\n",
       "      <td>SAMN03081427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2018-03-15</td>\n",
       "      <td>TAIR10.1</td>\n",
       "      <td>The Arabidopsis Information Resource (TAIR)</td>\n",
       "      <td>GCA_000001735.2</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>PRJNA70973</td>\n",
       "      <td>SAMN02953622</td>\n",
       "      <td>AAGU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2009-07-15</td>\n",
       "      <td>Loxafr3.0</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>GCA_000001905.1</td>\n",
       "      <td>different</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GCF_000007065.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>192952</td>\n",
       "      <td>2209</td>\n",
       "      <td>Methanosarcina mazei Go1</td>\n",
       "      <td>strain=Go1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2002-05-20</td>\n",
       "      <td>ASM706v1</td>\n",
       "      <td>Gottingen Genomics Laboratory</td>\n",
       "      <td>GCA_000007065.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>GCF_000007085.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>273068</td>\n",
       "      <td>911092</td>\n",
       "      <td>Caldanaerobacter subterraneus subsp. tengconge...</td>\n",
       "      <td>strain=MB4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2002-05-14</td>\n",
       "      <td>ASM708v1</td>\n",
       "      <td>Beijing Center.HGP</td>\n",
       "      <td>GCA_000007085.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>assembly from type material</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>GCF_000007105.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603579</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>264203</td>\n",
       "      <td>542</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821</td>\n",
       "      <td>strain=ZM4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2010-01-12</td>\n",
       "      <td>ASM710v1</td>\n",
       "      <td>Macrogen Inc.</td>\n",
       "      <td>GCA_000007105.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>GCF_000007125.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>224914</td>\n",
       "      <td>29459</td>\n",
       "      <td>Brucella melitensis bv. 1 str. 16M</td>\n",
       "      <td>strain=16M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2001-12-28</td>\n",
       "      <td>ASM712v1</td>\n",
       "      <td>Integrated Genomics</td>\n",
       "      <td>GCA_000007125.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>assembly from type material</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>GCF_000007145.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603845</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>190485</td>\n",
       "      <td>339</td>\n",
       "      <td>Xanthomonas campestris pv. campestris str. ATC...</td>\n",
       "      <td>strain=ATCC 33913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2002-05-24</td>\n",
       "      <td>ASM714v1</td>\n",
       "      <td>Sao Paulo state (Brazil) Consortium</td>\n",
       "      <td>GCA_000007145.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>assembly from type material</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   geneticAccessionNumber ncbi_bioproject ncbi_biosample    ncbi_nuccore  \\\n",
       "0         GCF_000001215.4        PRJNA164   SAMN02803731             NaN   \n",
       "1        GCF_000001405.40        PRJNA168            NaN             NaN   \n",
       "2        GCF_000001635.27        PRJNA169            NaN             NaN   \n",
       "3         GCF_000001735.4        PRJNA116   SAMN03081427             NaN   \n",
       "4         GCF_000001905.1      PRJNA70973   SAMN02953622  AAGU00000000.3   \n",
       "..                    ...             ...            ...             ...   \n",
       "95        GCF_000007065.1     PRJNA224116   SAMN02603290             NaN   \n",
       "96        GCF_000007085.1     PRJNA224116   SAMN02603009             NaN   \n",
       "97        GCF_000007105.1     PRJNA224116   SAMN02603579             NaN   \n",
       "98        GCF_000007125.1     PRJNA224116   SAMN02603416             NaN   \n",
       "99        GCF_000007145.1     PRJNA224116   SAMN02603845             NaN   \n",
       "\n",
       "     ncbi_refseq_category taxonID specific_host  \\\n",
       "0        reference genome    7227          7227   \n",
       "1        reference genome    9606          9606   \n",
       "2        reference genome   10090         10090   \n",
       "3        reference genome    3702          3702   \n",
       "4   representative genome    9785          9785   \n",
       "..                    ...     ...           ...   \n",
       "95  representative genome  192952          2209   \n",
       "96  representative genome  273068        911092   \n",
       "97  representative genome  264203           542   \n",
       "98  representative genome  224914         29459   \n",
       "99                     na  190485           339   \n",
       "\n",
       "                                       scientificName infraspecificEpithet  \\\n",
       "0                             Drosophila melanogaster                  NaN   \n",
       "1                                        Homo sapiens                  NaN   \n",
       "2                                        Mus musculus                  NaN   \n",
       "3                                Arabidopsis thaliana     ecotype=Columbia   \n",
       "4                                  Loxodonta africana                  NaN   \n",
       "..                                                ...                  ...   \n",
       "95                           Methanosarcina mazei Go1           strain=Go1   \n",
       "96  Caldanaerobacter subterraneus subsp. tengconge...           strain=MB4   \n",
       "97  Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821           strain=ZM4   \n",
       "98                 Brucella melitensis bv. 1 str. 16M           strain=16M   \n",
       "99  Xanthomonas campestris pv. campestris str. ATC...    strain=ATCC 33913   \n",
       "\n",
       "   source_mat_id  ... ncbi_genome_rep  eventDate     otherCatalogNumbers  \\\n",
       "0            NaN  ...            Full 2014-08-01  Release 6 plus ISO1 MT   \n",
       "1            NaN  ...            Full 2022-02-03              GRCh38.p14   \n",
       "2            NaN  ...            Full 2020-06-24                  GRCm39   \n",
       "3            NaN  ...            Full 2018-03-15                TAIR10.1   \n",
       "4     ISIS603380  ...            Full 2009-07-15               Loxafr3.0   \n",
       "..           ...  ...             ...        ...                     ...   \n",
       "95           NaN  ...            Full 2002-05-20                ASM706v1   \n",
       "96           NaN  ...            Full 2002-05-14                ASM708v1   \n",
       "97           NaN  ...            Full 2010-01-12                ASM710v1   \n",
       "98           NaN  ...            Full 2001-12-28                ASM712v1   \n",
       "99           NaN  ...            Full 2002-05-24                ASM714v1   \n",
       "\n",
       "                                          redcordedBy ncbi_gbrs_paired_asm  \\\n",
       "0   The FlyBase Consortium/Berkeley Drosophila Gen...      GCA_000001215.4   \n",
       "1                         Genome Reference Consortium     GCA_000001405.29   \n",
       "2                         Genome Reference Consortium      GCA_000001635.9   \n",
       "3         The Arabidopsis Information Resource (TAIR)      GCA_000001735.2   \n",
       "4                                     Broad Institute      GCA_000001905.1   \n",
       "..                                                ...                  ...   \n",
       "95                      Gottingen Genomics Laboratory      GCA_000007065.1   \n",
       "96                                 Beijing Center.HGP      GCA_000007085.1   \n",
       "97                                      Macrogen Inc.      GCA_000007105.1   \n",
       "98                                Integrated Genomics      GCA_000007125.1   \n",
       "99                Sao Paulo state (Brazil) Consortium      GCA_000007145.1   \n",
       "\n",
       "   ncbi_paired_asm_comp                                geneticAccessionURI  \\\n",
       "0             identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1             different  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "2             identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "3             identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "4             different  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "..                  ...                                                ...   \n",
       "95            identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "96            identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "97            identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "98            identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "99            identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "\n",
       "   ncbi_excluded_from_refseq ncbi_relation_to_type_material  \\\n",
       "0                        NaN                            NaN   \n",
       "1                        NaN                            NaN   \n",
       "2                        NaN                            NaN   \n",
       "3                        NaN                            NaN   \n",
       "4                        NaN                            NaN   \n",
       "..                       ...                            ...   \n",
       "95                       NaN                            NaN   \n",
       "96                       NaN    assembly from type material   \n",
       "97                       NaN                            NaN   \n",
       "98                       NaN    assembly from type material   \n",
       "99                       NaN    assembly from type material   \n",
       "\n",
       "   ncbiasm_not_live_date  \n",
       "0                     na  \n",
       "1                     na  \n",
       "2                     na  \n",
       "3                     na  \n",
       "4                     na  \n",
       "..                   ...  \n",
       "95                    na  \n",
       "96                    na  \n",
       "97                    na  \n",
       "98                    na  \n",
       "99                    na  \n",
       "\n",
       "[100 rows x 23 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7ad5498d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv('assembly_summary_refseq.csv', index= None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "42b29997",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"assembly_summary_refseq.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bed4d65e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ad4f9719",
   "metadata": {},
   "outputs": [],
   "source": [
    "dictn = {\"Category 3\": \"1km\",\n",
    "       \"Category 2\": \"10km\",\n",
    "       \"Category 1\": \"WITHHOLD\"\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "94900b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "select = ['Category 3','Category 2','Category 1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7fae7a7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['generalisation'] = df['sensitivityClass']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c88349d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['generalisation'] = df['generalisation'].replace(dictn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e501d475",
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitive = df[(df['speciesID'] == df['taxonID']) & df['sensitivityClass'].isin(select)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f479ccb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dcterms_rightsHolder</th>\n",
       "      <th>dcterms_rights</th>\n",
       "      <th>dcterms_language</th>\n",
       "      <th>dcterms_type</th>\n",
       "      <th>dcterms_modified</th>\n",
       "      <th>dcterms_available</th>\n",
       "      <th>speciesID</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>kingdomID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>...</th>\n",
       "      <th>protectedInNSW</th>\n",
       "      <th>sensitivityClass</th>\n",
       "      <th>TSProfileID</th>\n",
       "      <th>countryConservation</th>\n",
       "      <th>highThreatWeed</th>\n",
       "      <th>widelyCultivatedNativeSpecies</th>\n",
       "      <th>CAMBA</th>\n",
       "      <th>JAMBA</th>\n",
       "      <th>ROKAMBA</th>\n",
       "      <th>generalisation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2011-11-11T11:23:09+11:00</td>\n",
       "      <td>1995-12-15T12:48:15+11:00</td>\n",
       "      <td>6</td>\n",
       "      <td>Species</td>\n",
       "      <td>138</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10975.0</td>\n",
       "      <td>Endangered</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2011-11-11T11:23:06+11:00</td>\n",
       "      <td>1995-12-15T12:49:15+11:00</td>\n",
       "      <td>162</td>\n",
       "      <td>Species</td>\n",
       "      <td>138</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10562.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2012-07-23T18:04:52.683+10:00</td>\n",
       "      <td>1995-12-15T12:50:21+11:00</td>\n",
       "      <td>333</td>\n",
       "      <td>Species</td>\n",
       "      <td>138</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10821.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2011-11-11T11:23:06+11:00</td>\n",
       "      <td>1995-12-15T12:50:32+11:00</td>\n",
       "      <td>363</td>\n",
       "      <td>Species</td>\n",
       "      <td>138</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10561.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2011-11-11T11:23:04+11:00</td>\n",
       "      <td>1995-12-15T12:50:42+11:00</td>\n",
       "      <td>390</td>\n",
       "      <td>Species</td>\n",
       "      <td>138</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 2</td>\n",
       "      <td>10413.0</td>\n",
       "      <td>Vulnerable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21986</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2020-01-23T17:14:13.01+11:00</td>\n",
       "      <td>2019-06-24T10:36:48.61+10:00</td>\n",
       "      <td>23014</td>\n",
       "      <td>Subspecies</td>\n",
       "      <td>139</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>20148.0</td>\n",
       "      <td>Critically Endangered</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22658</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2020-06-23T19:56:43.85+10:00</td>\n",
       "      <td>2020-06-23T15:48:30.817+10:00</td>\n",
       "      <td>23737</td>\n",
       "      <td>Species</td>\n",
       "      <td>139</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22659</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2020-06-23T19:51:08.733+10:00</td>\n",
       "      <td>2020-06-23T19:47:12.607+10:00</td>\n",
       "      <td>23738</td>\n",
       "      <td>Species</td>\n",
       "      <td>139</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 2</td>\n",
       "      <td>20365.0</td>\n",
       "      <td>Endangered</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23131</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2021-12-31T11:53:58.467+11:00</td>\n",
       "      <td>2021-03-17T11:44:46.88+11:00</td>\n",
       "      <td>24268</td>\n",
       "      <td>Species</td>\n",
       "      <td>139</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Category 2</td>\n",
       "      <td>10668.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10km</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23133</th>\n",
       "      <td>NSW Dept of Planning, Industry and Environment</td>\n",
       "      <td>CC-BY 4.0</td>\n",
       "      <td>en</td>\n",
       "      <td>service</td>\n",
       "      <td>2021-03-17T14:00:40.507+11:00</td>\n",
       "      <td>2021-03-17T13:48:54.217+11:00</td>\n",
       "      <td>24270</td>\n",
       "      <td>Species</td>\n",
       "      <td>139</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10166.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1km</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 dcterms_rightsHolder dcterms_rights  \\\n",
       "5      NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "159    NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "326    NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "356    NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "382    NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "...                                               ...            ...   \n",
       "21986  NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "22658  NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "22659  NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "23131  NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "23133  NSW Dept of Planning, Industry and Environment      CC-BY 4.0   \n",
       "\n",
       "      dcterms_language dcterms_type               dcterms_modified  \\\n",
       "5                   en      service      2011-11-11T11:23:09+11:00   \n",
       "159                 en      service      2011-11-11T11:23:06+11:00   \n",
       "326                 en      service  2012-07-23T18:04:52.683+10:00   \n",
       "356                 en      service      2011-11-11T11:23:06+11:00   \n",
       "382                 en      service      2011-11-11T11:23:04+11:00   \n",
       "...                ...          ...                            ...   \n",
       "21986               en      service   2020-01-23T17:14:13.01+11:00   \n",
       "22658               en      service   2020-06-23T19:56:43.85+10:00   \n",
       "22659               en      service  2020-06-23T19:51:08.733+10:00   \n",
       "23131               en      service  2021-12-31T11:53:58.467+11:00   \n",
       "23133               en      service  2021-03-17T14:00:40.507+11:00   \n",
       "\n",
       "                   dcterms_available  speciesID   taxonRank  kingdomID  \\\n",
       "5          1995-12-15T12:48:15+11:00          6     Species        138   \n",
       "159        1995-12-15T12:49:15+11:00        162     Species        138   \n",
       "326        1995-12-15T12:50:21+11:00        333     Species        138   \n",
       "356        1995-12-15T12:50:32+11:00        363     Species        138   \n",
       "382        1995-12-15T12:50:42+11:00        390     Species        138   \n",
       "...                              ...        ...         ...        ...   \n",
       "21986   2019-06-24T10:36:48.61+10:00      23014  Subspecies        139   \n",
       "22658  2020-06-23T15:48:30.817+10:00      23737     Species        139   \n",
       "22659  2020-06-23T19:47:12.607+10:00      23738     Species        139   \n",
       "23131   2021-03-17T11:44:46.88+11:00      24268     Species        139   \n",
       "23133  2021-03-17T13:48:54.217+11:00      24270     Species        139   \n",
       "\n",
       "        kingdom  ...  protectedInNSW sensitivityClass  TSProfileID  \\\n",
       "5      Animalia  ...            True       Category 3      10975.0   \n",
       "159    Animalia  ...            True       Category 3      10562.0   \n",
       "326    Animalia  ...            True       Category 3      10821.0   \n",
       "356    Animalia  ...            True       Category 3      10561.0   \n",
       "382    Animalia  ...            True       Category 2      10413.0   \n",
       "...         ...  ...             ...              ...          ...   \n",
       "21986   Plantae  ...           False       Category 3      20148.0   \n",
       "22658   Plantae  ...            True       Category 2          NaN   \n",
       "22659   Plantae  ...            True       Category 2      20365.0   \n",
       "23131   Plantae  ...           False       Category 2      10668.0   \n",
       "23133   Plantae  ...           False       Category 3      10166.0   \n",
       "\n",
       "         countryConservation  highThreatWeed widelyCultivatedNativeSpecies  \\\n",
       "5                 Endangered             NaN                           NaN   \n",
       "159               Not Listed             NaN                           NaN   \n",
       "326               Not Listed             NaN                           NaN   \n",
       "356               Not Listed             NaN                           NaN   \n",
       "382               Vulnerable             NaN                           NaN   \n",
       "...                      ...             ...                           ...   \n",
       "21986  Critically Endangered             NaN                           NaN   \n",
       "22658             Not Listed             NaN                           NaN   \n",
       "22659             Endangered             NaN                           NaN   \n",
       "23131             Not Listed             NaN                           NaN   \n",
       "23133             Not Listed             NaN                           NaN   \n",
       "\n",
       "       CAMBA  JAMBA ROKAMBA  generalisation  \n",
       "5      False  False   False             1km  \n",
       "159    False  False   False             1km  \n",
       "326    False  False   False             1km  \n",
       "356    False  False   False             1km  \n",
       "382    False  False   False            10km  \n",
       "...      ...    ...     ...             ...  \n",
       "21986  False  False   False             1km  \n",
       "22658  False  False   False            10km  \n",
       "22659  False  False   False            10km  \n",
       "23131  False  False   False            10km  \n",
       "23133  False  False   False             1km  \n",
       "\n",
       "[197 rows x 52 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9db514cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitive = sensitive[['taxonRank', 'kingdom' ,'class','order','family','genus','scientificName','specificEpithet','vernacularName','establishmentMeans','stateConservation','protectedInNSW','sensitivityClass','TSProfileID','countryConservation','dcterms_modified','speciesID','taxonID','generalisation']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b19129a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/cha801/.local/lib/python3.8/site-packages/pandas/core/frame.py:3607: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._set_item(key, value)\n"
     ]
    }
   ],
   "source": [
    "sensitive['status'] = sensitive['stateConservation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d046dc87",
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitive = sensitive.rename(columns={\"stateConservation\": \"sourceStatus\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3586f378",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>specificEpithet</th>\n",
       "      <th>vernacularName</th>\n",
       "      <th>establishmentMeans</th>\n",
       "      <th>sourceStatus</th>\n",
       "      <th>protectedInNSW</th>\n",
       "      <th>sensitivityClass</th>\n",
       "      <th>TSProfileID</th>\n",
       "      <th>countryConservation</th>\n",
       "      <th>dcterms_modified</th>\n",
       "      <th>speciesID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>generalisation</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Species</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Psittaciformes</td>\n",
       "      <td>Cacatuidae</td>\n",
       "      <td>Callocephalon</td>\n",
       "      <td>Callocephalon fimbriatum</td>\n",
       "      <td>fimbriatum</td>\n",
       "      <td>Gang-gang Cockatoo</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Vulnerable</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10975.0</td>\n",
       "      <td>Endangered</td>\n",
       "      <td>2011-11-11T11:23:09+11:00</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1km</td>\n",
       "      <td>Vulnerable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>Species</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Strigiformes</td>\n",
       "      <td>Strigidae</td>\n",
       "      <td>Ninox</td>\n",
       "      <td>Ninox strenua</td>\n",
       "      <td>strenua</td>\n",
       "      <td>Powerful Owl</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Vulnerable</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10562.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>2011-11-11T11:23:06+11:00</td>\n",
       "      <td>162</td>\n",
       "      <td>162</td>\n",
       "      <td>1km</td>\n",
       "      <td>Vulnerable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>Species</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Strigiformes</td>\n",
       "      <td>Tytonidae</td>\n",
       "      <td>Tyto</td>\n",
       "      <td>Tyto tenebricosa</td>\n",
       "      <td>tenebricosa</td>\n",
       "      <td>Sooty Owl</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Vulnerable</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10821.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>2012-07-23T18:04:52.683+10:00</td>\n",
       "      <td>333</td>\n",
       "      <td>333</td>\n",
       "      <td>1km</td>\n",
       "      <td>Vulnerable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>Species</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Strigiformes</td>\n",
       "      <td>Strigidae</td>\n",
       "      <td>Ninox</td>\n",
       "      <td>Ninox connivens</td>\n",
       "      <td>connivens</td>\n",
       "      <td>Barking Owl</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Vulnerable</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10561.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>2011-11-11T11:23:06+11:00</td>\n",
       "      <td>363</td>\n",
       "      <td>363</td>\n",
       "      <td>1km</td>\n",
       "      <td>Vulnerable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>Species</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Reptilia</td>\n",
       "      <td>Squamata</td>\n",
       "      <td>Elapidae</td>\n",
       "      <td>Hoplocephalus</td>\n",
       "      <td>Hoplocephalus bungaroides</td>\n",
       "      <td>bungaroides</td>\n",
       "      <td>Broad-headed Snake</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Endangered</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 2</td>\n",
       "      <td>10413.0</td>\n",
       "      <td>Vulnerable</td>\n",
       "      <td>2011-11-11T11:23:04+11:00</td>\n",
       "      <td>390</td>\n",
       "      <td>390</td>\n",
       "      <td>10km</td>\n",
       "      <td>Endangered</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21986</th>\n",
       "      <td>Subspecies</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Anthosachne</td>\n",
       "      <td>Anthosachne kingiana subsp. kingiana</td>\n",
       "      <td>kingiana</td>\n",
       "      <td>Philip Island Wheat Grass</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Critically Endangered</td>\n",
       "      <td>False</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>20148.0</td>\n",
       "      <td>Critically Endangered</td>\n",
       "      <td>2020-01-23T17:14:13.01+11:00</td>\n",
       "      <td>23014</td>\n",
       "      <td>23014</td>\n",
       "      <td>1km</td>\n",
       "      <td>Critically Endangered</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22658</th>\n",
       "      <td>Species</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Orchidaceae</td>\n",
       "      <td>Caladenia</td>\n",
       "      <td>Caladenia tensa</td>\n",
       "      <td>tensa</td>\n",
       "      <td>Rigid Spider-orchid</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>2020-06-23T19:56:43.85+10:00</td>\n",
       "      <td>23737</td>\n",
       "      <td>23737</td>\n",
       "      <td>10km</td>\n",
       "      <td>Not Listed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22659</th>\n",
       "      <td>Species</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Orchidaceae</td>\n",
       "      <td>Caladenia</td>\n",
       "      <td>Caladenia atroclavia</td>\n",
       "      <td>atroclavia</td>\n",
       "      <td>Black-clubbed Spider-orchid</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>True</td>\n",
       "      <td>Category 2</td>\n",
       "      <td>20365.0</td>\n",
       "      <td>Endangered</td>\n",
       "      <td>2020-06-23T19:51:08.733+10:00</td>\n",
       "      <td>23738</td>\n",
       "      <td>23738</td>\n",
       "      <td>10km</td>\n",
       "      <td>Not Listed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23131</th>\n",
       "      <td>Species</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Orchidaceae</td>\n",
       "      <td>Prasophyllum</td>\n",
       "      <td>Prasophyllum sandrae</td>\n",
       "      <td>sandrae</td>\n",
       "      <td>Majors Creek Leek Orchid</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Critically Endangered</td>\n",
       "      <td>False</td>\n",
       "      <td>Category 2</td>\n",
       "      <td>10668.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>2021-12-31T11:53:58.467+11:00</td>\n",
       "      <td>24268</td>\n",
       "      <td>24268</td>\n",
       "      <td>10km</td>\n",
       "      <td>Critically Endangered</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23133</th>\n",
       "      <td>Species</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Flora</td>\n",
       "      <td>Myrtaceae</td>\n",
       "      <td>Backhousia</td>\n",
       "      <td>Backhousia subargentea</td>\n",
       "      <td>subargentea</td>\n",
       "      <td>Giant Ironwood</td>\n",
       "      <td>Alive in NSW, Native</td>\n",
       "      <td>Endangered</td>\n",
       "      <td>False</td>\n",
       "      <td>Category 3</td>\n",
       "      <td>10166.0</td>\n",
       "      <td>Not Listed</td>\n",
       "      <td>2021-03-17T14:00:40.507+11:00</td>\n",
       "      <td>24270</td>\n",
       "      <td>24270</td>\n",
       "      <td>1km</td>\n",
       "      <td>Endangered</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        taxonRank   kingdom     class           order       family  \\\n",
       "5         Species  Animalia      Aves  Psittaciformes   Cacatuidae   \n",
       "159       Species  Animalia      Aves    Strigiformes    Strigidae   \n",
       "326       Species  Animalia      Aves    Strigiformes    Tytonidae   \n",
       "356       Species  Animalia      Aves    Strigiformes    Strigidae   \n",
       "382       Species  Animalia  Reptilia        Squamata     Elapidae   \n",
       "...           ...       ...       ...             ...          ...   \n",
       "21986  Subspecies   Plantae     Flora           Flora      Poaceae   \n",
       "22658     Species   Plantae     Flora           Flora  Orchidaceae   \n",
       "22659     Species   Plantae     Flora           Flora  Orchidaceae   \n",
       "23131     Species   Plantae     Flora           Flora  Orchidaceae   \n",
       "23133     Species   Plantae     Flora           Flora    Myrtaceae   \n",
       "\n",
       "               genus                        scientificName specificEpithet  \\\n",
       "5      Callocephalon              Callocephalon fimbriatum      fimbriatum   \n",
       "159            Ninox                         Ninox strenua         strenua   \n",
       "326             Tyto                      Tyto tenebricosa     tenebricosa   \n",
       "356            Ninox                       Ninox connivens       connivens   \n",
       "382    Hoplocephalus             Hoplocephalus bungaroides     bungaroides   \n",
       "...              ...                                   ...             ...   \n",
       "21986    Anthosachne  Anthosachne kingiana subsp. kingiana        kingiana   \n",
       "22658      Caladenia                       Caladenia tensa           tensa   \n",
       "22659      Caladenia                  Caladenia atroclavia      atroclavia   \n",
       "23131   Prasophyllum                  Prasophyllum sandrae         sandrae   \n",
       "23133     Backhousia                Backhousia subargentea     subargentea   \n",
       "\n",
       "                    vernacularName    establishmentMeans  \\\n",
       "5               Gang-gang Cockatoo  Alive in NSW, Native   \n",
       "159                   Powerful Owl  Alive in NSW, Native   \n",
       "326                      Sooty Owl  Alive in NSW, Native   \n",
       "356                    Barking Owl  Alive in NSW, Native   \n",
       "382             Broad-headed Snake  Alive in NSW, Native   \n",
       "...                            ...                   ...   \n",
       "21986    Philip Island Wheat Grass  Alive in NSW, Native   \n",
       "22658          Rigid Spider-orchid  Alive in NSW, Native   \n",
       "22659  Black-clubbed Spider-orchid  Alive in NSW, Native   \n",
       "23131     Majors Creek Leek Orchid  Alive in NSW, Native   \n",
       "23133               Giant Ironwood  Alive in NSW, Native   \n",
       "\n",
       "                sourceStatus  protectedInNSW sensitivityClass  TSProfileID  \\\n",
       "5                 Vulnerable            True       Category 3      10975.0   \n",
       "159               Vulnerable            True       Category 3      10562.0   \n",
       "326               Vulnerable            True       Category 3      10821.0   \n",
       "356               Vulnerable            True       Category 3      10561.0   \n",
       "382               Endangered            True       Category 2      10413.0   \n",
       "...                      ...             ...              ...          ...   \n",
       "21986  Critically Endangered           False       Category 3      20148.0   \n",
       "22658             Not Listed            True       Category 2          NaN   \n",
       "22659             Not Listed            True       Category 2      20365.0   \n",
       "23131  Critically Endangered           False       Category 2      10668.0   \n",
       "23133             Endangered           False       Category 3      10166.0   \n",
       "\n",
       "         countryConservation               dcterms_modified  speciesID  \\\n",
       "5                 Endangered      2011-11-11T11:23:09+11:00          6   \n",
       "159               Not Listed      2011-11-11T11:23:06+11:00        162   \n",
       "326               Not Listed  2012-07-23T18:04:52.683+10:00        333   \n",
       "356               Not Listed      2011-11-11T11:23:06+11:00        363   \n",
       "382               Vulnerable      2011-11-11T11:23:04+11:00        390   \n",
       "...                      ...                            ...        ...   \n",
       "21986  Critically Endangered   2020-01-23T17:14:13.01+11:00      23014   \n",
       "22658             Not Listed   2020-06-23T19:56:43.85+10:00      23737   \n",
       "22659             Endangered  2020-06-23T19:51:08.733+10:00      23738   \n",
       "23131             Not Listed  2021-12-31T11:53:58.467+11:00      24268   \n",
       "23133             Not Listed  2021-03-17T14:00:40.507+11:00      24270   \n",
       "\n",
       "       taxonID generalisation                 status  \n",
       "5            6            1km             Vulnerable  \n",
       "159        162            1km             Vulnerable  \n",
       "326        333            1km             Vulnerable  \n",
       "356        363            1km             Vulnerable  \n",
       "382        390           10km             Endangered  \n",
       "...        ...            ...                    ...  \n",
       "21986    23014            1km  Critically Endangered  \n",
       "22658    23737           10km             Not Listed  \n",
       "22659    23738           10km             Not Listed  \n",
       "23131    24268           10km  Critically Endangered  \n",
       "23133    24270            1km             Endangered  \n",
       "\n",
       "[197 rows x 20 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5efdec00",
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitive.to_csv('sensitive-NSW.csv',index = False,encoding='utf-8-sig')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f34c7b7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.9.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
