{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9df0567f",
   "metadata": {},
   "source": [
    "# NCBI Genome RefSeq ingestion script\n",
    "\n",
    "This script reads the listing for the NCBI genome/refseq via the URL https://ftp.ncbi.nih.gov/genomes/refseq/assembly_summary_refseq.txt and attempts to map the columns to Darwin Core (DwC), using any of the DwC extenesions that might apply. More info in [this issue](https://github.com/AtlasOfLivingAustralia/arga-issues/issues/8)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e4998905",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_table(\"https://ftp.ncbi.nih.gov/genomes/refseq/assembly_summary_refseq.txt\", header=1, dtype = 'object', parse_dates=['seq_rel_date'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at the data types detected by pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a29aa8a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 261398 entries, 0 to 261397\n",
      "Data columns (total 33 columns):\n",
      " #   Column                          Non-Null Count   Dtype         \n",
      "---  ------                          --------------   -----         \n",
      " 0   geneticAccessionNumber          261398 non-null  object        \n",
      " 1   ncbi_bioproject                 261260 non-null  object        \n",
      " 2   ncbi_biosample                  249679 non-null  object        \n",
      " 3   ncbi_nuccore                    218468 non-null  object        \n",
      " 4   ncbi_refseq_category            261398 non-null  object        \n",
      " 5   taxonID                         261398 non-null  object        \n",
      " 6   specific_host                   261398 non-null  object        \n",
      " 7   scientificName                  261398 non-null  object        \n",
      " 8   cloneStrain                     242405 non-null  object        \n",
      " 9   source_mat_id                   17724 non-null   object        \n",
      " 10  ncbi_version_status             261398 non-null  object        \n",
      " 11  ncbi_assembly_level             261398 non-null  object        \n",
      " 12  ncbi_release_type               261398 non-null  object        \n",
      " 13  ncbi_genome_rep                 261398 non-null  object        \n",
      " 14  eventDate                       261398 non-null  datetime64[ns]\n",
      " 15  otherCatalogNumbers             261202 non-null  object        \n",
      " 16  redcordedBy                     249498 non-null  object        \n",
      " 17  ncbi_gbrs_paired_asm            261398 non-null  object        \n",
      " 18  ncbi_paired_asm_comp            261398 non-null  object        \n",
      " 19  geneticAccessionURI             261398 non-null  object        \n",
      " 20  ncbi_excluded_from_refseq       498 non-null     object        \n",
      " 21  ncbi_relation_to_type_material  27081 non-null   object        \n",
      " 22  ncbia_sm_not_live_date          261398 non-null  object        \n",
      " 23  ncbi_assembly_accession         261398 non-null  object        \n",
      " 24  ncbi_taxid                      261398 non-null  object        \n",
      " 25  ncbi_species_taxid              261398 non-null  object        \n",
      " 26  ncbi_organism_name              261398 non-null  object        \n",
      " 27  ncbi_infraspecific_name         242405 non-null  object        \n",
      " 28  ncbi_isolate                    17724 non-null   object        \n",
      " 29  ncbi_seq_rel_date               261398 non-null  datetime64[ns]\n",
      " 30  ncbi_asm_name                   261202 non-null  object        \n",
      " 31  ncbi_submitter                  249498 non-null  object        \n",
      " 32  ncbi_ftp_path                   261398 non-null  object        \n",
      "dtypes: datetime64[ns](2), object(31)\n",
      "memory usage: 65.8+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "53073a18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>paired_asm_comp</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>excluded_from_refseq</th>\n",
       "      <th>relation_to_type_material</th>\n",
       "      <th>asm_not_live_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>PRJNA164</td>\n",
       "      <td>SAMN02803731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2014-08-01</td>\n",
       "      <td>Release 6 plus ISO1 MT</td>\n",
       "      <td>The FlyBase Consortium/Berkeley Drosophila Gen...</td>\n",
       "      <td>GCA_000001215.4</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>PRJNA168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-02-03</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>GCA_000001405.29</td>\n",
       "      <td>different</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>PRJNA169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2020-06-24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>GCA_000001635.9</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>PRJNA116</td>\n",
       "      <td>SAMN03081427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2018-03-15</td>\n",
       "      <td>TAIR10.1</td>\n",
       "      <td>The Arabidopsis Information Resource (TAIR)</td>\n",
       "      <td>GCA_000001735.2</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>PRJNA70973</td>\n",
       "      <td>SAMN02953622</td>\n",
       "      <td>AAGU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2009-07-15</td>\n",
       "      <td>Loxafr3.0</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>GCA_000001905.1</td>\n",
       "      <td>different</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GCF_000007065.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>192952</td>\n",
       "      <td>2209</td>\n",
       "      <td>Methanosarcina mazei Go1</td>\n",
       "      <td>strain=Go1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2002-05-20</td>\n",
       "      <td>ASM706v1</td>\n",
       "      <td>Gottingen Genomics Laboratory</td>\n",
       "      <td>GCA_000007065.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>GCF_000007085.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>273068</td>\n",
       "      <td>911092</td>\n",
       "      <td>Caldanaerobacter subterraneus subsp. tengconge...</td>\n",
       "      <td>strain=MB4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2002-05-14</td>\n",
       "      <td>ASM708v1</td>\n",
       "      <td>Beijing Center.HGP</td>\n",
       "      <td>GCA_000007085.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>assembly from type material</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>GCF_000007105.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603579</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>264203</td>\n",
       "      <td>542</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821</td>\n",
       "      <td>strain=ZM4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2010-01-12</td>\n",
       "      <td>ASM710v1</td>\n",
       "      <td>Macrogen Inc.</td>\n",
       "      <td>GCA_000007105.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>GCF_000007125.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>224914</td>\n",
       "      <td>29459</td>\n",
       "      <td>Brucella melitensis bv. 1 str. 16M</td>\n",
       "      <td>strain=16M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2001-12-28</td>\n",
       "      <td>ASM712v1</td>\n",
       "      <td>Integrated Genomics</td>\n",
       "      <td>GCA_000007125.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>assembly from type material</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>GCF_000007145.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603845</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>190485</td>\n",
       "      <td>339</td>\n",
       "      <td>Xanthomonas campestris pv. campestris str. ATC...</td>\n",
       "      <td>strain=ATCC 33913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2002-05-24</td>\n",
       "      <td>ASM714v1</td>\n",
       "      <td>Sao Paulo state (Brazil) Consortium</td>\n",
       "      <td>GCA_000007145.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>assembly from type material</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   # assembly_accession   bioproject     biosample      wgs_master  \\\n",
       "0       GCF_000001215.4     PRJNA164  SAMN02803731             NaN   \n",
       "1      GCF_000001405.40     PRJNA168           NaN             NaN   \n",
       "2      GCF_000001635.27     PRJNA169           NaN             NaN   \n",
       "3       GCF_000001735.4     PRJNA116  SAMN03081427             NaN   \n",
       "4       GCF_000001905.1   PRJNA70973  SAMN02953622  AAGU00000000.3   \n",
       "..                  ...          ...           ...             ...   \n",
       "95      GCF_000007065.1  PRJNA224116  SAMN02603290             NaN   \n",
       "96      GCF_000007085.1  PRJNA224116  SAMN02603009             NaN   \n",
       "97      GCF_000007105.1  PRJNA224116  SAMN02603579             NaN   \n",
       "98      GCF_000007125.1  PRJNA224116  SAMN02603416             NaN   \n",
       "99      GCF_000007145.1  PRJNA224116  SAMN02603845             NaN   \n",
       "\n",
       "          refseq_category   taxid species_taxid  \\\n",
       "0        reference genome    7227          7227   \n",
       "1        reference genome    9606          9606   \n",
       "2        reference genome   10090         10090   \n",
       "3        reference genome    3702          3702   \n",
       "4   representative genome    9785          9785   \n",
       "..                    ...     ...           ...   \n",
       "95  representative genome  192952          2209   \n",
       "96  representative genome  273068        911092   \n",
       "97  representative genome  264203           542   \n",
       "98  representative genome  224914         29459   \n",
       "99                     na  190485           339   \n",
       "\n",
       "                                        organism_name infraspecific_name  \\\n",
       "0                             Drosophila melanogaster                NaN   \n",
       "1                                        Homo sapiens                NaN   \n",
       "2                                        Mus musculus                NaN   \n",
       "3                                Arabidopsis thaliana   ecotype=Columbia   \n",
       "4                                  Loxodonta africana                NaN   \n",
       "..                                                ...                ...   \n",
       "95                           Methanosarcina mazei Go1         strain=Go1   \n",
       "96  Caldanaerobacter subterraneus subsp. tengconge...         strain=MB4   \n",
       "97  Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821         strain=ZM4   \n",
       "98                 Brucella melitensis bv. 1 str. 16M         strain=16M   \n",
       "99  Xanthomonas campestris pv. campestris str. ATC...  strain=ATCC 33913   \n",
       "\n",
       "       isolate  ... genome_rep seq_rel_date                asm_name  \\\n",
       "0          NaN  ...       Full   2014-08-01  Release 6 plus ISO1 MT   \n",
       "1          NaN  ...       Full   2022-02-03              GRCh38.p14   \n",
       "2          NaN  ...       Full   2020-06-24                  GRCm39   \n",
       "3          NaN  ...       Full   2018-03-15                TAIR10.1   \n",
       "4   ISIS603380  ...       Full   2009-07-15               Loxafr3.0   \n",
       "..         ...  ...        ...          ...                     ...   \n",
       "95         NaN  ...       Full   2002-05-20                ASM706v1   \n",
       "96         NaN  ...       Full   2002-05-14                ASM708v1   \n",
       "97         NaN  ...       Full   2010-01-12                ASM710v1   \n",
       "98         NaN  ...       Full   2001-12-28                ASM712v1   \n",
       "99         NaN  ...       Full   2002-05-24                ASM714v1   \n",
       "\n",
       "                                            submitter   gbrs_paired_asm  \\\n",
       "0   The FlyBase Consortium/Berkeley Drosophila Gen...   GCA_000001215.4   \n",
       "1                         Genome Reference Consortium  GCA_000001405.29   \n",
       "2                         Genome Reference Consortium   GCA_000001635.9   \n",
       "3         The Arabidopsis Information Resource (TAIR)   GCA_000001735.2   \n",
       "4                                     Broad Institute   GCA_000001905.1   \n",
       "..                                                ...               ...   \n",
       "95                      Gottingen Genomics Laboratory   GCA_000007065.1   \n",
       "96                                 Beijing Center.HGP   GCA_000007085.1   \n",
       "97                                      Macrogen Inc.   GCA_000007105.1   \n",
       "98                                Integrated Genomics   GCA_000007125.1   \n",
       "99                Sao Paulo state (Brazil) Consortium   GCA_000007145.1   \n",
       "\n",
       "   paired_asm_comp                                           ftp_path  \\\n",
       "0        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "1        different  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "2        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "3        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "4        different  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "..             ...                                                ...   \n",
       "95       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "96       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "97       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "98       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "99       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "\n",
       "   excluded_from_refseq    relation_to_type_material asm_not_live_date  \n",
       "0                   NaN                          NaN                na  \n",
       "1                   NaN                          NaN                na  \n",
       "2                   NaN                          NaN                na  \n",
       "3                   NaN                          NaN                na  \n",
       "4                   NaN                          NaN                na  \n",
       "..                  ...                          ...               ...  \n",
       "95                  NaN                          NaN                na  \n",
       "96                  NaN  assembly from type material                na  \n",
       "97                  NaN                          NaN                na  \n",
       "98                  NaN  assembly from type material                na  \n",
       "99                  NaN  assembly from type material                na  \n",
       "\n",
       "[100 rows x 23 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b8c9c37",
   "metadata": {},
   "source": [
    "## Field mappings\n",
    "\n",
    "Attempt to map NCBI fields to known [DwC terms](https://dwc.tdwg.org/list/), including extensions\n",
    "\n",
    "NCBI fields described at https://ftp.ncbi.nih.gov/genomes/README_assembly_summary.txt\n",
    "\n",
    "| # | field name | field description |\n",
    "| ---- | ---- | ---- |\n",
    "| 1 | \"assembly_accession\" | Assembly accession: the assembly accession.version reported in this field is a unique identifier for the set of sequences in this particular version of the genome assembly. |\n",
    "| 2 | \"bioproject\" | BioProject: accession for the BioProject which produced the sequences in the genome assembly. A BioProject is a collection of biological data related to a single initiative, originating from a single organization or from a consortium. A BioProject record provides users a single place to find links to the diverse data types generated for that project. The record can be retrieved from the NCBI BioProject resource: https://www.ncbi.nlm.nih.gov/bioproject/ |\n",
    "| 3 | \"biosample\" | BioSample: accession for the BioSample from which the sequences in the genome assembly were obtained. A BioSample record contains a description of the biological source material used in experimental assays. The record can be retrieved from the NCBI BioSample resource: https://www.ncbi.nlm.nih.gov/biosample/ |\n",
    "| 4 | \"wgs_master\" | WGS-master: the GenBank Nucleotide accession and version for the master record of the Whole Genome Shotgun (WGS) project for the genome assembly. The master record can be retrieved from the NCBI Nucleotide resource: https://www.ncbi.nlm.nih.gov/nuccore Genome assemblies that are complete genomes, and those that are clone-based, do not have WGS-master records in which case this field will be empty. |\n",
    "| 5 | \"refseq_category\" | RefSeq Category: whether the assembly is a reference or representative genome in the NCBI Reference Sequence (RefSeq) project classification. Values: **reference genome** - a manually selected high quality genome assembly that NCBI and the community have identified as being important as a standard against which other data are compared; **representative genome** - a genome computationally or manually selected as a representative from among the best genomes available for a species or clade that does not have a designated reference genome; **na** - no RefSeq category assigned to this assembly Prokaryotes may have more than one reference or representative genome per species. For more information see: https://www.ncbi.nlm.nih.gov/refseq/about/prokaryotes/#referencegenome Eukaryotes have no more than one reference or representative genome per species. If there are no assemblies in RefSeq for a particular eukaryotic species, then the GenBank assembly that RefSeq would select as the best available for that species will be designated as the representative genome. Viruses may have one or more reference genomes per species. The representative genome designation is not applied to viruses. |\n",
    "| 6 | \"taxid\" | Taxonomy ID: the NCBI taxonomy identifier for the organism from which the genome assembly was derived. The NCBI Taxonomy Database is a curated classification and nomenclature for all of the organisms in the public sequence databases. The taxonomy record can be retrieved from the NCBI Taxonomy resource: https://www.ncbi.nlm.nih.gov/taxonomy/ |\n",
    "| 7 | \"species_taxid\" | Species taxonomy ID: the NCBI taxonomy identifier for the species from which the genome assembly was derived. The species taxid will differ from the organism taxid (column 6) only when the organism was reported at a sub- species or strain level. |\n",
    "| 8 | \"organism_name\" | Organism name: the scientific name of the organism from which the sequences in the genome assembly were derived. This name is taken from the NCBI Taxonomy record for the taxid specified in column 6. Some older taxids were assigned at the strain level and for these the organism name will include the strain. Current practice is only to assign taxids at the species level; for these the organism name will be just the species, however, the strain name will be reported in the infraspecific_name field (column 9). |\n",
    "| 9 | \"infraspecific_name\" | Infraspecific name: the strain, breed, cultivar or ecotype of the organism from which the sequences in the genome assembly were derived. Data are reported in the form tag=value, e.g. strain=AF16. Strain, breed, cultivar and ecotype are not expected to be used together, however, if they are then they will be reported in a list separated by \", /\". Empty if no strain, breed, cultivar or ecotype is specified on the genomic sequence records. |\n",
    "| 10 | \"isolate\" | Isolate: the individual isolate from which the sequences in the genome assembly were derived. Empty if no isolate is specified on the genomic sequence records. |\n",
    "| 11 | \"version_status\" | Version status: the release status for the genome assembly version. Values: **latest** - the most recent of all the versions for this assembly chain; **replaced** - this version has been replaced by a newer version of the assembly in the same chain; **suppressed** - this version of the assembly has been suppressed An assembly chain is the collection of all versions for the same assembly accession. |\n",
    "| 12 | \"assembly_level\" | Assembly level: the highest level of assembly for any object in the genome assembly. Values: **Complete genome** - all chromosomes are gapless and have no runs of 10 or more ambiguous bases (Ns), there are no unplaced or unlocalized scaffolds, and all the expected chromosomes are present (i.e. the assembly is not noted as having partial genome representation). Plasmids and organelles may or may not be included in the assembly but if present then the sequences are gapless. **Chromosome** - there is sequence for one or more chromosomes. This could be a completely sequenced chromosome without gaps or a chromosome containing scaffolds or contigs with gaps between them. There may also be unplaced or unlocalized scaffolds. Scaffold - some sequence contigs have been connected across gaps to create scaffolds, but the scaffolds are all unplaced or unlocalized. **Contig** - nothing is assembled beyond the level of sequence contigs |\n",
    "| 13 | \"release_type\" | Release type: whether this version of the genome assembly is a major, minor or patch release. Values: **Major** - changes from the previous assembly version result in a significant change to the coordinate system. The first version of an assembly is always a major release. Most subsequent genome assembly updates are also major releases. **Minor** - changes from the previous assembly version are limited to the following changes, none of which result in a significant change to the coordinate system of the primary assembly-unit: - adding, removing or changing a non-nuclear assembly-unit - dropping unplaced or unlocalized scaffolds - adding up to 50 unplaced or unlocalized scaffolds which are shorter than the current scaffold-N50 value - replacing a component with a gap of the same length Patch - the only change is the addition or modification of a patch assembly-unit. See the NCBI Assembly model web page (https://www.ncbi.nlm.nih.gov/assembly/ model/#asmb_def) for definitions of assembly-units and genome patches. |\n",
    "| 14 | \"genome_rep\" | Genome representation: whether the goal for the assembly was to represent the whole genome or only part of it. Values: **Full** - the data used to generate the assembly was obtained from the whole genome, as in Whole Genome Shotgun (WGS) assemblies for example. There may still be gaps in the assembly. **Partial** - the data used to generate the assembly came from only part of the genome. Most assemblies have full genome representation with a minority being partial genome representation. See the Assembly help web page (https://www.ncbi.nlm.nih.gov/assembly/help/) for reasons that the genome representation would be set to partial. |\n",
    "| 15 | \"seq_rel_date\" | Sequence release date: the date the sequences in the genome assembly were released in the International Nucleotide Sequence Database Collaboration (INSDC) databases, i.e. DDBJ, ENA or GenBank. |\n",
    "| 16 | \"asm_name\" | Assembly name: the submitter's name for the genome assembly, when one was provided, otherwise a default name, in the form ASM#####v#, is provided by NCBI. Assembly names are not unique. |\n",
    "| 17 | \"submitter\" | Submitter: the submitting consortium or first position if a list of organizations. The full submitter information is available in the NCBI BioProject resource: www.ncbi.nlm.nih.gov/bioproject/ |\n",
    "| 18 | \"gbrs_paired_asm\" | GenBank/RefSeq paired assembly: the accession.version of the GenBank assembly that is paired to the given RefSeq assembly, or vice-versa. \"na\" is reported if the assembly is unpaired. |\n",
    "| 19 | \"paired_asm_comp\" | Paired assembly comparison: whether the paired GenBank & RefSeq assemblies are identical or different. Values: **identical** - GenBank and RefSeq assemblies are identical; **different** - GenBank and RefSeq assemblies are not identical; **na** - not applicable since the assembly is unpaired |\n",
    "| 20 | \"ftp_path\" | FTP path: the path to the directory on the NCBI genomes FTP site from which data for this genome assembly can be downloaded. |\n",
    "| 21 | \"excluded_from_refseq\" | Excluded from RefSeq: reasons the assembly was excluded from the NCBI Reference Sequence (RefSeq) project, including any assembly anomalies. See: https://www.ncbi.nlm.nih.gov/assembly/help/anomnotrefseq/ |\n",
    "| 22 | \"relation_to_type_material\" | Relation to type material: contains a value if the sequences in the genome assembly were derived from type material. Values: **assembly from type material** - the sequences in the genome assembly were derived from type material; **assembly from synonym type material** - the sequences in the genome assembly were derived from synonym type material; **assembly from pathotype material** - the sequences in the genome assembly were derived from pathovar material;** **assembly designated as neotype** - the sequences in the genome assembly were derived from neotype material; **assembly designated as reftype** - the sequences in the genome assembly were derived from reference material where type material never was available and is not likely to ever be available ICTV species exemplar - the International Committee on Taxonomy of Viruses (ICTV) designated the genome assembly as the exemplar for the virus species ICTV additional isolate - the International Committee on Taxonomy of Viruses (ICTV) designated the genome assembly an additional isolate for the virus species |\n",
    "| 23 | \"asm_not_live_date\" | Assembly no longer live date: the date the assembly transitioned from version_status latest to either replaced or suppressed. When the assembly is in status latest, \"na\" is reported. |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "348ef5b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1:1 mapping to DwC and extension terms\n",
    "new_names = {\n",
    "  '# assembly_accession': 'geneticAccessionNumber',\n",
    "  'bioproject': 'ncbi_bioproject',\n",
    "  'biosample': 'ncbi_biosample',\n",
    "  'wgs_master': 'ncbi_nuccore',\n",
    "  'refseq_category': 'ncbi_refseq_category',\n",
    "  'taxid': 'taxonID',\n",
    "  'species_taxid': 'specific_host',\n",
    "  'organism_name': 'scientificName',\n",
    "  'infraspecific_name': 'cloneStrain',\n",
    "  'isolate': 'source_mat_id',\n",
    "  'version_status': 'ncbi_version_status',\n",
    "  'assembly_level': 'ncbi_assembly_level',\n",
    "  'release_type': 'ncbi_release_type',\n",
    "  'genome_rep': 'ncbi_genome_rep',\n",
    "  'seq_rel_date': 'eventDate',\n",
    "  'asm_name': 'otherCatalogNumbers',\n",
    "  'submitter': 'redcordedBy',\n",
    "  'gbrs_paired_asm': 'ncbi_gbrs_paired_asm',\n",
    "  'paired_asm_comp': 'ncbi_paired_asm_comp',\n",
    "  'ftp_path': 'geneticAccessionURI',\n",
    "  'excluded_from_refseq': 'ncbi_excluded_from_refseq',\n",
    "  'relation_to_type_material': 'ncbi_relation_to_type_material',\n",
    "  'asm_not_live_date': 'ncbia_sm_not_live_date'\n",
    "}\n",
    "\n",
    "# keep copies of DwC matched columns (just in case)\n",
    "df['ncbi_assembly_accession'] = df['# assembly_accession'] \n",
    "df['ncbi_taxid'] = df['taxid'] \n",
    "df['ncbi_species_taxid'] = df['species_taxid'] \n",
    "df['ncbi_organism_name'] = df['organism_name']\n",
    "df['ncbi_infraspecific_name'] = df['infraspecific_name']\n",
    "df['ncbi_isolate'] = df['isolate']\n",
    "df['ncbi_seq_rel_date'] = df['seq_rel_date'] \n",
    "df['ncbi_asm_name'] = df['asm_name']\n",
    "df['ncbi_submitter'] = df['submitter']\n",
    "df['ncbi_ftp_path'] = df['ftp_path']\n",
    "\n",
    "# do rename\n",
    "df.rename(\n",
    "  columns = new_names, \n",
    "  inplace = True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b56ad59b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 261398 entries, 0 to 261397\n",
      "Data columns (total 33 columns):\n",
      " #   Column                          Non-Null Count   Dtype         \n",
      "---  ------                          --------------   -----         \n",
      " 0   geneticAccessionNumber          261398 non-null  object        \n",
      " 1   ncbi_bioproject                 261260 non-null  object        \n",
      " 2   ncbi_biosample                  249679 non-null  object        \n",
      " 3   ncbi_nuccore                    218468 non-null  object        \n",
      " 4   ncbi_refseq_category            261398 non-null  object        \n",
      " 5   taxonID                         261398 non-null  object        \n",
      " 6   specific_host                   261398 non-null  object        \n",
      " 7   scientificName                  261398 non-null  object        \n",
      " 8   cloneStrain                     242405 non-null  object        \n",
      " 9   source_mat_id                   17724 non-null   object        \n",
      " 10  ncbi_version_status             261398 non-null  object        \n",
      " 11  ncbi_assembly_level             261398 non-null  object        \n",
      " 12  ncbi_release_type               261398 non-null  object        \n",
      " 13  ncbi_genome_rep                 261398 non-null  object        \n",
      " 14  eventDate                       261398 non-null  datetime64[ns]\n",
      " 15  otherCatalogNumbers             261202 non-null  object        \n",
      " 16  redcordedBy                     249498 non-null  object        \n",
      " 17  ncbi_gbrs_paired_asm            261398 non-null  object        \n",
      " 18  ncbi_paired_asm_comp            261398 non-null  object        \n",
      " 19  geneticAccessionURI             261398 non-null  object        \n",
      " 20  ncbi_excluded_from_refseq       498 non-null     object        \n",
      " 21  ncbi_relation_to_type_material  27081 non-null   object        \n",
      " 22  ncbia_sm_not_live_date          261398 non-null  object        \n",
      " 23  ncbi_assembly_accession         261398 non-null  object        \n",
      " 24  ncbi_taxid                      261398 non-null  object        \n",
      " 25  ncbi_species_taxid              261398 non-null  object        \n",
      " 26  ncbi_organism_name              261398 non-null  object        \n",
      " 27  ncbi_infraspecific_name         242405 non-null  object        \n",
      " 28  ncbi_isolate                    17724 non-null   object        \n",
      " 29  ncbi_seq_rel_date               261398 non-null  datetime64[ns]\n",
      " 30  ncbi_asm_name                   261202 non-null  object        \n",
      " 31  ncbi_submitter                  249498 non-null  object        \n",
      " 32  ncbi_ftp_path                   261398 non-null  object        \n",
      "dtypes: datetime64[ns](2), object(31)\n",
      "memory usage: 65.8+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "3ca73f5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneticAccessionNumber</th>\n",
       "      <th>ncbi_bioproject</th>\n",
       "      <th>ncbi_biosample</th>\n",
       "      <th>ncbi_nuccore</th>\n",
       "      <th>ncbi_refseq_category</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>specific_host</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>cloneStrain</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>...</th>\n",
       "      <th>ncbi_assembly_accession</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>ncbi_species_taxid</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "      <th>ncbi_infraspecific_name</th>\n",
       "      <th>ncbi_isolate</th>\n",
       "      <th>ncbi_seq_rel_date</th>\n",
       "      <th>ncbi_asm_name</th>\n",
       "      <th>ncbi_submitter</th>\n",
       "      <th>ncbi_ftp_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>PRJNA164</td>\n",
       "      <td>SAMN02803731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-08-01</td>\n",
       "      <td>Release 6 plus ISO1 MT</td>\n",
       "      <td>The FlyBase Consortium/Berkeley Drosophila Gen...</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>PRJNA168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-03</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>PRJNA169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-06-24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>PRJNA116</td>\n",
       "      <td>SAMN03081427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-03-15</td>\n",
       "      <td>TAIR10.1</td>\n",
       "      <td>The Arabidopsis Information Resource (TAIR)</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>PRJNA70973</td>\n",
       "      <td>SAMN02953622</td>\n",
       "      <td>AAGU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>2009-07-15</td>\n",
       "      <td>Loxafr3.0</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GCF_000007065.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>192952</td>\n",
       "      <td>2209</td>\n",
       "      <td>Methanosarcina mazei Go1</td>\n",
       "      <td>strain=Go1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000007065.1</td>\n",
       "      <td>192952</td>\n",
       "      <td>2209</td>\n",
       "      <td>Methanosarcina mazei Go1</td>\n",
       "      <td>strain=Go1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2002-05-20</td>\n",
       "      <td>ASM706v1</td>\n",
       "      <td>Gottingen Genomics Laboratory</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>GCF_000007085.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>273068</td>\n",
       "      <td>911092</td>\n",
       "      <td>Caldanaerobacter subterraneus subsp. tengconge...</td>\n",
       "      <td>strain=MB4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000007085.1</td>\n",
       "      <td>273068</td>\n",
       "      <td>911092</td>\n",
       "      <td>Caldanaerobacter subterraneus subsp. tengconge...</td>\n",
       "      <td>strain=MB4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2002-05-14</td>\n",
       "      <td>ASM708v1</td>\n",
       "      <td>Beijing Center.HGP</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>GCF_000007105.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603579</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>264203</td>\n",
       "      <td>542</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821</td>\n",
       "      <td>strain=ZM4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000007105.1</td>\n",
       "      <td>264203</td>\n",
       "      <td>542</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821</td>\n",
       "      <td>strain=ZM4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010-01-12</td>\n",
       "      <td>ASM710v1</td>\n",
       "      <td>Macrogen Inc.</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>GCF_000007125.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>224914</td>\n",
       "      <td>29459</td>\n",
       "      <td>Brucella melitensis bv. 1 str. 16M</td>\n",
       "      <td>strain=16M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000007125.1</td>\n",
       "      <td>224914</td>\n",
       "      <td>29459</td>\n",
       "      <td>Brucella melitensis bv. 1 str. 16M</td>\n",
       "      <td>strain=16M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2001-12-28</td>\n",
       "      <td>ASM712v1</td>\n",
       "      <td>Integrated Genomics</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>GCF_000007145.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603845</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>190485</td>\n",
       "      <td>339</td>\n",
       "      <td>Xanthomonas campestris pv. campestris str. ATC...</td>\n",
       "      <td>strain=ATCC 33913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000007145.1</td>\n",
       "      <td>190485</td>\n",
       "      <td>339</td>\n",
       "      <td>Xanthomonas campestris pv. campestris str. ATC...</td>\n",
       "      <td>strain=ATCC 33913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2002-05-24</td>\n",
       "      <td>ASM714v1</td>\n",
       "      <td>Sao Paulo state (Brazil) Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   geneticAccessionNumber ncbi_bioproject ncbi_biosample    ncbi_nuccore  \\\n",
       "0         GCF_000001215.4        PRJNA164   SAMN02803731             NaN   \n",
       "1        GCF_000001405.40        PRJNA168            NaN             NaN   \n",
       "2        GCF_000001635.27        PRJNA169            NaN             NaN   \n",
       "3         GCF_000001735.4        PRJNA116   SAMN03081427             NaN   \n",
       "4         GCF_000001905.1      PRJNA70973   SAMN02953622  AAGU00000000.3   \n",
       "..                    ...             ...            ...             ...   \n",
       "95        GCF_000007065.1     PRJNA224116   SAMN02603290             NaN   \n",
       "96        GCF_000007085.1     PRJNA224116   SAMN02603009             NaN   \n",
       "97        GCF_000007105.1     PRJNA224116   SAMN02603579             NaN   \n",
       "98        GCF_000007125.1     PRJNA224116   SAMN02603416             NaN   \n",
       "99        GCF_000007145.1     PRJNA224116   SAMN02603845             NaN   \n",
       "\n",
       "     ncbi_refseq_category taxonID specific_host  \\\n",
       "0        reference genome    7227          7227   \n",
       "1        reference genome    9606          9606   \n",
       "2        reference genome   10090         10090   \n",
       "3        reference genome    3702          3702   \n",
       "4   representative genome    9785          9785   \n",
       "..                    ...     ...           ...   \n",
       "95  representative genome  192952          2209   \n",
       "96  representative genome  273068        911092   \n",
       "97  representative genome  264203           542   \n",
       "98  representative genome  224914         29459   \n",
       "99                     na  190485           339   \n",
       "\n",
       "                                       scientificName        cloneStrain  \\\n",
       "0                             Drosophila melanogaster                NaN   \n",
       "1                                        Homo sapiens                NaN   \n",
       "2                                        Mus musculus                NaN   \n",
       "3                                Arabidopsis thaliana   ecotype=Columbia   \n",
       "4                                  Loxodonta africana                NaN   \n",
       "..                                                ...                ...   \n",
       "95                           Methanosarcina mazei Go1         strain=Go1   \n",
       "96  Caldanaerobacter subterraneus subsp. tengconge...         strain=MB4   \n",
       "97  Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821         strain=ZM4   \n",
       "98                 Brucella melitensis bv. 1 str. 16M         strain=16M   \n",
       "99  Xanthomonas campestris pv. campestris str. ATC...  strain=ATCC 33913   \n",
       "\n",
       "   source_mat_id  ... ncbi_assembly_accession ncbi_taxid ncbi_species_taxid  \\\n",
       "0            NaN  ...         GCF_000001215.4       7227               7227   \n",
       "1            NaN  ...        GCF_000001405.40       9606               9606   \n",
       "2            NaN  ...        GCF_000001635.27      10090              10090   \n",
       "3            NaN  ...         GCF_000001735.4       3702               3702   \n",
       "4     ISIS603380  ...         GCF_000001905.1       9785               9785   \n",
       "..           ...  ...                     ...        ...                ...   \n",
       "95           NaN  ...         GCF_000007065.1     192952               2209   \n",
       "96           NaN  ...         GCF_000007085.1     273068             911092   \n",
       "97           NaN  ...         GCF_000007105.1     264203                542   \n",
       "98           NaN  ...         GCF_000007125.1     224914              29459   \n",
       "99           NaN  ...         GCF_000007145.1     190485                339   \n",
       "\n",
       "                                   ncbi_organism_name ncbi_infraspecific_name  \\\n",
       "0                             Drosophila melanogaster                     NaN   \n",
       "1                                        Homo sapiens                     NaN   \n",
       "2                                        Mus musculus                     NaN   \n",
       "3                                Arabidopsis thaliana        ecotype=Columbia   \n",
       "4                                  Loxodonta africana                     NaN   \n",
       "..                                                ...                     ...   \n",
       "95                           Methanosarcina mazei Go1              strain=Go1   \n",
       "96  Caldanaerobacter subterraneus subsp. tengconge...              strain=MB4   \n",
       "97  Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821              strain=ZM4   \n",
       "98                 Brucella melitensis bv. 1 str. 16M              strain=16M   \n",
       "99  Xanthomonas campestris pv. campestris str. ATC...       strain=ATCC 33913   \n",
       "\n",
       "   ncbi_isolate ncbi_seq_rel_date           ncbi_asm_name  \\\n",
       "0           NaN        2014-08-01  Release 6 plus ISO1 MT   \n",
       "1           NaN        2022-02-03              GRCh38.p14   \n",
       "2           NaN        2020-06-24                  GRCm39   \n",
       "3           NaN        2018-03-15                TAIR10.1   \n",
       "4    ISIS603380        2009-07-15               Loxafr3.0   \n",
       "..          ...               ...                     ...   \n",
       "95          NaN        2002-05-20                ASM706v1   \n",
       "96          NaN        2002-05-14                ASM708v1   \n",
       "97          NaN        2010-01-12                ASM710v1   \n",
       "98          NaN        2001-12-28                ASM712v1   \n",
       "99          NaN        2002-05-24                ASM714v1   \n",
       "\n",
       "                                       ncbi_submitter  \\\n",
       "0   The FlyBase Consortium/Berkeley Drosophila Gen...   \n",
       "1                         Genome Reference Consortium   \n",
       "2                         Genome Reference Consortium   \n",
       "3         The Arabidopsis Information Resource (TAIR)   \n",
       "4                                     Broad Institute   \n",
       "..                                                ...   \n",
       "95                      Gottingen Genomics Laboratory   \n",
       "96                                 Beijing Center.HGP   \n",
       "97                                      Macrogen Inc.   \n",
       "98                                Integrated Genomics   \n",
       "99                Sao Paulo state (Brazil) Consortium   \n",
       "\n",
       "                                        ncbi_ftp_path  \n",
       "0   https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "1   https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "2   https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "3   https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "4   https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "..                                                ...  \n",
       "95  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "96  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "97  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "98  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "99  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "\n",
       "[100 rows x 33 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7ad5498d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../data/assembly_summary_refseq.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "42b29997",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sw/fck698gd1wv_56c640ksf4r0009m0h/T/ipykernel_17474/1373742857.py:1: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"../data/assembly_summary_refseq.csv\")\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"../data/assembly_summary_refseq.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "bed4d65e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneticAccessionNumber</th>\n",
       "      <th>ncbi_bioproject</th>\n",
       "      <th>ncbi_biosample</th>\n",
       "      <th>ncbi_nuccore</th>\n",
       "      <th>ncbi_refseq_category</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>specific_host</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>cloneStrain</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>...</th>\n",
       "      <th>ncbi_assembly_accession</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>ncbi_species_taxid</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "      <th>ncbi_infraspecific_name</th>\n",
       "      <th>ncbi_isolate</th>\n",
       "      <th>ncbi_seq_rel_date</th>\n",
       "      <th>ncbi_asm_name</th>\n",
       "      <th>ncbi_submitter</th>\n",
       "      <th>ncbi_ftp_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>PRJNA164</td>\n",
       "      <td>SAMN02803731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-08-01</td>\n",
       "      <td>Release 6 plus ISO1 MT</td>\n",
       "      <td>The FlyBase Consortium/Berkeley Drosophila Gen...</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>PRJNA168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-03</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>PRJNA169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-06-24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>PRJNA116</td>\n",
       "      <td>SAMN03081427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-03-15</td>\n",
       "      <td>TAIR10.1</td>\n",
       "      <td>The Arabidopsis Information Resource (TAIR)</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>PRJNA70973</td>\n",
       "      <td>SAMN02953622</td>\n",
       "      <td>AAGU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>2009-07-15</td>\n",
       "      <td>Loxafr3.0</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261393</th>\n",
       "      <td>GCF_936394065.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA13317838</td>\n",
       "      <td>CAKYAI000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1280</td>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>164625</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_936394065.1</td>\n",
       "      <td>1280</td>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>164625</td>\n",
       "      <td>2022-04-16</td>\n",
       "      <td>Assembly164625</td>\n",
       "      <td>istituto zooprofilattico sperimentale della lo...</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261394</th>\n",
       "      <td>GCF_937468385.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA10017782</td>\n",
       "      <td>CALBWS000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2880965</td>\n",
       "      <td>2880965</td>\n",
       "      <td>Neobacillus sp. JJ-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CIP111895</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_937468385.1</td>\n",
       "      <td>2880965</td>\n",
       "      <td>2880965</td>\n",
       "      <td>Neobacillus sp. JJ-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CIP111895</td>\n",
       "      <td>2022-05-02</td>\n",
       "      <td>JJ-3</td>\n",
       "      <td>Institut Pasteur</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261395</th>\n",
       "      <td>GCF_938192085.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14037638</td>\n",
       "      <td>CAKZRX000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fecal samples</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_938192085.1</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fecal samples</td>\n",
       "      <td>2022-05-02</td>\n",
       "      <td>LBC9d_2</td>\n",
       "      <td>umr1137</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261396</th>\n",
       "      <td>GCF_939535315.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA8919069</td>\n",
       "      <td>CALNEA000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=ROAR-440 / Onovel10:O92 / fimH1284 / 11...</td>\n",
       "      <td>Faeces</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_939535315.1</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=ROAR-440 / Onovel10:O92 / fimH1284 / 11...</td>\n",
       "      <td>Faeces</td>\n",
       "      <td>2022-05-03</td>\n",
       "      <td>ROAR-440</td>\n",
       "      <td>UMR 1137 IAME</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261397</th>\n",
       "      <td>GCF_939535325.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA8918993</td>\n",
       "      <td>CALNDZ000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Faeces</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_939535325.1</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Faeces</td>\n",
       "      <td>2022-05-03</td>\n",
       "      <td>ROAR-343</td>\n",
       "      <td>UMR 1137 IAME</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>261398 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       geneticAccessionNumber ncbi_bioproject ncbi_biosample  \\\n",
       "0             GCF_000001215.4        PRJNA164   SAMN02803731   \n",
       "1            GCF_000001405.40        PRJNA168            NaN   \n",
       "2            GCF_000001635.27        PRJNA169            NaN   \n",
       "3             GCF_000001735.4        PRJNA116   SAMN03081427   \n",
       "4             GCF_000001905.1      PRJNA70973   SAMN02953622   \n",
       "...                       ...             ...            ...   \n",
       "261393        GCF_936394065.1     PRJNA224116  SAMEA13317838   \n",
       "261394        GCF_937468385.1     PRJNA224116  SAMEA10017782   \n",
       "261395        GCF_938192085.1     PRJNA224116  SAMEA14037638   \n",
       "261396        GCF_939535315.1     PRJNA224116   SAMEA8919069   \n",
       "261397        GCF_939535325.1     PRJNA224116   SAMEA8918993   \n",
       "\n",
       "             ncbi_nuccore   ncbi_refseq_category  taxonID specific_host  \\\n",
       "0                     NaN       reference genome     7227          7227   \n",
       "1                     NaN       reference genome     9606          9606   \n",
       "2                     NaN       reference genome    10090         10090   \n",
       "3                     NaN       reference genome     3702          3702   \n",
       "4          AAGU00000000.3  representative genome     9785          9785   \n",
       "...                   ...                    ...      ...           ...   \n",
       "261393  CAKYAI000000000.1                     na     1280          1280   \n",
       "261394  CALBWS000000000.1                     na  2880965       2880965   \n",
       "261395  CAKZRX000000000.1                     na      562           562   \n",
       "261396  CALNEA000000000.1                     na      562           562   \n",
       "261397  CALNDZ000000000.1                     na      562           562   \n",
       "\n",
       "                 scientificName  \\\n",
       "0       Drosophila melanogaster   \n",
       "1                  Homo sapiens   \n",
       "2                  Mus musculus   \n",
       "3          Arabidopsis thaliana   \n",
       "4            Loxodonta africana   \n",
       "...                         ...   \n",
       "261393    Staphylococcus aureus   \n",
       "261394     Neobacillus sp. JJ-3   \n",
       "261395         Escherichia coli   \n",
       "261396         Escherichia coli   \n",
       "261397         Escherichia coli   \n",
       "\n",
       "                                              cloneStrain  source_mat_id  ...  \\\n",
       "0                                                     NaN            NaN  ...   \n",
       "1                                                     NaN            NaN  ...   \n",
       "2                                                     NaN            NaN  ...   \n",
       "3                                        ecotype=Columbia            NaN  ...   \n",
       "4                                                     NaN     ISIS603380  ...   \n",
       "...                                                   ...            ...  ...   \n",
       "261393                                                NaN         164625  ...   \n",
       "261394                                                NaN      CIP111895  ...   \n",
       "261395                                                NaN  Fecal samples  ...   \n",
       "261396  strain=ROAR-440 / Onovel10:O92 / fimH1284 / 11...         Faeces  ...   \n",
       "261397                                                NaN         Faeces  ...   \n",
       "\n",
       "       ncbi_assembly_accession ncbi_taxid ncbi_species_taxid  \\\n",
       "0              GCF_000001215.4       7227               7227   \n",
       "1             GCF_000001405.40       9606               9606   \n",
       "2             GCF_000001635.27      10090              10090   \n",
       "3              GCF_000001735.4       3702               3702   \n",
       "4              GCF_000001905.1       9785               9785   \n",
       "...                        ...        ...                ...   \n",
       "261393         GCF_936394065.1       1280               1280   \n",
       "261394         GCF_937468385.1    2880965            2880965   \n",
       "261395         GCF_938192085.1        562                562   \n",
       "261396         GCF_939535315.1        562                562   \n",
       "261397         GCF_939535325.1        562                562   \n",
       "\n",
       "             ncbi_organism_name  \\\n",
       "0       Drosophila melanogaster   \n",
       "1                  Homo sapiens   \n",
       "2                  Mus musculus   \n",
       "3          Arabidopsis thaliana   \n",
       "4            Loxodonta africana   \n",
       "...                         ...   \n",
       "261393    Staphylococcus aureus   \n",
       "261394     Neobacillus sp. JJ-3   \n",
       "261395         Escherichia coli   \n",
       "261396         Escherichia coli   \n",
       "261397         Escherichia coli   \n",
       "\n",
       "                                  ncbi_infraspecific_name   ncbi_isolate  \\\n",
       "0                                                     NaN            NaN   \n",
       "1                                                     NaN            NaN   \n",
       "2                                                     NaN            NaN   \n",
       "3                                        ecotype=Columbia            NaN   \n",
       "4                                                     NaN     ISIS603380   \n",
       "...                                                   ...            ...   \n",
       "261393                                                NaN         164625   \n",
       "261394                                                NaN      CIP111895   \n",
       "261395                                                NaN  Fecal samples   \n",
       "261396  strain=ROAR-440 / Onovel10:O92 / fimH1284 / 11...         Faeces   \n",
       "261397                                                NaN         Faeces   \n",
       "\n",
       "       ncbi_seq_rel_date           ncbi_asm_name  \\\n",
       "0             2014-08-01  Release 6 plus ISO1 MT   \n",
       "1             2022-02-03              GRCh38.p14   \n",
       "2             2020-06-24                  GRCm39   \n",
       "3             2018-03-15                TAIR10.1   \n",
       "4             2009-07-15               Loxafr3.0   \n",
       "...                  ...                     ...   \n",
       "261393        2022-04-16          Assembly164625   \n",
       "261394        2022-05-02                    JJ-3   \n",
       "261395        2022-05-02                 LBC9d_2   \n",
       "261396        2022-05-03                ROAR-440   \n",
       "261397        2022-05-03                ROAR-343   \n",
       "\n",
       "                                           ncbi_submitter  \\\n",
       "0       The FlyBase Consortium/Berkeley Drosophila Gen...   \n",
       "1                             Genome Reference Consortium   \n",
       "2                             Genome Reference Consortium   \n",
       "3             The Arabidopsis Information Resource (TAIR)   \n",
       "4                                         Broad Institute   \n",
       "...                                                   ...   \n",
       "261393  istituto zooprofilattico sperimentale della lo...   \n",
       "261394                                   Institut Pasteur   \n",
       "261395                                            umr1137   \n",
       "261396                                      UMR 1137 IAME   \n",
       "261397                                      UMR 1137 IAME   \n",
       "\n",
       "                                            ncbi_ftp_path  \n",
       "0       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "2       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "3       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "4       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "...                                                   ...  \n",
       "261393  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "261394  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "261395  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "261396  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "261397  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...  \n",
       "\n",
       "[261398 rows x 33 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0c3259c",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.9.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
